<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.552">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>MED3007 V25 - R Lab (day 4): Clustering</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">MED3007 V25</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../get_started/get_started.html"> 
<span class="menu-text">Get Started</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../lab/overview.html"> 
<span class="menu-text">Lab and Code</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/ocbe-uio/course_med3007"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#exercise-1-food" id="toc-exercise-1-food" class="nav-link active" data-scroll-target="#exercise-1-food">Exercise 1: Food</a>
  <ul class="collapse">
  <li><a href="#distances" id="toc-distances" class="nav-link" data-scroll-target="#distances">Distances</a></li>
  <li><a href="#hierarchical-clustering" id="toc-hierarchical-clustering" class="nav-link" data-scroll-target="#hierarchical-clustering">Hierarchical clustering</a></li>
  <li><a href="#linkage-dissimilarity-scaling" id="toc-linkage-dissimilarity-scaling" class="nav-link" data-scroll-target="#linkage-dissimilarity-scaling">Linkage, dissimilarity, scaling</a></li>
  <li><a href="#heatmap" id="toc-heatmap" class="nav-link" data-scroll-target="#heatmap">Heatmap</a></li>
  </ul></li>
  <li><a href="#exercise-2-nci60" id="toc-exercise-2-nci60" class="nav-link" data-scroll-target="#exercise-2-nci60">Exercise 2: NCI60</a>
  <ul class="collapse">
  <li><a href="#hierarchical-clustering-1" id="toc-hierarchical-clustering-1" class="nav-link" data-scroll-target="#hierarchical-clustering-1">Hierarchical clustering</a></li>
  <li><a href="#k-means-clustering" id="toc-k-means-clustering" class="nav-link" data-scroll-target="#k-means-clustering">K-means clustering</a></li>
  <li><a href="#compare-with-hierarchical-clustering" id="toc-compare-with-hierarchical-clustering" class="nav-link" data-scroll-target="#compare-with-hierarchical-clustering">Compare with hierarchical clustering</a></li>
  <li><a href="#visualize-clusters" id="toc-visualize-clusters" class="nav-link" data-scroll-target="#visualize-clusters">Visualize clusters</a></li>
  <li><a href="#heatmap-1" id="toc-heatmap-1" class="nav-link" data-scroll-target="#heatmap-1">Heatmap</a></li>
  </ul></li>
  <li><a href="#exercise-3-gene-expression-data" id="toc-exercise-3-gene-expression-data" class="nav-link" data-scroll-target="#exercise-3-gene-expression-data">Exercise 3: Gene expression data</a>
  <ul class="collapse">
  <li><a href="#hierarchical-clustering-2" id="toc-hierarchical-clustering-2" class="nav-link" data-scroll-target="#hierarchical-clustering-2">Hierarchical clustering</a></li>
  <li><a href="#k-means" id="toc-k-means" class="nav-link" data-scroll-target="#k-means">K-means</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">R Lab (day 4): Clustering</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>Download datasets <a href="https://github.com/ocbe-uio/course_med3007/tree/main/lab/data">here</a>, or from Canvas.</p>
<p>R script: <a href="https://github.com/ocbe-uio/course_med3007/blob/main/lab/code/MED3007_Lab3_clustering.R">Code</a></p>
<p>Presentation (2024 version by Chi): <a href="presentation/Lab_clustering.pdf">Slides</a></p>
<section id="exercise-1-food" class="level2">
<h2 class="anchored" data-anchor-id="exercise-1-food">Exercise 1: Food</h2>
<p>We use the same <code>Food.txt</code> data to illustrate two concepts: hierarchical clustering, and heatmap.</p>
<p>This is not a genomics dataset, but for the ease of interpretability, we use it for teaching purposes.</p>
<p>Let us load the dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>food <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">'data/Food.txt'</span>, <span class="at">header=</span>T)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co"># we change the name from pulses to a more common name, legume</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(food)[<span class="dv">7</span>] <span class="ot">&lt;-</span> <span class="st">'Legume'</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(food) <span class="co"># print first 6 lines </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               Meat Pigs Eggs Milk Fish Cereals Legume Fruit
Albania        10.1  1.4  0.5  8.9  0.2    42.3    5.5   1.7
Austria         8.9 14.0  4.3 19.9  2.1    28.0    1.3   4.3
Belg.Lux.      13.5  9.3  4.1 17.5  4.5    26.6    2.1   4.0
Bulgaria        7.8  6.0  1.6  8.3  1.2    56.7    3.7   4.2
Czechoslovakia  9.7 11.4  2.8 12.5  2.0    34.3    1.1   4.0
Denmark        10.6 10.8  3.7 25.0  9.9    21.9    0.7   2.4</code></pre>
</div>
</div>
<p>We <strong>scale</strong> the data (also called standardize, or normalize sometimes) so that each column (feature, variable) has 0 mean and 1 variance. We call the scaled data <code>food_s</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>food_s <span class="ot">&lt;-</span> <span class="fu">scale</span>(food)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(food_s) <span class="co"># print first 6 lines</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                      Meat       Pigs       Eggs        Milk        Fish
Albania         0.08126490 -1.8299828 -2.2437259 -1.15570645 -1.20028213
Austria        -0.27725673  1.6636208  1.2335002  0.39161231 -0.64187467
Belg.Lux.       1.09707621  0.3604512  1.0504883  0.05401549  0.06348211
Bulgaria       -0.60590157 -0.5545403 -1.2371605 -1.24010566 -0.90638347
Czechoslovakia -0.03824231  0.9427184 -0.1390890 -0.64931122 -0.67126454
Denmark         0.23064892  0.7763564  0.6844645  1.10900556  1.65053488
                  Cereals     Legume       Fruit
Albania         0.9159176  1.2227536 -1.35040507
Austria        -0.3870690 -0.8923886  0.09091397
Belg.Lux.      -0.5146342 -0.4895043 -0.07539207
Bulgaria        2.2280161  0.3162641  0.03547862
Czechoslovakia  0.1869740 -0.9931096 -0.07539207
Denmark        -0.9428885 -1.1945517 -0.96235764</code></pre>
</div>
</div>
<section id="distances" class="level3">
<h3 class="anchored" data-anchor-id="distances">Distances</h3>
<p>To do hierarchical clustering, the most convenient command is <code>hclust()</code>. As input you would need a <strong>distance</strong> between the subjects (patients, or countries in this example). We do it on the scaled data.</p>
<p>The command to compute pair-wise distance is <code>dist()</code>. By default, the distance being computed is the Euclidean distance (<em>details optional</em>). Euclidean distance is possibly the most commonly used metric, but there are others. See <code>?dist()</code> to find out more options.</p>
<p>We can present the pair-wise distances in a matrix format. You can see that this matrix is symmetric, with 0 on the diagonal - this should be intuitive: the distance between A - B is the same as B - A, and the distance between A and itself is 0.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># compute distance</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>food_dist <span class="ot">&lt;-</span> <span class="fu">dist</span>(food_s)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co"># round(food_dist, digits = 2) # try this yourself to see what it does</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># alternatively, look at this as a matrix</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>food_dist_matrix <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(food_dist)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(food_dist_matrix[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>], <span class="at">digits =</span> <span class="dv">2</span>) <span class="co"># first 5 row 5 col</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               Albania Austria Belg.Lux. Bulgaria Czechoslovakia
Albania           0.00    5.95      5.13     2.77           4.44
Austria           5.95    0.00      2.11     4.71           1.98
Belg.Lux.         5.13    2.11      0.00     4.45           2.20
Bulgaria          2.77    4.71      4.45     0.00           3.17
Czechoslovakia    4.44    1.98      2.20     3.17           0.00</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Optional: Euclidean disance
</div>
</div>
<div class="callout-body-container callout-body">
<p>You can check the Euclidean distance between Albania and Austria is indeed 5.95. This distance is the square root of the sum of squared differences between two subjects in all their measurements.</p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(food_s[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>,], <span class="at">digits =</span> <span class="dv">2</span>) <span class="co"># we only keep first 2 digits</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Meat  Pigs  Eggs  Milk  Fish Cereals Legume Fruit
Albania  0.08 -1.83 -2.24 -1.16 -1.20    0.92   1.22 -1.35
Austria -0.28  1.66  1.23  0.39 -0.64   -0.39  -0.89  0.09</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># take the data for two countries each </span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>albania <span class="ot">&lt;-</span> <span class="fu">round</span>(food_s[<span class="dv">1</span>,], <span class="at">digits =</span> <span class="dv">2</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>austria <span class="ot">&lt;-</span> <span class="fu">round</span>(food_s[<span class="dv">2</span>,], <span class="at">digits =</span> <span class="dv">2</span>)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co"># compute difference between each col</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> albania <span class="sc">-</span> austria</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>d</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Meat    Pigs    Eggs    Milk    Fish Cereals  Legume   Fruit 
   0.36   -3.49   -3.47   -1.55   -0.56    1.31    2.11   -1.44 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># euclidean distance: square each element, sum together, and take a square root</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sqrt</span>(<span class="fu">sum</span>(d<span class="sc">^</span><span class="dv">2</span>)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 5.942096</code></pre>
</div>
</div>
</section>
<section id="hierarchical-clustering" class="level3">
<h3 class="anchored" data-anchor-id="hierarchical-clustering">Hierarchical clustering</h3>
<p>Now that we have computed the distance <code>food_dist</code>, we plug it in the clustering algorithm, <code>hclust()</code>.</p>
<p>We try the complete linkage method, by specifying <code>method = 'complete'</code>. The result is saved as <code>hc.complete</code>. You can visualize it, and add label of the country names to make it easier to read.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>hc.complete <span class="ot">&lt;-</span> <span class="fu">hclust</span>(food_dist, <span class="at">method=</span><span class="st">"complete"</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc.complete, <span class="at">labels=</span><span class="fu">rownames</span>(food), <span class="at">main=</span><span class="st">"Complete Linkage"</span>, <span class="at">xlab=</span><span class="st">""</span>, <span class="at">sub=</span><span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day4_clustering_files/figure-html/hc-food-complete-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="linkage-dissimilarity-scaling" class="level3">
<h3 class="anchored" data-anchor-id="linkage-dissimilarity-scaling">Linkage, dissimilarity, scaling</h3>
<p>Hierarchical clustering is a class of methods, and there are a variety of options to set.</p>
<ul>
<li>Linkage (by seting <code>method</code> inside <code>hclust()</code>): complete, single, average</li>
<li>Dissimilarity: Euclidean, correlation, …</li>
<li>Scaling: scaled data (mean 0 variance 1) or unscaled, original data</li>
</ul>
<p>There is no definite guide on which combination works the best, hence you can try them out and see what could make most sense. Again, in unsupervised learning data do not have outcome labels, so the interpretation is left for the domain experts to make.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># single linkage</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>hc.single <span class="ot">&lt;-</span> <span class="fu">hclust</span>(food_dist, <span class="at">method=</span><span class="st">"single"</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc.single, <span class="at">labels=</span><span class="fu">rownames</span>(food), <span class="at">main=</span><span class="st">"Single Linkage"</span>, <span class="at">xlab=</span><span class="st">""</span>, <span class="at">sub=</span><span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day4_clustering_files/figure-html/hc-food-linkage-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># average linkage</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>hc.average <span class="ot">&lt;-</span> <span class="fu">hclust</span>(food_dist, <span class="at">method=</span><span class="st">"average"</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc.average, <span class="at">labels=</span><span class="fu">rownames</span>(food), <span class="at">main=</span><span class="st">"Average Linkage"</span>, <span class="at">xlab=</span><span class="st">""</span>, <span class="at">sub=</span><span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day4_clustering_files/figure-html/hc-food-linkage-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># unscaled data, complete linkage</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>hc.unscaled <span class="ot">&lt;-</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(food), <span class="at">method=</span><span class="st">"complete"</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc.unscaled, <span class="at">labels=</span><span class="fu">rownames</span>(food), <span class="at">main=</span><span class="st">"Complete linkage with unscaled features"</span>, <span class="at">xlab=</span><span class="st">""</span>, <span class="at">sub=</span><span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day4_clustering_files/figure-html/hc-food-unscaled-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># correlation as dissimiarity, rather than euclidean distance</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>dd <span class="ot">&lt;-</span> <span class="fu">as.dist</span>(<span class="dv">1</span><span class="sc">-</span><span class="fu">cor</span>(<span class="fu">t</span>(food_s))) <span class="co"># compute the metric</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>hc.corr <span class="ot">&lt;-</span> <span class="fu">hclust</span>(dd, <span class="at">method=</span><span class="st">"complete"</span>) <span class="co"># cluster</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc.corr, <span class="at">labels=</span><span class="fu">rownames</span>(food), <span class="at">main=</span><span class="st">"Complete linkage with correlation-based distance"</span>, <span class="at">xlab=</span><span class="st">""</span>, <span class="at">sub=</span><span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day4_clustering_files/figure-html/hc-food-cor-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="heatmap" class="level3">
<h3 class="anchored" data-anchor-id="heatmap">Heatmap</h3>
<p>Heatmap is a visualization tool to plot data of similar values in similar colors, so that you can identify visualy if there is any pattern. It can also be combined with hierarchical clustering - this is actually the default outcome: dendrograms are displayed for both rows and column.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make heatmap on the scaled data</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(food_s)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day4_clustering_files/figure-html/hc-food-heatmap1-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>To preserve the original ordering of the columns and rows, you can specify <code>Rowv = NA, Colv = NA</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># no clustering for row or col, this preserves the original ordering</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(food_s, <span class="at">Rowv =</span> <span class="cn">NA</span>, <span class="at">Colv =</span> <span class="cn">NA</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day4_clustering_files/figure-html/hc-food-heatmap2-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Can also only do clustering for row only (or column only).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># only clustering for row</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(food_s, <span class="at">Colv =</span> <span class="cn">NA</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day4_clustering_files/figure-html/hc-food-heatmap3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="exercise-2-nci60" class="level2">
<h2 class="anchored" data-anchor-id="exercise-2-nci60">Exercise 2: NCI60</h2>
<p>We look at the NCI60 data again. First load the dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ISLR)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="co"># or, load('data/NCI60.RData')</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>nci.labs <span class="ot">&lt;-</span> NCI60<span class="sc">$</span>labs <span class="co"># Sample labels (tissue type)</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>nci.data <span class="ot">&lt;-</span> NCI60<span class="sc">$</span>data <span class="co"># Gene expression data set</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="hierarchical-clustering-1" class="level3">
<h3 class="anchored" data-anchor-id="hierarchical-clustering-1">Hierarchical clustering</h3>
<p>We start by scaling the data, and calculate the distance matrix (using the Euclidean distance), and then investigate different linkage methods.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Scale the data to zero mean and unit variance:</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>sd.data <span class="ot">&lt;-</span> <span class="fu">scale</span>(nci.data)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the distance matrix </span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="co"># equivalent: data.dist &lt;- dist(sd.data, method="euclidean")</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>data.dist <span class="ot">&lt;-</span> <span class="fu">dist</span>(sd.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next we perform hierarchical clustering with distance matrix as input. The function we use is <code>hclust()</code>. We specify the linkage method to be <code>complete</code>.</p>
<p>Once the result is saved in <code>hc.complete</code> object, you can plot the dendrogram.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform clustering</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>hc.complete <span class="ot">&lt;-</span> <span class="fu">hclust</span>(data.dist, <span class="at">method=</span><span class="st">"complete"</span>)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="co"># names(hc.complete)</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc.complete, <span class="at">labels=</span>nci.labs, <span class="at">main=</span><span class="st">"Complete Linkage"</span>, <span class="at">xlab=</span><span class="st">""</span>, <span class="at">sub=</span><span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day4_clustering_files/figure-html/clust-nci60-hccomplete-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The object <code>hc.complete</code> contains a lot of information. To get the information, you can use the <code>$</code> operator.</p>
<p>You should refer to the documentation for <code>hclust()</code> to see a complete list of output. Use <code>?hclust</code> to get the documentation on how to use the function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>hc.complete<span class="sc">$</span>dist.method <span class="co"># distance method</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "euclidean"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># hc.complete$merge  # order of aggregations of samples / clusters</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="co"># hc.complete$height # distance at which aggregations happen</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="co"># hc.complete$labels # labels (numeric, since we don't know the original categories!)</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="co"># hc.complete$method</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="co"># hc.complete$call</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can try different linkage methods and see how the clustering results differ. Change the <code>method</code> argument in the function, and plot the results.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>hc.average <span class="ot">&lt;-</span> <span class="fu">hclust</span>(data.dist, <span class="at">method=</span><span class="st">"average"</span>)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>hc.single <span class="ot">&lt;-</span> <span class="fu">hclust</span>(data.dist, <span class="at">method=</span><span class="st">"single"</span>)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc.average, <span class="at">labels=</span>nci.labs, <span class="at">main=</span><span class="st">"Average Linkage"</span>, <span class="at">xlab=</span><span class="st">""</span>, <span class="at">sub=</span><span class="st">""</span>)</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc.single, <span class="at">labels=</span>nci.labs,  <span class="at">main=</span><span class="st">"Single Linkage"</span>, <span class="at">xlab=</span><span class="st">""</span>, <span class="at">sub=</span><span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we focus on <strong>complete linkage</strong> only.</p>
<p>First, we use <code>cutree()</code> to compare the results when the data are separated into either 2 or 4 clusters.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare 2 clusters and 4 clusters:</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>hc.clusters <span class="ot">&lt;-</span> <span class="fu">cutree</span>(hc.complete, <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">4</span>))</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(hc.clusters) <span class="co"># print first 6 results</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   2 4
V1 1 1
V2 1 1
V3 1 1
V4 1 1
V5 1 2
V6 1 2</code></pre>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># cross tabulation</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(hc.clusters[,<span class="st">"2"</span>], hc.clusters[,<span class="st">"4"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   
     1  2  3  4
  1 40  7  0  0
  2  0  0  8  9</code></pre>
</div>
</div>
<p>It is more straightforward to check the results with a dendrogram.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># visualize the cuts</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="co"># how do you know where to draw the line? check height</span></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>heights <span class="ot">&lt;-</span> hc.complete<span class="sc">$</span>height</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(heights, <span class="dv">4</span>)  <span class="co"># print the last 4</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 137.5633 141.2472 142.9218 162.2074</code></pre>
</div>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc.complete, <span class="at">labels=</span>nci.labs, <span class="at">main=</span><span class="st">"Complete Linkage"</span>, <span class="at">xlab=</span><span class="st">""</span>, <span class="at">sub=</span><span class="st">""</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h=</span><span class="dv">140</span>, <span class="at">col=</span><span class="st">"red"</span>)  <span class="co"># 4 clusters</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h=</span><span class="dv">150</span>, <span class="at">col=</span><span class="st">"blue"</span>) <span class="co"># 2 clusters</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day4_clustering_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The way to interpret the height variable is simple: it is where two clusters are merged into one. For example, the largest cluster corresponds to the last value of <code>height</code> (162.2) - if you check the figure, it is exactly where the horizontal line is merging the two groups. Similarly, 142.9 is where three groups became two, 141.2 is where four groups became three. If you draw a line at 140, it points out the four clusters.</p>
<p>How are the labels distributed between clusters? We can focus on 4 cluster situation, and use <code>table()</code> to list out which cancer types is merged in which of the four clusters.</p>
<p>For example, breast cancer appears in all but 3rd cluster; melanoma only appears in the first clsuter; so on so forth.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(hc.clusters[,<span class="st">"4"</span>], nci.labs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   nci.labs
    BREAST CNS COLON K562A-repro K562B-repro LEUKEMIA MCF7A-repro MCF7D-repro
  1      2   3     2           0           0        0           0           0
  2      3   2     0           0           0        0           0           0
  3      0   0     0           1           1        6           0           0
  4      2   0     5           0           0        0           1           1
   nci.labs
    MELANOMA NSCLC OVARIAN PROSTATE RENAL UNKNOWN
  1        8     8       6        2     8       1
  2        0     1       0        0     1       0
  3        0     0       0        0     0       0
  4        0     0       0        0     0       0</code></pre>
</div>
</div>
<p>Finally, we see what happens if we use <strong>unscaled data</strong> instead of scaled data, or if we use a <strong>correlation-based distance metric</strong> instead of the Euclidean distance.</p>
<p>Compare the dendrograms: How different are the resulting clusterings? Do you recognise subclusters that are consistent?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare scaled data versus non-scaled data:</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>hc.unscaled <span class="ot">&lt;-</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(nci.data), <span class="at">method=</span><span class="st">"complete"</span>)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc.unscaled, <span class="at">labels=</span>nci.labs, <span class="at">main=</span><span class="st">"Complete linkage with unscaled features"</span>, <span class="at">xlab=</span><span class="st">""</span>, <span class="at">sub=</span><span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day4_clustering_files/figure-html/clust-nci60-unscale-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare Euclidean distance with correlation-based distance:</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>dd <span class="ot">&lt;-</span> <span class="fu">as.dist</span>(<span class="dv">1</span><span class="sc">-</span><span class="fu">cor</span>(<span class="fu">t</span>(sd.data)))</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>hc.corr <span class="ot">&lt;-</span> <span class="fu">hclust</span>(dd, <span class="at">method=</span><span class="st">"complete"</span>)</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc.corr, <span class="at">labels=</span>nci.labs, <span class="at">main=</span><span class="st">"Complete linkage with correlation-based distance"</span>, <span class="at">xlab=</span><span class="st">""</span>, <span class="at">sub=</span><span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day4_clustering_files/figure-html/clust-nci60-unscale-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="k-means-clustering" class="level3">
<h3 class="anchored" data-anchor-id="k-means-clustering">K-means clustering</h3>
<p>In this section we explore the K-means clustering on the same dataset.</p>
<p>In contrast to the hierarchical clustering which requires a <strong>distance</strong> as input, with K-means you would provide the data matrix. The data matrix can be scaled (centered and with unit variance), or unscaled.</p>
<p>In this example we use scaled data computed from before, <code>sd.data</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">4</span>) <span class="co"># set random seed</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>km.out4 <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(sd.data, <span class="at">centers =</span> <span class="dv">4</span>, <span class="at">nstart=</span><span class="dv">20</span>)</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>km.out4<span class="sc">$</span>cluster</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> V1  V2  V3  V4  V5  V6  V7  V8  V9 V10 V11 V12 V13 V14 V15 V16 V17 V18 V19 V20 
  4   4   4   4   4   4   4   4   4   4   4   4   4   4   4   4   4   4   4   4 
V21 V22 V23 V24 V25 V26 V27 V28 V29 V30 V31 V32 V33 V34 V35 V36 V37 V38 V39 V40 
  4   4   4   1   1   4   1   4   1   4   4   1   1   3   3   3   3   3   3   3 
V41 V42 V43 V44 V45 V46 V47 V48 V49 V50 V51 V52 V53 V54 V55 V56 V57 V58 V59 V60 
  3   1   1   1   1   1   1   1   1   1   1   1   1   1   1   2   2   2   2   2 
V61 V62 V63 V64 
  2   2   2   2 </code></pre>
</div>
</div>
<p>Read the help file <code>?kmeans</code> to understand what the argument nstart=20 does. Comparing an analysis with nstart=20 versus nstart=1 demonstrates how the cluster results can be improved if we allow more evaluations with different randomly chosen starting centroids.</p>
<p>Set a different random seed, say 3 (as long as it’s different from the one you used before), and run the analysis again. This time we use a different <code>nstart</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># different starting centroids improve the clustering:</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">3</span>)</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>km.out <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(sd.data, <span class="at">centers =</span> <span class="dv">4</span>, <span class="at">nstart=</span><span class="dv">1</span>)</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>km.out<span class="sc">$</span>cluster <span class="co"># cluster label</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> V1  V2  V3  V4  V5  V6  V7  V8  V9 V10 V11 V12 V13 V14 V15 V16 V17 V18 V19 V20 
  1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1 
V21 V22 V23 V24 V25 V26 V27 V28 V29 V30 V31 V32 V33 V34 V35 V36 V37 V38 V39 V40 
  1   1   1   3   3   1   3   1   3   1   1   3   3   4   4   4   4   4   4   4 
V41 V42 V43 V44 V45 V46 V47 V48 V49 V50 V51 V52 V53 V54 V55 V56 V57 V58 V59 V60 
  4   3   3   3   3   3   3   3   3   3   3   3   3   3   3   2   2   2   2   2 
V61 V62 V63 V64 
  2   2   2   2 </code></pre>
</div>
</div>
</section>
<section id="compare-with-hierarchical-clustering" class="level3">
<h3 class="anchored" data-anchor-id="compare-with-hierarchical-clustering">Compare with hierarchical clustering</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># we can directly compare the k-means result (along rows)</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="co"># with the hierarchical clustering result (along columns)</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(km.out4<span class="sc">$</span>cluster, hc.clusters[,<span class="st">"4"</span>], <span class="at">deparse.level=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               hc.clusters[, "4"]
km.out4$cluster  1  2  3  4
              1 11  0  0  9
              2  9  0  0  0
              3  0  0  8  0
              4 20  7  0  0</code></pre>
</div>
</div>
<p>From the results, you can see that the results are slightly different between the two methods. Keep in mind that in unsupervised learning you do not have the real outcome label (such as the cancer types here), so you need to try a few different methods and compare the outputs, and make interpretations accordingly.</p>
</section>
<section id="visualize-clusters" class="level3">
<h3 class="anchored" data-anchor-id="visualize-clusters">Visualize clusters</h3>
<p>We can visualise the K-means clustering results of high-dimensional data by using PCA for dimension reduction. We plot the first two principal components and colour the data points (= individual cell lines) by their assigned cluster from K-means.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first, run PCA again on the NCI60 data</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>pr.out <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(nci.data, <span class="at">scale=</span><span class="cn">TRUE</span>)</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a><span class="co"># more cluster options</span></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>km.out2 <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(sd.data, <span class="dv">2</span>, <span class="at">nstart=</span><span class="dv">20</span>)</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>km.out3 <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(sd.data, <span class="dv">3</span>, <span class="at">nstart=</span><span class="dv">20</span>)</span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a><span class="co"># we can now visualise the K-Means results by labelling the data points</span></span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a><span class="co"># in a plot of the scores of the first 2 principal components:</span></span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>))</span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pr.out<span class="sc">$</span>x[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="at">col=</span>(km.out2<span class="sc">$</span>cluster<span class="sc">+</span><span class="dv">1</span>), <span class="at">main=</span><span class="st">"K-Means with K=2"</span>,</span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">"PC 1"</span>, <span class="at">ylab=</span><span class="st">"PC 2"</span>, <span class="at">pch=</span><span class="dv">20</span>)</span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pr.out<span class="sc">$</span>x[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="at">col=</span>(km.out3<span class="sc">$</span>cluster<span class="sc">+</span><span class="dv">1</span>), <span class="at">main=</span><span class="st">"K-Means with K=3"</span>,</span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">"PC 1"</span>,  <span class="at">ylab=</span><span class="st">"PC 2"</span>, <span class="at">pch=</span><span class="dv">20</span>)</span>
<span id="cb45-15"><a href="#cb45-15" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pr.out<span class="sc">$</span>x[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="at">col=</span>(km.out4<span class="sc">$</span>cluster<span class="sc">+</span><span class="dv">1</span>), <span class="at">main=</span><span class="st">"K-Means with K=4"</span>,</span>
<span id="cb45-16"><a href="#cb45-16" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">"PC 1"</span>, <span class="at">ylab=</span><span class="st">"PC 2"</span>, <span class="at">pch=</span><span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day4_clustering_files/figure-html/clust-nci60-km-pca-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Compare with the plot from Exercise 2 yesterday (left panel) along with the cancer type labels. The clusters from K-means seem to correspond decently to partition the data into groups.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>Cols<span class="ot">=</span><span class="cf">function</span>(vec){</span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>  cols<span class="ot">=</span><span class="fu">rainbow</span>(<span class="fu">length</span>(<span class="fu">unique</span>(vec)))</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(cols[<span class="fu">as.numeric</span>(<span class="fu">as.factor</span>(vec))])</span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pr.out<span class="sc">$</span>x[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="at">col=</span><span class="fu">Cols</span>(nci.labs), <span class="at">pch=</span><span class="dv">19</span>,<span class="at">xlab=</span><span class="st">"PC 1"</span>,<span class="at">ylab=</span><span class="st">" PC 2"</span>)</span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">'topleft'</span>, <span class="at">col=</span><span class="fu">rainbow</span>(<span class="fu">length</span>(<span class="fu">unique</span>(nci.labs))), <span class="at">legend=</span><span class="fu">unique</span>(nci.labs), <span class="at">bty=</span><span class="st">'n'</span>, <span class="at">lwd=</span><span class="dv">2</span>, <span class="at">cex=</span>.<span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day4_clustering_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="heatmap-1" class="level3">
<h3 class="anchored" data-anchor-id="heatmap-1">Heatmap</h3>
<p>A heatmap is another way to visualize the clusters from the data. We use the principal components rather than the raw data, as PCs are already explaining a large amount of variability in the over 6000 features.</p>
<p>Similar values are presented with similar colors.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="do">## We use the scores of the PCA on the NCI60 data, to reduce dimension</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>scores <span class="ot">&lt;-</span> pr.out<span class="sc">$</span>x</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>scores[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>] <span class="co"># first 5 pc, first 5 measurements</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         PC1       PC2         PC3         PC4        PC5
V1 -19.68245  3.527748  -9.7354382   0.8177816 -12.511081
V2 -22.90812  6.390938 -13.3725378  -5.5911088  -7.972471
V3 -27.24077  2.445809  -3.5053437   1.3311502 -12.466296
V4 -42.48098 -9.691742  -0.8830921  -3.4180227 -41.938370
V5 -54.98387 -5.158121 -20.9291076 -15.7253986 -10.361364</code></pre>
</div>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co">#  default choices</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(pr.out<span class="sc">$</span>x)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day4_clustering_files/figure-html/clust-nci60-heatmap-pca-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>You can remove the dendrogram on the PCs, only keeping the ones for cancer types. Now you see that the PCs have kept their original order from 1 to 64.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># hc.corr is the result from hclust. check the section on hierarchical clustering</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(pr.out<span class="sc">$</span>x, <span class="at">Rowv =</span> <span class="fu">as.dendrogram</span>(hc.corr), <span class="at">Colv =</span> <span class="cn">NA</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day4_clustering_files/figure-html/clust-nci60-heatmap-dendro-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>You can also reduce the number of PCs, and add titles to the plot annd y-axis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">cex.main =</span> .<span class="dv">7</span>)</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(pr.out<span class="sc">$</span>x[,<span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>], <span class="at">Rowv =</span> <span class="fu">as.dendrogram</span>(hc.corr), <span class="at">Colv =</span> <span class="cn">NA</span>,</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">labRow =</span> nci.labs, <span class="at">main =</span> <span class="st">'Heatmap of the scores of the first 40 PCs on the NCI60 data'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day4_clustering_files/figure-html/clust-nci60-heatmap-type-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="exercise-3-gene-expression-data" class="level2">
<h2 class="anchored" data-anchor-id="exercise-3-gene-expression-data">Exercise 3: Gene expression data</h2>
<p>(CH12Ex13 from statistical learning)</p>
<p>We use the <code>Ch12Ex13.csv</code> data to repeat some of the clustering analysis we did.</p>
<p>The first 20 samples are from healthy patients, while the second 20 are from a diseased group.</p>
<p>Load in the data using read.csv(). You will need to select header=F. Alternatively: load in the data using “Import dataset” in the upper right window, and click “no” on the “Heading” option.</p>
<p>Carry out both hierarchical clustering and K-means clustering. You should choose the most meaningful number of clusters (think about how many groups of patients we have!). Compare the results.</p>
<p>Note: remember that the data has genes on the rows and patients on the columns. You need to transpose the data so that the orders are reversed.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load in the data using read.csv(). You will need to select header=F.</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"data/Ch12Ex13.csv"</span>, <span class="at">header=</span><span class="cn">FALSE</span>)</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1000   40</code></pre>
</div>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># transpose the data, so that we have each row is one patient (subject)</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">t</span>(data) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now the first 20 rows are measurements from healthy patients (group 0), and 21-50 rows are the disease patients (group 1). We can denote this information in a vector like this.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>true.groups <span class="ot">&lt;-</span> <span class="fu">c</span>( <span class="fu">rep</span>(<span class="dv">0</span>,<span class="dv">20</span>), <span class="fu">rep</span>(<span class="dv">1</span>,<span class="dv">20</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="hierarchical-clustering-2" class="level3">
<h3 class="anchored" data-anchor-id="hierarchical-clustering-2">Hierarchical clustering</h3>
<p>You can use different linkage options and distance metrics of your choosing. For example, with complete linkage the code is like this.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>data.dist <span class="ot">&lt;-</span> <span class="fu">dist</span>(data) <span class="co"># need to compute the distance matrix</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>hclust.df <span class="ot">&lt;-</span> <span class="fu">hclust</span>(data.dist, <span class="at">method=</span><span class="st">"complete"</span> )</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a><span class="co">#alternatives:</span></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a><span class="co">#hclust.df &lt;- hclust( D, method="average" )</span></span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a><span class="co">#hclust.df &lt;- hclust( D, method="single" )</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can keep 2 clusters with <code>cutree</code>. Then do a cross tabulation of the true labels and clustered results: how well do they correspond?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># find the clusters</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>predicted <span class="ot">&lt;-</span> <span class="fu">cutree</span>( hclust.df, <span class="at">k=</span><span class="dv">2</span> )</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a><span class="co"># How well does our clustering predict health vs. diseased</span></span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(predicted, true.groups )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         true.groups
predicted  0  1
        1 20  0
        2  0 20</code></pre>
</div>
</div>
</section>
<section id="k-means" class="level3">
<h3 class="anchored" data-anchor-id="k-means">K-means</h3>
<p>Now you can use K-means to identify 2 clusters.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>predicted.kmean <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(data, <span class="dv">2</span>, <span class="at">nstart=</span><span class="dv">20</span>)<span class="sc">$</span>cluster</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="co"># agreement with true label</span></span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(predicted.kmean, true.groups )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               true.groups
predicted.kmean  0  1
              1 20  0
              2  0 20</code></pre>
</div>
</div>
<p>Both methods seem to do work decently for the task.</p>


<!-- -->

</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb61" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "R Lab (day 4): Clustering"</span></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span><span class="co"> </span></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a><span class="co">    code-fold: false</span></span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools: true</span></span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a>Download datasets <span class="co">[</span><span class="ot">here</span><span class="co">](https://github.com/ocbe-uio/course_med3007/tree/main/lab/data)</span>, or from Canvas.</span>
<span id="cb61-10"><a href="#cb61-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-11"><a href="#cb61-11" aria-hidden="true" tabindex="-1"></a>R script: <span class="co">[</span><span class="ot">Code</span><span class="co">](https://github.com/ocbe-uio/course_med3007/blob/main/lab/code/MED3007_Lab3_clustering.R)</span></span>
<span id="cb61-12"><a href="#cb61-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-13"><a href="#cb61-13" aria-hidden="true" tabindex="-1"></a>Presentation (2024 version by Chi): <span class="co">[</span><span class="ot">Slides</span><span class="co">](presentation/Lab_clustering.pdf)</span></span>
<span id="cb61-14"><a href="#cb61-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-15"><a href="#cb61-15" aria-hidden="true" tabindex="-1"></a><span class="fu">## Exercise 1: Food</span></span>
<span id="cb61-16"><a href="#cb61-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-17"><a href="#cb61-17" aria-hidden="true" tabindex="-1"></a>We use the same <span class="in">`Food.txt`</span> data to illustrate two concepts: hierarchical clustering, and heatmap. </span>
<span id="cb61-18"><a href="#cb61-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-19"><a href="#cb61-19" aria-hidden="true" tabindex="-1"></a>This is not a genomics dataset, but for the ease of interpretability, we use it for teaching purposes.</span>
<span id="cb61-20"><a href="#cb61-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-21"><a href="#cb61-21" aria-hidden="true" tabindex="-1"></a>Let us load the dataset.</span>
<span id="cb61-22"><a href="#cb61-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-25"><a href="#cb61-25" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-26"><a href="#cb61-26" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: hc-food-loaddata</span></span>
<span id="cb61-27"><a href="#cb61-27" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-28"><a href="#cb61-28" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-29"><a href="#cb61-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-30"><a href="#cb61-30" aria-hidden="true" tabindex="-1"></a>food <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">'data/Food.txt'</span>, <span class="at">header=</span>T)</span>
<span id="cb61-31"><a href="#cb61-31" aria-hidden="true" tabindex="-1"></a><span class="co"># we change the name from pulses to a more common name, legume</span></span>
<span id="cb61-32"><a href="#cb61-32" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(food)[<span class="dv">7</span>] <span class="ot">&lt;-</span> <span class="st">'Legume'</span></span>
<span id="cb61-33"><a href="#cb61-33" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(food) <span class="co"># print first 6 lines </span></span>
<span id="cb61-34"><a href="#cb61-34" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-35"><a href="#cb61-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-36"><a href="#cb61-36" aria-hidden="true" tabindex="-1"></a>We **scale** the data (also called standardize, or normalize sometimes) so that each column (feature, variable) has 0 mean and 1 variance. We call the scaled data <span class="in">`food_s`</span>.</span>
<span id="cb61-37"><a href="#cb61-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-40"><a href="#cb61-40" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-41"><a href="#cb61-41" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: hc-food-scale</span></span>
<span id="cb61-42"><a href="#cb61-42" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-43"><a href="#cb61-43" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-44"><a href="#cb61-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-45"><a href="#cb61-45" aria-hidden="true" tabindex="-1"></a>food_s <span class="ot">&lt;-</span> <span class="fu">scale</span>(food)</span>
<span id="cb61-46"><a href="#cb61-46" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(food_s) <span class="co"># print first 6 lines</span></span>
<span id="cb61-47"><a href="#cb61-47" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-48"><a href="#cb61-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-49"><a href="#cb61-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-50"><a href="#cb61-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-51"><a href="#cb61-51" aria-hidden="true" tabindex="-1"></a><span class="fu">### Distances</span></span>
<span id="cb61-52"><a href="#cb61-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-53"><a href="#cb61-53" aria-hidden="true" tabindex="-1"></a>To do hierarchical clustering, the most convenient command is <span class="in">`hclust()`</span>. As input you would need a **distance** between the subjects (patients, or countries in this example). We do it on the scaled data.</span>
<span id="cb61-54"><a href="#cb61-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-55"><a href="#cb61-55" aria-hidden="true" tabindex="-1"></a>The command to compute pair-wise distance is <span class="in">`dist()`</span>. By default, the distance being computed is the Euclidean distance (*details optional*). Euclidean distance is possibly the most commonly used metric, but there are others. See <span class="in">`?dist()`</span> to find out more options.</span>
<span id="cb61-56"><a href="#cb61-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-57"><a href="#cb61-57" aria-hidden="true" tabindex="-1"></a>We can present the pair-wise distances in a matrix format. You can see that this matrix is symmetric, with 0 on the diagonal - this should be intuitive: the distance between A - B is the same as B - A, and the distance between A and itself is 0.</span>
<span id="cb61-58"><a href="#cb61-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-59"><a href="#cb61-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-62"><a href="#cb61-62" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-63"><a href="#cb61-63" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: hc-food-dist</span></span>
<span id="cb61-64"><a href="#cb61-64" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-65"><a href="#cb61-65" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-66"><a href="#cb61-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-67"><a href="#cb61-67" aria-hidden="true" tabindex="-1"></a><span class="co"># compute distance</span></span>
<span id="cb61-68"><a href="#cb61-68" aria-hidden="true" tabindex="-1"></a>food_dist <span class="ot">&lt;-</span> <span class="fu">dist</span>(food_s)</span>
<span id="cb61-69"><a href="#cb61-69" aria-hidden="true" tabindex="-1"></a><span class="co"># round(food_dist, digits = 2) # try this yourself to see what it does</span></span>
<span id="cb61-70"><a href="#cb61-70" aria-hidden="true" tabindex="-1"></a><span class="co"># alternatively, look at this as a matrix</span></span>
<span id="cb61-71"><a href="#cb61-71" aria-hidden="true" tabindex="-1"></a>food_dist_matrix <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(food_dist)</span>
<span id="cb61-72"><a href="#cb61-72" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(food_dist_matrix[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>], <span class="at">digits =</span> <span class="dv">2</span>) <span class="co"># first 5 row 5 col</span></span>
<span id="cb61-73"><a href="#cb61-73" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-74"><a href="#cb61-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-75"><a href="#cb61-75" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb61-76"><a href="#cb61-76" aria-hidden="true" tabindex="-1"></a><span class="fu">## Optional: Euclidean disance</span></span>
<span id="cb61-77"><a href="#cb61-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-78"><a href="#cb61-78" aria-hidden="true" tabindex="-1"></a>You can check the Euclidean distance between Albania and Austria is indeed 5.95. This distance is the square root of the sum of squared differences between two subjects in all their measurements. </span>
<span id="cb61-79"><a href="#cb61-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-80"><a href="#cb61-80" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb61-81"><a href="#cb61-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-84"><a href="#cb61-84" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-85"><a href="#cb61-85" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: hc-food-dist2</span></span>
<span id="cb61-86"><a href="#cb61-86" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-87"><a href="#cb61-87" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-88"><a href="#cb61-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-89"><a href="#cb61-89" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(food_s[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>,], <span class="at">digits =</span> <span class="dv">2</span>) <span class="co"># we only keep first 2 digits</span></span>
<span id="cb61-90"><a href="#cb61-90" aria-hidden="true" tabindex="-1"></a><span class="co"># take the data for two countries each </span></span>
<span id="cb61-91"><a href="#cb61-91" aria-hidden="true" tabindex="-1"></a>albania <span class="ot">&lt;-</span> <span class="fu">round</span>(food_s[<span class="dv">1</span>,], <span class="at">digits =</span> <span class="dv">2</span>)</span>
<span id="cb61-92"><a href="#cb61-92" aria-hidden="true" tabindex="-1"></a>austria <span class="ot">&lt;-</span> <span class="fu">round</span>(food_s[<span class="dv">2</span>,], <span class="at">digits =</span> <span class="dv">2</span>)</span>
<span id="cb61-93"><a href="#cb61-93" aria-hidden="true" tabindex="-1"></a><span class="co"># compute difference between each col</span></span>
<span id="cb61-94"><a href="#cb61-94" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> albania <span class="sc">-</span> austria</span>
<span id="cb61-95"><a href="#cb61-95" aria-hidden="true" tabindex="-1"></a>d</span>
<span id="cb61-96"><a href="#cb61-96" aria-hidden="true" tabindex="-1"></a><span class="co"># euclidean distance: square each element, sum together, and take a square root</span></span>
<span id="cb61-97"><a href="#cb61-97" aria-hidden="true" tabindex="-1"></a><span class="fu">sqrt</span>(<span class="fu">sum</span>(d<span class="sc">^</span><span class="dv">2</span>)) </span>
<span id="cb61-98"><a href="#cb61-98" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-99"><a href="#cb61-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-100"><a href="#cb61-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-101"><a href="#cb61-101" aria-hidden="true" tabindex="-1"></a><span class="fu">### Hierarchical clustering</span></span>
<span id="cb61-102"><a href="#cb61-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-103"><a href="#cb61-103" aria-hidden="true" tabindex="-1"></a>Now that we have computed the distance <span class="in">`food_dist`</span>, we plug it in the clustering algorithm, <span class="in">`hclust()`</span>. </span>
<span id="cb61-104"><a href="#cb61-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-105"><a href="#cb61-105" aria-hidden="true" tabindex="-1"></a>We try the complete linkage method, by specifying <span class="in">`method = 'complete'`</span>. The result is saved as <span class="in">`hc.complete`</span>. You can visualize it, and add label of the country names to make it easier to read.</span>
<span id="cb61-106"><a href="#cb61-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-109"><a href="#cb61-109" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-110"><a href="#cb61-110" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: hc-food-complete</span></span>
<span id="cb61-111"><a href="#cb61-111" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-112"><a href="#cb61-112" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-113"><a href="#cb61-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-114"><a href="#cb61-114" aria-hidden="true" tabindex="-1"></a>hc.complete <span class="ot">&lt;-</span> <span class="fu">hclust</span>(food_dist, <span class="at">method=</span><span class="st">"complete"</span>)</span>
<span id="cb61-115"><a href="#cb61-115" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc.complete, <span class="at">labels=</span><span class="fu">rownames</span>(food), <span class="at">main=</span><span class="st">"Complete Linkage"</span>, <span class="at">xlab=</span><span class="st">""</span>, <span class="at">sub=</span><span class="st">""</span>)</span>
<span id="cb61-116"><a href="#cb61-116" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-117"><a href="#cb61-117" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-118"><a href="#cb61-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-119"><a href="#cb61-119" aria-hidden="true" tabindex="-1"></a><span class="fu">### Linkage, dissimilarity, scaling</span></span>
<span id="cb61-120"><a href="#cb61-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-121"><a href="#cb61-121" aria-hidden="true" tabindex="-1"></a>Hierarchical clustering is a class of methods, and there are a variety of options to set. </span>
<span id="cb61-122"><a href="#cb61-122" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-123"><a href="#cb61-123" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Linkage (by seting <span class="in">`method`</span> inside <span class="in">`hclust()`</span>): complete, single, average</span>
<span id="cb61-124"><a href="#cb61-124" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Dissimilarity: Euclidean, correlation, ...</span>
<span id="cb61-125"><a href="#cb61-125" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Scaling: scaled data (mean 0 variance 1) or unscaled, original data</span>
<span id="cb61-126"><a href="#cb61-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-127"><a href="#cb61-127" aria-hidden="true" tabindex="-1"></a>There is no definite guide on which combination works the best, hence you can try them out and see what could make most sense. Again, in unsupervised learning data do not have outcome labels, so the interpretation is left for the domain experts to make.</span>
<span id="cb61-128"><a href="#cb61-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-129"><a href="#cb61-129" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-132"><a href="#cb61-132" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-133"><a href="#cb61-133" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: hc-food-linkage</span></span>
<span id="cb61-134"><a href="#cb61-134" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-135"><a href="#cb61-135" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-136"><a href="#cb61-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-137"><a href="#cb61-137" aria-hidden="true" tabindex="-1"></a><span class="co"># single linkage</span></span>
<span id="cb61-138"><a href="#cb61-138" aria-hidden="true" tabindex="-1"></a>hc.single <span class="ot">&lt;-</span> <span class="fu">hclust</span>(food_dist, <span class="at">method=</span><span class="st">"single"</span>)</span>
<span id="cb61-139"><a href="#cb61-139" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc.single, <span class="at">labels=</span><span class="fu">rownames</span>(food), <span class="at">main=</span><span class="st">"Single Linkage"</span>, <span class="at">xlab=</span><span class="st">""</span>, <span class="at">sub=</span><span class="st">""</span>)</span>
<span id="cb61-140"><a href="#cb61-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-141"><a href="#cb61-141" aria-hidden="true" tabindex="-1"></a><span class="co"># average linkage</span></span>
<span id="cb61-142"><a href="#cb61-142" aria-hidden="true" tabindex="-1"></a>hc.average <span class="ot">&lt;-</span> <span class="fu">hclust</span>(food_dist, <span class="at">method=</span><span class="st">"average"</span>)</span>
<span id="cb61-143"><a href="#cb61-143" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc.average, <span class="at">labels=</span><span class="fu">rownames</span>(food), <span class="at">main=</span><span class="st">"Average Linkage"</span>, <span class="at">xlab=</span><span class="st">""</span>, <span class="at">sub=</span><span class="st">""</span>)</span>
<span id="cb61-144"><a href="#cb61-144" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-145"><a href="#cb61-145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-146"><a href="#cb61-146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-147"><a href="#cb61-147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-150"><a href="#cb61-150" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-151"><a href="#cb61-151" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: hc-food-unscaled</span></span>
<span id="cb61-152"><a href="#cb61-152" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-153"><a href="#cb61-153" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-154"><a href="#cb61-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-155"><a href="#cb61-155" aria-hidden="true" tabindex="-1"></a><span class="co"># unscaled data, complete linkage</span></span>
<span id="cb61-156"><a href="#cb61-156" aria-hidden="true" tabindex="-1"></a>hc.unscaled <span class="ot">&lt;-</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(food), <span class="at">method=</span><span class="st">"complete"</span>)</span>
<span id="cb61-157"><a href="#cb61-157" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc.unscaled, <span class="at">labels=</span><span class="fu">rownames</span>(food), <span class="at">main=</span><span class="st">"Complete linkage with unscaled features"</span>, <span class="at">xlab=</span><span class="st">""</span>, <span class="at">sub=</span><span class="st">""</span>)</span>
<span id="cb61-158"><a href="#cb61-158" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-159"><a href="#cb61-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-160"><a href="#cb61-160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-163"><a href="#cb61-163" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-164"><a href="#cb61-164" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: hc-food-cor</span></span>
<span id="cb61-165"><a href="#cb61-165" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-166"><a href="#cb61-166" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-167"><a href="#cb61-167" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-168"><a href="#cb61-168" aria-hidden="true" tabindex="-1"></a><span class="co"># correlation as dissimiarity, rather than euclidean distance</span></span>
<span id="cb61-169"><a href="#cb61-169" aria-hidden="true" tabindex="-1"></a>dd <span class="ot">&lt;-</span> <span class="fu">as.dist</span>(<span class="dv">1</span><span class="sc">-</span><span class="fu">cor</span>(<span class="fu">t</span>(food_s))) <span class="co"># compute the metric</span></span>
<span id="cb61-170"><a href="#cb61-170" aria-hidden="true" tabindex="-1"></a>hc.corr <span class="ot">&lt;-</span> <span class="fu">hclust</span>(dd, <span class="at">method=</span><span class="st">"complete"</span>) <span class="co"># cluster</span></span>
<span id="cb61-171"><a href="#cb61-171" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc.corr, <span class="at">labels=</span><span class="fu">rownames</span>(food), <span class="at">main=</span><span class="st">"Complete linkage with correlation-based distance"</span>, <span class="at">xlab=</span><span class="st">""</span>, <span class="at">sub=</span><span class="st">""</span>)</span>
<span id="cb61-172"><a href="#cb61-172" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-173"><a href="#cb61-173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-174"><a href="#cb61-174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-175"><a href="#cb61-175" aria-hidden="true" tabindex="-1"></a><span class="fu">### Heatmap</span></span>
<span id="cb61-176"><a href="#cb61-176" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-177"><a href="#cb61-177" aria-hidden="true" tabindex="-1"></a>Heatmap is a visualization tool to plot data of similar values in similar colors, so that you can identify visualy if there is any pattern. It can also be combined with hierarchical clustering - this is actually the default outcome: dendrograms are displayed for both rows and column.</span>
<span id="cb61-178"><a href="#cb61-178" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-181"><a href="#cb61-181" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-182"><a href="#cb61-182" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: hc-food-heatmap1</span></span>
<span id="cb61-183"><a href="#cb61-183" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-184"><a href="#cb61-184" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-185"><a href="#cb61-185" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-186"><a href="#cb61-186" aria-hidden="true" tabindex="-1"></a><span class="co"># make heatmap on the scaled data</span></span>
<span id="cb61-187"><a href="#cb61-187" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(food_s)</span>
<span id="cb61-188"><a href="#cb61-188" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-189"><a href="#cb61-189" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-190"><a href="#cb61-190" aria-hidden="true" tabindex="-1"></a>To preserve the original ordering of the columns and rows, you can specify <span class="in">`Rowv = NA, Colv = NA`</span>.</span>
<span id="cb61-191"><a href="#cb61-191" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-194"><a href="#cb61-194" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-195"><a href="#cb61-195" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: hc-food-heatmap2</span></span>
<span id="cb61-196"><a href="#cb61-196" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-197"><a href="#cb61-197" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-198"><a href="#cb61-198" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-199"><a href="#cb61-199" aria-hidden="true" tabindex="-1"></a><span class="co"># no clustering for row or col, this preserves the original ordering</span></span>
<span id="cb61-200"><a href="#cb61-200" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(food_s, <span class="at">Rowv =</span> <span class="cn">NA</span>, <span class="at">Colv =</span> <span class="cn">NA</span>)</span>
<span id="cb61-201"><a href="#cb61-201" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-202"><a href="#cb61-202" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-203"><a href="#cb61-203" aria-hidden="true" tabindex="-1"></a>Can also only do clustering for row only (or column only). </span>
<span id="cb61-204"><a href="#cb61-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-207"><a href="#cb61-207" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-208"><a href="#cb61-208" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: hc-food-heatmap3</span></span>
<span id="cb61-209"><a href="#cb61-209" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-210"><a href="#cb61-210" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-211"><a href="#cb61-211" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-212"><a href="#cb61-212" aria-hidden="true" tabindex="-1"></a><span class="co"># only clustering for row</span></span>
<span id="cb61-213"><a href="#cb61-213" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(food_s, <span class="at">Colv =</span> <span class="cn">NA</span>)</span>
<span id="cb61-214"><a href="#cb61-214" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-215"><a href="#cb61-215" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-216"><a href="#cb61-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-217"><a href="#cb61-217" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-218"><a href="#cb61-218" aria-hidden="true" tabindex="-1"></a><span class="fu">## Exercise 2: NCI60</span></span>
<span id="cb61-219"><a href="#cb61-219" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-220"><a href="#cb61-220" aria-hidden="true" tabindex="-1"></a>We look at the NCI60 data again. First load the dataset.</span>
<span id="cb61-221"><a href="#cb61-221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-224"><a href="#cb61-224" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-225"><a href="#cb61-225" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: clust-nci60-loaddata</span></span>
<span id="cb61-226"><a href="#cb61-226" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-227"><a href="#cb61-227" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-228"><a href="#cb61-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-229"><a href="#cb61-229" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ISLR)</span>
<span id="cb61-230"><a href="#cb61-230" aria-hidden="true" tabindex="-1"></a><span class="co"># or, load('data/NCI60.RData')</span></span>
<span id="cb61-231"><a href="#cb61-231" aria-hidden="true" tabindex="-1"></a>nci.labs <span class="ot">&lt;-</span> NCI60<span class="sc">$</span>labs <span class="co"># Sample labels (tissue type)</span></span>
<span id="cb61-232"><a href="#cb61-232" aria-hidden="true" tabindex="-1"></a>nci.data <span class="ot">&lt;-</span> NCI60<span class="sc">$</span>data <span class="co"># Gene expression data set</span></span>
<span id="cb61-233"><a href="#cb61-233" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-234"><a href="#cb61-234" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-235"><a href="#cb61-235" aria-hidden="true" tabindex="-1"></a><span class="fu">### Hierarchical clustering</span></span>
<span id="cb61-236"><a href="#cb61-236" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-237"><a href="#cb61-237" aria-hidden="true" tabindex="-1"></a>We start by scaling the data, and calculate the distance matrix (using the Euclidean distance), and then investigate different linkage methods.</span>
<span id="cb61-238"><a href="#cb61-238" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-239"><a href="#cb61-239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-242"><a href="#cb61-242" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-243"><a href="#cb61-243" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: clust-nci60-scale</span></span>
<span id="cb61-244"><a href="#cb61-244" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-245"><a href="#cb61-245" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-246"><a href="#cb61-246" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-247"><a href="#cb61-247" aria-hidden="true" tabindex="-1"></a><span class="co"># Scale the data to zero mean and unit variance:</span></span>
<span id="cb61-248"><a href="#cb61-248" aria-hidden="true" tabindex="-1"></a>sd.data <span class="ot">&lt;-</span> <span class="fu">scale</span>(nci.data)</span>
<span id="cb61-249"><a href="#cb61-249" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-250"><a href="#cb61-250" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the distance matrix </span></span>
<span id="cb61-251"><a href="#cb61-251" aria-hidden="true" tabindex="-1"></a><span class="co"># equivalent: data.dist &lt;- dist(sd.data, method="euclidean")</span></span>
<span id="cb61-252"><a href="#cb61-252" aria-hidden="true" tabindex="-1"></a>data.dist <span class="ot">&lt;-</span> <span class="fu">dist</span>(sd.data)</span>
<span id="cb61-253"><a href="#cb61-253" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-254"><a href="#cb61-254" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-255"><a href="#cb61-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-256"><a href="#cb61-256" aria-hidden="true" tabindex="-1"></a>Next we perform hierarchical clustering with distance matrix as input. The function we use is <span class="in">`hclust()`</span>. We specify the linkage method to be <span class="in">`complete`</span>.</span>
<span id="cb61-257"><a href="#cb61-257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-258"><a href="#cb61-258" aria-hidden="true" tabindex="-1"></a>Once the result is saved in <span class="in">`hc.complete`</span> object, you can plot the dendrogram.</span>
<span id="cb61-259"><a href="#cb61-259" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-262"><a href="#cb61-262" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-263"><a href="#cb61-263" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: clust-nci60-hccomplete</span></span>
<span id="cb61-264"><a href="#cb61-264" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-265"><a href="#cb61-265" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-266"><a href="#cb61-266" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-267"><a href="#cb61-267" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform clustering</span></span>
<span id="cb61-268"><a href="#cb61-268" aria-hidden="true" tabindex="-1"></a>hc.complete <span class="ot">&lt;-</span> <span class="fu">hclust</span>(data.dist, <span class="at">method=</span><span class="st">"complete"</span>)</span>
<span id="cb61-269"><a href="#cb61-269" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-270"><a href="#cb61-270" aria-hidden="true" tabindex="-1"></a><span class="co"># names(hc.complete)</span></span>
<span id="cb61-271"><a href="#cb61-271" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc.complete, <span class="at">labels=</span>nci.labs, <span class="at">main=</span><span class="st">"Complete Linkage"</span>, <span class="at">xlab=</span><span class="st">""</span>, <span class="at">sub=</span><span class="st">""</span>)</span>
<span id="cb61-272"><a href="#cb61-272" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-273"><a href="#cb61-273" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-274"><a href="#cb61-274" aria-hidden="true" tabindex="-1"></a>The object <span class="in">`hc.complete`</span> contains a lot of information. To get the information, you can use the <span class="in">`$`</span> operator. </span>
<span id="cb61-275"><a href="#cb61-275" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-276"><a href="#cb61-276" aria-hidden="true" tabindex="-1"></a>You should refer to the documentation for <span class="in">`hclust()`</span> to see a complete list of output. Use <span class="in">`?hclust`</span> to get the documentation on how to use the function.</span>
<span id="cb61-277"><a href="#cb61-277" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-280"><a href="#cb61-280" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-281"><a href="#cb61-281" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: clust-nci60-hccomplete-2</span></span>
<span id="cb61-282"><a href="#cb61-282" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-283"><a href="#cb61-283" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-284"><a href="#cb61-284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-285"><a href="#cb61-285" aria-hidden="true" tabindex="-1"></a>hc.complete<span class="sc">$</span>dist.method <span class="co"># distance method</span></span>
<span id="cb61-286"><a href="#cb61-286" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-287"><a href="#cb61-287" aria-hidden="true" tabindex="-1"></a><span class="co"># hc.complete$merge  # order of aggregations of samples / clusters</span></span>
<span id="cb61-288"><a href="#cb61-288" aria-hidden="true" tabindex="-1"></a><span class="co"># hc.complete$height # distance at which aggregations happen</span></span>
<span id="cb61-289"><a href="#cb61-289" aria-hidden="true" tabindex="-1"></a><span class="co"># hc.complete$labels # labels (numeric, since we don't know the original categories!)</span></span>
<span id="cb61-290"><a href="#cb61-290" aria-hidden="true" tabindex="-1"></a><span class="co"># hc.complete$method</span></span>
<span id="cb61-291"><a href="#cb61-291" aria-hidden="true" tabindex="-1"></a><span class="co"># hc.complete$call</span></span>
<span id="cb61-292"><a href="#cb61-292" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-293"><a href="#cb61-293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-294"><a href="#cb61-294" aria-hidden="true" tabindex="-1"></a>We can try different linkage methods and see how the clustering results differ. Change the <span class="in">`method`</span> argument in the function, and plot the results.</span>
<span id="cb61-295"><a href="#cb61-295" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-298"><a href="#cb61-298" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-299"><a href="#cb61-299" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: clust-nci60-otherlink</span></span>
<span id="cb61-300"><a href="#cb61-300" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-301"><a href="#cb61-301" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-302"><a href="#cb61-302" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb61-303"><a href="#cb61-303" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-304"><a href="#cb61-304" aria-hidden="true" tabindex="-1"></a>hc.average <span class="ot">&lt;-</span> <span class="fu">hclust</span>(data.dist, <span class="at">method=</span><span class="st">"average"</span>)</span>
<span id="cb61-305"><a href="#cb61-305" aria-hidden="true" tabindex="-1"></a>hc.single <span class="ot">&lt;-</span> <span class="fu">hclust</span>(data.dist, <span class="at">method=</span><span class="st">"single"</span>)</span>
<span id="cb61-306"><a href="#cb61-306" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-307"><a href="#cb61-307" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc.average, <span class="at">labels=</span>nci.labs, <span class="at">main=</span><span class="st">"Average Linkage"</span>, <span class="at">xlab=</span><span class="st">""</span>, <span class="at">sub=</span><span class="st">""</span>)</span>
<span id="cb61-308"><a href="#cb61-308" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc.single, <span class="at">labels=</span>nci.labs,  <span class="at">main=</span><span class="st">"Single Linkage"</span>, <span class="at">xlab=</span><span class="st">""</span>, <span class="at">sub=</span><span class="st">""</span>)</span>
<span id="cb61-309"><a href="#cb61-309" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-310"><a href="#cb61-310" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-311"><a href="#cb61-311" aria-hidden="true" tabindex="-1"></a>Now we focus on **complete linkage** only.</span>
<span id="cb61-312"><a href="#cb61-312" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-313"><a href="#cb61-313" aria-hidden="true" tabindex="-1"></a>First, we use <span class="in">`cutree()`</span> to compare the results when the data are separated into either 2 or 4 clusters.</span>
<span id="cb61-314"><a href="#cb61-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-317"><a href="#cb61-317" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-318"><a href="#cb61-318" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: clust-nci60-cutree</span></span>
<span id="cb61-319"><a href="#cb61-319" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-320"><a href="#cb61-320" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-321"><a href="#cb61-321" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-322"><a href="#cb61-322" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare 2 clusters and 4 clusters:</span></span>
<span id="cb61-323"><a href="#cb61-323" aria-hidden="true" tabindex="-1"></a>hc.clusters <span class="ot">&lt;-</span> <span class="fu">cutree</span>(hc.complete, <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">4</span>))</span>
<span id="cb61-324"><a href="#cb61-324" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(hc.clusters) <span class="co"># print first 6 results</span></span>
<span id="cb61-325"><a href="#cb61-325" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-326"><a href="#cb61-326" aria-hidden="true" tabindex="-1"></a><span class="co"># cross tabulation</span></span>
<span id="cb61-327"><a href="#cb61-327" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(hc.clusters[,<span class="st">"2"</span>], hc.clusters[,<span class="st">"4"</span>])</span>
<span id="cb61-328"><a href="#cb61-328" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-329"><a href="#cb61-329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-330"><a href="#cb61-330" aria-hidden="true" tabindex="-1"></a>It is more straightforward to check the results with a dendrogram.</span>
<span id="cb61-331"><a href="#cb61-331" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-334"><a href="#cb61-334" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-335"><a href="#cb61-335" aria-hidden="true" tabindex="-1"></a><span class="co"># visualize the cuts</span></span>
<span id="cb61-336"><a href="#cb61-336" aria-hidden="true" tabindex="-1"></a><span class="co"># how do you know where to draw the line? check height</span></span>
<span id="cb61-337"><a href="#cb61-337" aria-hidden="true" tabindex="-1"></a>heights <span class="ot">&lt;-</span> hc.complete<span class="sc">$</span>height</span>
<span id="cb61-338"><a href="#cb61-338" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(heights, <span class="dv">4</span>)  <span class="co"># print the last 4</span></span>
<span id="cb61-339"><a href="#cb61-339" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc.complete, <span class="at">labels=</span>nci.labs, <span class="at">main=</span><span class="st">"Complete Linkage"</span>, <span class="at">xlab=</span><span class="st">""</span>, <span class="at">sub=</span><span class="st">""</span>)</span>
<span id="cb61-340"><a href="#cb61-340" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h=</span><span class="dv">140</span>, <span class="at">col=</span><span class="st">"red"</span>)  <span class="co"># 4 clusters</span></span>
<span id="cb61-341"><a href="#cb61-341" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h=</span><span class="dv">150</span>, <span class="at">col=</span><span class="st">"blue"</span>) <span class="co"># 2 clusters</span></span>
<span id="cb61-342"><a href="#cb61-342" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-343"><a href="#cb61-343" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-344"><a href="#cb61-344" aria-hidden="true" tabindex="-1"></a>The way to interpret the height variable is simple: it is where two clusters are merged into one. For example, the largest cluster corresponds to the last value of <span class="in">`height`</span> (162.2) - if you check the figure, it is exactly where the horizontal line is merging the two groups. Similarly, 142.9 is where three groups became two, 141.2 is where four groups became three. If you draw a line at 140, it points out the four clusters.</span>
<span id="cb61-345"><a href="#cb61-345" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-346"><a href="#cb61-346" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-347"><a href="#cb61-347" aria-hidden="true" tabindex="-1"></a>How are the labels distributed between clusters? We can focus on 4 cluster situation, and use <span class="in">`table()`</span> to list out which cancer types is merged in which of the four clusters.</span>
<span id="cb61-348"><a href="#cb61-348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-349"><a href="#cb61-349" aria-hidden="true" tabindex="-1"></a>For example, breast cancer appears in all but 3rd cluster; melanoma only appears in the first clsuter; so on so forth.</span>
<span id="cb61-350"><a href="#cb61-350" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-353"><a href="#cb61-353" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-354"><a href="#cb61-354" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: clust-nci60-cutree2</span></span>
<span id="cb61-355"><a href="#cb61-355" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-356"><a href="#cb61-356" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-357"><a href="#cb61-357" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-358"><a href="#cb61-358" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(hc.clusters[,<span class="st">"4"</span>], nci.labs)</span>
<span id="cb61-359"><a href="#cb61-359" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-360"><a href="#cb61-360" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-361"><a href="#cb61-361" aria-hidden="true" tabindex="-1"></a>Finally, we see what happens if we use **unscaled data** instead of scaled data, or if we use a **correlation-based distance metric** instead of the Euclidean distance.</span>
<span id="cb61-362"><a href="#cb61-362" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-363"><a href="#cb61-363" aria-hidden="true" tabindex="-1"></a>Compare the dendrograms: How different are the resulting clusterings? Do you recognise subclusters that are consistent?</span>
<span id="cb61-364"><a href="#cb61-364" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-365"><a href="#cb61-365" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-368"><a href="#cb61-368" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-369"><a href="#cb61-369" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: clust-nci60-unscale</span></span>
<span id="cb61-370"><a href="#cb61-370" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-371"><a href="#cb61-371" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-372"><a href="#cb61-372" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-373"><a href="#cb61-373" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare scaled data versus non-scaled data:</span></span>
<span id="cb61-374"><a href="#cb61-374" aria-hidden="true" tabindex="-1"></a>hc.unscaled <span class="ot">&lt;-</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(nci.data), <span class="at">method=</span><span class="st">"complete"</span>)</span>
<span id="cb61-375"><a href="#cb61-375" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc.unscaled, <span class="at">labels=</span>nci.labs, <span class="at">main=</span><span class="st">"Complete linkage with unscaled features"</span>, <span class="at">xlab=</span><span class="st">""</span>, <span class="at">sub=</span><span class="st">""</span>)</span>
<span id="cb61-376"><a href="#cb61-376" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-377"><a href="#cb61-377" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare Euclidean distance with correlation-based distance:</span></span>
<span id="cb61-378"><a href="#cb61-378" aria-hidden="true" tabindex="-1"></a>dd <span class="ot">&lt;-</span> <span class="fu">as.dist</span>(<span class="dv">1</span><span class="sc">-</span><span class="fu">cor</span>(<span class="fu">t</span>(sd.data)))</span>
<span id="cb61-379"><a href="#cb61-379" aria-hidden="true" tabindex="-1"></a>hc.corr <span class="ot">&lt;-</span> <span class="fu">hclust</span>(dd, <span class="at">method=</span><span class="st">"complete"</span>)</span>
<span id="cb61-380"><a href="#cb61-380" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc.corr, <span class="at">labels=</span>nci.labs, <span class="at">main=</span><span class="st">"Complete linkage with correlation-based distance"</span>, <span class="at">xlab=</span><span class="st">""</span>, <span class="at">sub=</span><span class="st">""</span>)</span>
<span id="cb61-381"><a href="#cb61-381" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-382"><a href="#cb61-382" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-383"><a href="#cb61-383" aria-hidden="true" tabindex="-1"></a><span class="fu">### K-means clustering</span></span>
<span id="cb61-384"><a href="#cb61-384" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-385"><a href="#cb61-385" aria-hidden="true" tabindex="-1"></a>In this section we explore the K-means clustering on the same dataset. </span>
<span id="cb61-386"><a href="#cb61-386" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-387"><a href="#cb61-387" aria-hidden="true" tabindex="-1"></a>In contrast to the hierarchical clustering which requires a **distance** as input, with K-means you would provide the data matrix. The data matrix can be scaled (centered and with unit variance), or unscaled.</span>
<span id="cb61-388"><a href="#cb61-388" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-389"><a href="#cb61-389" aria-hidden="true" tabindex="-1"></a>In this example we use scaled data computed from before, <span class="in">`sd.data`</span>. </span>
<span id="cb61-390"><a href="#cb61-390" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-393"><a href="#cb61-393" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-394"><a href="#cb61-394" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: clust-nci60-km4</span></span>
<span id="cb61-395"><a href="#cb61-395" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-396"><a href="#cb61-396" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-397"><a href="#cb61-397" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-398"><a href="#cb61-398" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">4</span>) <span class="co"># set random seed</span></span>
<span id="cb61-399"><a href="#cb61-399" aria-hidden="true" tabindex="-1"></a>km.out4 <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(sd.data, <span class="at">centers =</span> <span class="dv">4</span>, <span class="at">nstart=</span><span class="dv">20</span>)</span>
<span id="cb61-400"><a href="#cb61-400" aria-hidden="true" tabindex="-1"></a>km.out4<span class="sc">$</span>cluster</span>
<span id="cb61-401"><a href="#cb61-401" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-402"><a href="#cb61-402" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-403"><a href="#cb61-403" aria-hidden="true" tabindex="-1"></a>Read the help file <span class="in">`?kmeans`</span> to understand what the argument nstart=20 does. Comparing an analysis with nstart=20 versus nstart=1 demonstrates how the cluster results can be improved if we allow more evaluations with different randomly chosen starting centroids.</span>
<span id="cb61-404"><a href="#cb61-404" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-405"><a href="#cb61-405" aria-hidden="true" tabindex="-1"></a>Set a different random seed, say 3 (as long as it's different from the one you used before), and run the analysis again. This time we use a different <span class="in">`nstart`</span></span>
<span id="cb61-406"><a href="#cb61-406" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-409"><a href="#cb61-409" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-410"><a href="#cb61-410" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: clust-nci60-km3</span></span>
<span id="cb61-411"><a href="#cb61-411" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-412"><a href="#cb61-412" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-413"><a href="#cb61-413" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-414"><a href="#cb61-414" aria-hidden="true" tabindex="-1"></a><span class="co"># different starting centroids improve the clustering:</span></span>
<span id="cb61-415"><a href="#cb61-415" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">3</span>)</span>
<span id="cb61-416"><a href="#cb61-416" aria-hidden="true" tabindex="-1"></a>km.out <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(sd.data, <span class="at">centers =</span> <span class="dv">4</span>, <span class="at">nstart=</span><span class="dv">1</span>)</span>
<span id="cb61-417"><a href="#cb61-417" aria-hidden="true" tabindex="-1"></a>km.out<span class="sc">$</span>cluster <span class="co"># cluster label</span></span>
<span id="cb61-418"><a href="#cb61-418" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-419"><a href="#cb61-419" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-420"><a href="#cb61-420" aria-hidden="true" tabindex="-1"></a><span class="fu">### Compare with hierarchical clustering</span></span>
<span id="cb61-421"><a href="#cb61-421" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-424"><a href="#cb61-424" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-425"><a href="#cb61-425" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: clust-nci60-km-res</span></span>
<span id="cb61-426"><a href="#cb61-426" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-427"><a href="#cb61-427" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-428"><a href="#cb61-428" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-429"><a href="#cb61-429" aria-hidden="true" tabindex="-1"></a><span class="co"># we can directly compare the k-means result (along rows)</span></span>
<span id="cb61-430"><a href="#cb61-430" aria-hidden="true" tabindex="-1"></a><span class="co"># with the hierarchical clustering result (along columns)</span></span>
<span id="cb61-431"><a href="#cb61-431" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(km.out4<span class="sc">$</span>cluster, hc.clusters[,<span class="st">"4"</span>], <span class="at">deparse.level=</span><span class="dv">2</span>)</span>
<span id="cb61-432"><a href="#cb61-432" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-433"><a href="#cb61-433" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-434"><a href="#cb61-434" aria-hidden="true" tabindex="-1"></a>From the results, you can see that the results are slightly different between the two methods. Keep in mind that in unsupervised learning you do not have the real outcome label (such as the cancer types here), so you need to try a few different methods and compare the outputs, and make interpretations accordingly.</span>
<span id="cb61-435"><a href="#cb61-435" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-436"><a href="#cb61-436" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-437"><a href="#cb61-437" aria-hidden="true" tabindex="-1"></a><span class="fu">### Visualize clusters</span></span>
<span id="cb61-438"><a href="#cb61-438" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-439"><a href="#cb61-439" aria-hidden="true" tabindex="-1"></a>We can visualise the K-means clustering results of high-dimensional data by using PCA for dimension reduction. We plot the first two principal components and colour the data points (= individual cell lines) by their assigned cluster from K-means.</span>
<span id="cb61-440"><a href="#cb61-440" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-443"><a href="#cb61-443" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-444"><a href="#cb61-444" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: clust-nci60-km-pca</span></span>
<span id="cb61-445"><a href="#cb61-445" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-446"><a href="#cb61-446" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-447"><a href="#cb61-447" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-448"><a href="#cb61-448" aria-hidden="true" tabindex="-1"></a><span class="co"># first, run PCA again on the NCI60 data</span></span>
<span id="cb61-449"><a href="#cb61-449" aria-hidden="true" tabindex="-1"></a>pr.out <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(nci.data, <span class="at">scale=</span><span class="cn">TRUE</span>)</span>
<span id="cb61-450"><a href="#cb61-450" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-451"><a href="#cb61-451" aria-hidden="true" tabindex="-1"></a><span class="co"># more cluster options</span></span>
<span id="cb61-452"><a href="#cb61-452" aria-hidden="true" tabindex="-1"></a>km.out2 <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(sd.data, <span class="dv">2</span>, <span class="at">nstart=</span><span class="dv">20</span>)</span>
<span id="cb61-453"><a href="#cb61-453" aria-hidden="true" tabindex="-1"></a>km.out3 <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(sd.data, <span class="dv">3</span>, <span class="at">nstart=</span><span class="dv">20</span>)</span>
<span id="cb61-454"><a href="#cb61-454" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-455"><a href="#cb61-455" aria-hidden="true" tabindex="-1"></a><span class="co"># we can now visualise the K-Means results by labelling the data points</span></span>
<span id="cb61-456"><a href="#cb61-456" aria-hidden="true" tabindex="-1"></a><span class="co"># in a plot of the scores of the first 2 principal components:</span></span>
<span id="cb61-457"><a href="#cb61-457" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>))</span>
<span id="cb61-458"><a href="#cb61-458" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pr.out<span class="sc">$</span>x[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="at">col=</span>(km.out2<span class="sc">$</span>cluster<span class="sc">+</span><span class="dv">1</span>), <span class="at">main=</span><span class="st">"K-Means with K=2"</span>,</span>
<span id="cb61-459"><a href="#cb61-459" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">"PC 1"</span>, <span class="at">ylab=</span><span class="st">"PC 2"</span>, <span class="at">pch=</span><span class="dv">20</span>)</span>
<span id="cb61-460"><a href="#cb61-460" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pr.out<span class="sc">$</span>x[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="at">col=</span>(km.out3<span class="sc">$</span>cluster<span class="sc">+</span><span class="dv">1</span>), <span class="at">main=</span><span class="st">"K-Means with K=3"</span>,</span>
<span id="cb61-461"><a href="#cb61-461" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">"PC 1"</span>,  <span class="at">ylab=</span><span class="st">"PC 2"</span>, <span class="at">pch=</span><span class="dv">20</span>)</span>
<span id="cb61-462"><a href="#cb61-462" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pr.out<span class="sc">$</span>x[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="at">col=</span>(km.out4<span class="sc">$</span>cluster<span class="sc">+</span><span class="dv">1</span>), <span class="at">main=</span><span class="st">"K-Means with K=4"</span>,</span>
<span id="cb61-463"><a href="#cb61-463" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">"PC 1"</span>, <span class="at">ylab=</span><span class="st">"PC 2"</span>, <span class="at">pch=</span><span class="dv">20</span>)</span>
<span id="cb61-464"><a href="#cb61-464" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-465"><a href="#cb61-465" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-466"><a href="#cb61-466" aria-hidden="true" tabindex="-1"></a>Compare with the plot from Exercise 2 yesterday (left panel) along with the cancer type labels. The clusters from K-means seem to correspond decently to partition the data into groups.</span>
<span id="cb61-467"><a href="#cb61-467" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-470"><a href="#cb61-470" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-471"><a href="#cb61-471" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="cb61-472"><a href="#cb61-472" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-473"><a href="#cb61-473" aria-hidden="true" tabindex="-1"></a>Cols<span class="ot">=</span><span class="cf">function</span>(vec){</span>
<span id="cb61-474"><a href="#cb61-474" aria-hidden="true" tabindex="-1"></a>  cols<span class="ot">=</span><span class="fu">rainbow</span>(<span class="fu">length</span>(<span class="fu">unique</span>(vec)))</span>
<span id="cb61-475"><a href="#cb61-475" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(cols[<span class="fu">as.numeric</span>(<span class="fu">as.factor</span>(vec))])</span>
<span id="cb61-476"><a href="#cb61-476" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb61-477"><a href="#cb61-477" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pr.out<span class="sc">$</span>x[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="at">col=</span><span class="fu">Cols</span>(nci.labs), <span class="at">pch=</span><span class="dv">19</span>,<span class="at">xlab=</span><span class="st">"PC 1"</span>,<span class="at">ylab=</span><span class="st">" PC 2"</span>)</span>
<span id="cb61-478"><a href="#cb61-478" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">'topleft'</span>, <span class="at">col=</span><span class="fu">rainbow</span>(<span class="fu">length</span>(<span class="fu">unique</span>(nci.labs))), <span class="at">legend=</span><span class="fu">unique</span>(nci.labs), <span class="at">bty=</span><span class="st">'n'</span>, <span class="at">lwd=</span><span class="dv">2</span>, <span class="at">cex=</span>.<span class="dv">6</span>)</span>
<span id="cb61-479"><a href="#cb61-479" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-480"><a href="#cb61-480" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-481"><a href="#cb61-481" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-482"><a href="#cb61-482" aria-hidden="true" tabindex="-1"></a><span class="fu">### Heatmap </span></span>
<span id="cb61-483"><a href="#cb61-483" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-484"><a href="#cb61-484" aria-hidden="true" tabindex="-1"></a>A heatmap is another way to visualize the clusters from the data. We use the principal components rather than the raw data, as PCs are already explaining a large amount of variability in the over 6000 features.</span>
<span id="cb61-485"><a href="#cb61-485" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-486"><a href="#cb61-486" aria-hidden="true" tabindex="-1"></a>Similar values are presented with similar colors. </span>
<span id="cb61-487"><a href="#cb61-487" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-490"><a href="#cb61-490" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-491"><a href="#cb61-491" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: clust-nci60-heatmap-pca</span></span>
<span id="cb61-492"><a href="#cb61-492" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-493"><a href="#cb61-493" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-494"><a href="#cb61-494" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-495"><a href="#cb61-495" aria-hidden="true" tabindex="-1"></a><span class="do">## We use the scores of the PCA on the NCI60 data, to reduce dimension</span></span>
<span id="cb61-496"><a href="#cb61-496" aria-hidden="true" tabindex="-1"></a>scores <span class="ot">&lt;-</span> pr.out<span class="sc">$</span>x</span>
<span id="cb61-497"><a href="#cb61-497" aria-hidden="true" tabindex="-1"></a>scores[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>] <span class="co"># first 5 pc, first 5 measurements</span></span>
<span id="cb61-498"><a href="#cb61-498" aria-hidden="true" tabindex="-1"></a><span class="co">#  default choices</span></span>
<span id="cb61-499"><a href="#cb61-499" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(pr.out<span class="sc">$</span>x)</span>
<span id="cb61-500"><a href="#cb61-500" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-501"><a href="#cb61-501" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-502"><a href="#cb61-502" aria-hidden="true" tabindex="-1"></a>You can remove the dendrogram on the PCs, only keeping the ones for cancer types. Now you see that the PCs have kept their original order from 1 to 64.</span>
<span id="cb61-503"><a href="#cb61-503" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-506"><a href="#cb61-506" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-507"><a href="#cb61-507" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: clust-nci60-heatmap-dendro</span></span>
<span id="cb61-508"><a href="#cb61-508" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-509"><a href="#cb61-509" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-510"><a href="#cb61-510" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-511"><a href="#cb61-511" aria-hidden="true" tabindex="-1"></a><span class="co"># hc.corr is the result from hclust. check the section on hierarchical clustering</span></span>
<span id="cb61-512"><a href="#cb61-512" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(pr.out<span class="sc">$</span>x, <span class="at">Rowv =</span> <span class="fu">as.dendrogram</span>(hc.corr), <span class="at">Colv =</span> <span class="cn">NA</span>)</span>
<span id="cb61-513"><a href="#cb61-513" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-514"><a href="#cb61-514" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-515"><a href="#cb61-515" aria-hidden="true" tabindex="-1"></a>You can also reduce the number of PCs, and add titles to the plot annd y-axis.</span>
<span id="cb61-516"><a href="#cb61-516" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-519"><a href="#cb61-519" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-520"><a href="#cb61-520" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: clust-nci60-heatmap-type</span></span>
<span id="cb61-521"><a href="#cb61-521" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-522"><a href="#cb61-522" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-523"><a href="#cb61-523" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-524"><a href="#cb61-524" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">cex.main =</span> .<span class="dv">7</span>)</span>
<span id="cb61-525"><a href="#cb61-525" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(pr.out<span class="sc">$</span>x[,<span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>], <span class="at">Rowv =</span> <span class="fu">as.dendrogram</span>(hc.corr), <span class="at">Colv =</span> <span class="cn">NA</span>,</span>
<span id="cb61-526"><a href="#cb61-526" aria-hidden="true" tabindex="-1"></a>        <span class="at">labRow =</span> nci.labs, <span class="at">main =</span> <span class="st">'Heatmap of the scores of the first 40 PCs on the NCI60 data'</span>)</span>
<span id="cb61-527"><a href="#cb61-527" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-528"><a href="#cb61-528" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-529"><a href="#cb61-529" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-530"><a href="#cb61-530" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-531"><a href="#cb61-531" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-532"><a href="#cb61-532" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-533"><a href="#cb61-533" aria-hidden="true" tabindex="-1"></a><span class="fu">## Exercise 3: Gene expression data</span></span>
<span id="cb61-534"><a href="#cb61-534" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-535"><a href="#cb61-535" aria-hidden="true" tabindex="-1"></a>(CH12Ex13 from statistical learning)</span>
<span id="cb61-536"><a href="#cb61-536" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-537"><a href="#cb61-537" aria-hidden="true" tabindex="-1"></a>We use the <span class="in">`Ch12Ex13.csv`</span> data to repeat some of the clustering analysis we did.</span>
<span id="cb61-538"><a href="#cb61-538" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-539"><a href="#cb61-539" aria-hidden="true" tabindex="-1"></a>The first 20 samples are from healthy patients, while the second 20 are from a diseased group.</span>
<span id="cb61-540"><a href="#cb61-540" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-541"><a href="#cb61-541" aria-hidden="true" tabindex="-1"></a>Load in the data using read.csv(). You will need to select header=F. Alternatively: load in the data using “Import dataset” in the upper right window, and click “no” on the “Heading” option.</span>
<span id="cb61-542"><a href="#cb61-542" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-543"><a href="#cb61-543" aria-hidden="true" tabindex="-1"></a>Carry out both hierarchical clustering and K-means clustering. You should choose the most meaningful number of clusters (think about how many groups of patients we have!). Compare the results. </span>
<span id="cb61-544"><a href="#cb61-544" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-545"><a href="#cb61-545" aria-hidden="true" tabindex="-1"></a>Note: remember that the data has genes on the rows and patients on the columns. You need to transpose the data so that the orders are reversed.</span>
<span id="cb61-546"><a href="#cb61-546" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-549"><a href="#cb61-549" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-550"><a href="#cb61-550" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: clust-ge-loaddata</span></span>
<span id="cb61-551"><a href="#cb61-551" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-552"><a href="#cb61-552" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-553"><a href="#cb61-553" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-554"><a href="#cb61-554" aria-hidden="true" tabindex="-1"></a><span class="co"># load in the data using read.csv(). You will need to select header=F.</span></span>
<span id="cb61-555"><a href="#cb61-555" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"data/Ch12Ex13.csv"</span>, <span class="at">header=</span><span class="cn">FALSE</span>)</span>
<span id="cb61-556"><a href="#cb61-556" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(data)</span>
<span id="cb61-557"><a href="#cb61-557" aria-hidden="true" tabindex="-1"></a><span class="co"># transpose the data, so that we have each row is one patient (subject)</span></span>
<span id="cb61-558"><a href="#cb61-558" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">t</span>(data) </span>
<span id="cb61-559"><a href="#cb61-559" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-560"><a href="#cb61-560" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-561"><a href="#cb61-561" aria-hidden="true" tabindex="-1"></a>Now the first 20 rows are measurements from healthy patients (group 0), and 21-50 rows are the disease patients (group 1). We can denote this information in a vector like this.</span>
<span id="cb61-562"><a href="#cb61-562" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-565"><a href="#cb61-565" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-566"><a href="#cb61-566" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: clust-ge-truelabel</span></span>
<span id="cb61-567"><a href="#cb61-567" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-568"><a href="#cb61-568" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-569"><a href="#cb61-569" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-570"><a href="#cb61-570" aria-hidden="true" tabindex="-1"></a>true.groups <span class="ot">&lt;-</span> <span class="fu">c</span>( <span class="fu">rep</span>(<span class="dv">0</span>,<span class="dv">20</span>), <span class="fu">rep</span>(<span class="dv">1</span>,<span class="dv">20</span>))</span>
<span id="cb61-571"><a href="#cb61-571" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-572"><a href="#cb61-572" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-573"><a href="#cb61-573" aria-hidden="true" tabindex="-1"></a><span class="fu">### Hierarchical clustering </span></span>
<span id="cb61-574"><a href="#cb61-574" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-575"><a href="#cb61-575" aria-hidden="true" tabindex="-1"></a>You can use different linkage options and distance metrics of your choosing. For example, with complete linkage the code is like this.</span>
<span id="cb61-576"><a href="#cb61-576" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-579"><a href="#cb61-579" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-580"><a href="#cb61-580" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: clust-ge-clust</span></span>
<span id="cb61-581"><a href="#cb61-581" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-582"><a href="#cb61-582" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-583"><a href="#cb61-583" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-584"><a href="#cb61-584" aria-hidden="true" tabindex="-1"></a>data.dist <span class="ot">&lt;-</span> <span class="fu">dist</span>(data) <span class="co"># need to compute the distance matrix</span></span>
<span id="cb61-585"><a href="#cb61-585" aria-hidden="true" tabindex="-1"></a>hclust.df <span class="ot">&lt;-</span> <span class="fu">hclust</span>(data.dist, <span class="at">method=</span><span class="st">"complete"</span> )</span>
<span id="cb61-586"><a href="#cb61-586" aria-hidden="true" tabindex="-1"></a><span class="co">#alternatives:</span></span>
<span id="cb61-587"><a href="#cb61-587" aria-hidden="true" tabindex="-1"></a><span class="co">#hclust.df &lt;- hclust( D, method="average" )</span></span>
<span id="cb61-588"><a href="#cb61-588" aria-hidden="true" tabindex="-1"></a><span class="co">#hclust.df &lt;- hclust( D, method="single" )</span></span>
<span id="cb61-589"><a href="#cb61-589" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-590"><a href="#cb61-590" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-591"><a href="#cb61-591" aria-hidden="true" tabindex="-1"></a>We can keep 2 clusters with <span class="in">`cutree`</span>. Then do a cross tabulation of the true labels and clustered results: how well do they correspond?</span>
<span id="cb61-592"><a href="#cb61-592" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-595"><a href="#cb61-595" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-596"><a href="#cb61-596" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: clust-ge-clust-res</span></span>
<span id="cb61-597"><a href="#cb61-597" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-598"><a href="#cb61-598" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-599"><a href="#cb61-599" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-600"><a href="#cb61-600" aria-hidden="true" tabindex="-1"></a><span class="co"># find the clusters</span></span>
<span id="cb61-601"><a href="#cb61-601" aria-hidden="true" tabindex="-1"></a>predicted <span class="ot">&lt;-</span> <span class="fu">cutree</span>( hclust.df, <span class="at">k=</span><span class="dv">2</span> )</span>
<span id="cb61-602"><a href="#cb61-602" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-603"><a href="#cb61-603" aria-hidden="true" tabindex="-1"></a><span class="co"># How well does our clustering predict health vs. diseased</span></span>
<span id="cb61-604"><a href="#cb61-604" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(predicted, true.groups )</span>
<span id="cb61-605"><a href="#cb61-605" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-606"><a href="#cb61-606" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-607"><a href="#cb61-607" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-608"><a href="#cb61-608" aria-hidden="true" tabindex="-1"></a><span class="fu">### K-means</span></span>
<span id="cb61-609"><a href="#cb61-609" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-610"><a href="#cb61-610" aria-hidden="true" tabindex="-1"></a>Now you can use K-means to identify 2 clusters. </span>
<span id="cb61-611"><a href="#cb61-611" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-614"><a href="#cb61-614" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb61-615"><a href="#cb61-615" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: clust-ge-km</span></span>
<span id="cb61-616"><a href="#cb61-616" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb61-617"><a href="#cb61-617" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb61-618"><a href="#cb61-618" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-619"><a href="#cb61-619" aria-hidden="true" tabindex="-1"></a>predicted.kmean <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(data, <span class="dv">2</span>, <span class="at">nstart=</span><span class="dv">20</span>)<span class="sc">$</span>cluster</span>
<span id="cb61-620"><a href="#cb61-620" aria-hidden="true" tabindex="-1"></a><span class="co"># agreement with true label</span></span>
<span id="cb61-621"><a href="#cb61-621" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(predicted.kmean, true.groups )</span>
<span id="cb61-622"><a href="#cb61-622" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb61-623"><a href="#cb61-623" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-624"><a href="#cb61-624" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-625"><a href="#cb61-625" aria-hidden="true" tabindex="-1"></a>Both methods seem to do work decently for the task. </span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->




</body></html>