<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.552">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>MED3007 V24 - R Lab (day 2): Multiple Testing</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">MED3007 V24</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../get_started/get_started.html"> 
<span class="menu-text">Get Started</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../lab/overview.html"> 
<span class="menu-text">Lab and Code</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/ocbe-uio/course_med3007"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#exercise-1-brain-data" id="toc-exercise-1-brain-data" class="nav-link active" data-scroll-target="#exercise-1-brain-data">Exercise 1: brain data</a>
  <ul class="collapse">
  <li><a href="#t-test" id="toc-t-test" class="nav-link" data-scroll-target="#t-test">t-test</a></li>
  </ul></li>
  <li><a href="#exercise-2" id="toc-exercise-2" class="nav-link" data-scroll-target="#exercise-2">Exercise 2</a></li>
  <li><a href="#exercise-3-nci60" id="toc-exercise-3-nci60" class="nav-link" data-scroll-target="#exercise-3-nci60">Exercise 3: NCI60</a>
  <ul class="collapse">
  <li><a href="#t-test-for-genes" id="toc-t-test-for-genes" class="nav-link" data-scroll-target="#t-test-for-genes">t-test for genes</a></li>
  <li><a href="#multiple-testing-correction" id="toc-multiple-testing-correction" class="nav-link" data-scroll-target="#multiple-testing-correction">Multiple testing correction</a></li>
  </ul></li>
  <li><a href="#exercise-4" id="toc-exercise-4" class="nav-link" data-scroll-target="#exercise-4">Exercise 4</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">R Lab (day 2): Multiple Testing</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>Download datasets <a href="https://github.com/ocbe-uio/course_med3007/tree/main/lab/data">here</a> or from Canvas.</p>
<p>R scripts</p>
<ul>
<li><a href="https://github.com/ocbe-uio/course_med3007/blob/main/lab/code/MED3007_Lab1.R">Code</a></li>
<li><a href="https://github.com/ocbe-uio/course_med3007/blob/main/lab/code/MED3007_Lab1_exercise_solution.R">Code (solution)</a></li>
</ul>
<section id="exercise-1-brain-data" class="level2">
<h2 class="anchored" data-anchor-id="exercise-1-brain-data">Exercise 1: brain data</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">'data/data_brainshake.RData'</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(mydatanew)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  ID kjonn Alder  BMI time tot.kol LDL.kol HDL.kol triglyserider Hexenal
1  1     0    20 24.4   t1       5      12       2             2      35
2  2     1    22 20.7   t1      27      22      16             5      41
3  3     1    20 20.0   t1      18      20       6             9      31
4  4     0    20 22.3   t1       4       6       7             3      48
6  6     1    23 22.3   t1      20      23       5             6      30
7  7     1    25 19.3   t1       6       5       9             6      38
  Nonenal X8isoPGF.creat  GSH   GR    GPx   CAT X18_1_9 X18_2 X18_3 X20_4 X20_5
1       3              1 1.18 2.80 115.76 10.11      19    39    28     8    38
2       6             53 1.29 9.51 131.62 10.25      10    52    30    12    14
3       2             45 1.68 6.08 111.83 11.00      47    26     7    35     3
4       1              4 1.53 8.81  93.38  7.32      43    12     9    31    35
6      10              5 1.02 6.86 107.73  7.86      27    44    14    12     5
7       9             14 1.62 6.81 119.88 10.28      26    28    20    19    18
  X22_5 X22_6
1    20    36
2     9    12
3     2     6
4    25    33
6     3     4
7    23    32</code></pre>
</div>
</div>
<p>Some descriptive statistics of each variable</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(data) <span class="co"># summary for each variable</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       ID       kjonn       Alder           BMI        time       tot.kol     
 Min.   : 1.0   0: 45   Min.   :19.0   Min.   :17.30   t1:54   Min.   : 1.00  
 1st Qu.:18.0   1:117   1st Qu.:22.0   1st Qu.:21.00   t2:54   1st Qu.: 9.00  
 Median :43.5           Median :25.0   Median :22.40   t3:54   Median :13.00  
 Mean   :41.8           Mean   :26.8   Mean   :22.63           Mean   :13.66  
 3rd Qu.:63.0           3rd Qu.:30.0   3rd Qu.:23.80           3rd Qu.:18.00  
 Max.   :90.0           Max.   :49.0   Max.   :30.10           Max.   :30.00  
                                                                              
    LDL.kol         HDL.kol       triglyserider       Hexenal     
 Min.   : 1.00   Min.   : 1.000   Min.   : 1.000   Min.   : 1.00  
 1st Qu.:10.00   1st Qu.: 5.000   1st Qu.: 4.000   1st Qu.:11.00  
 Median :14.00   Median : 8.000   Median : 6.000   Median :24.00  
 Mean   :14.03   Mean   : 8.167   Mean   : 7.457   Mean   :24.27  
 3rd Qu.:18.00   3rd Qu.:10.750   3rd Qu.:10.000   3rd Qu.:37.00  
 Max.   :27.00   Max.   :19.000   Max.   :21.000   Max.   :52.00  
                                                                  
    Nonenal      X8isoPGF.creat       GSH              GR        
 Min.   : 1.00   Min.   : 1.00   Min.   :0.540   Min.   : 2.800  
 1st Qu.:11.25   1st Qu.:12.25   1st Qu.:1.180   1st Qu.: 7.058  
 Median :24.50   Median :25.50   Median :1.500   Median : 7.900  
 Mean   :24.18   Mean   :25.76   Mean   :1.523   Mean   : 7.878  
 3rd Qu.:36.00   3rd Qu.:39.00   3rd Qu.:1.860   3rd Qu.: 8.740  
 Max.   :52.00   Max.   :54.00   Max.   :2.890   Max.   :10.580  
                                 NA's   :7       NA's   :8       
      GPx              CAT            X18_1_9          X18_2      
 Min.   : 81.33   Min.   : 6.950   Min.   : 1.00   Min.   : 1.00  
 1st Qu.:106.33   1st Qu.: 8.880   1st Qu.:14.00   1st Qu.:14.00  
 Median :115.75   Median : 9.595   Median :27.00   Median :27.00  
 Mean   :115.99   Mean   : 9.615   Mean   :27.02   Mean   :27.32  
 3rd Qu.:123.76   3rd Qu.:10.370   3rd Qu.:40.00   3rd Qu.:40.75  
 Max.   :181.79   Max.   :14.290   Max.   :55.00   Max.   :55.00  
 NA's   :7        NA's   :10                                      
     X18_3           X20_4           X20_5           X22_5      
 Min.   : 1.00   Min.   : 1.00   Min.   : 1.00   Min.   : 1.00  
 1st Qu.: 9.00   1st Qu.:13.25   1st Qu.:11.00   1st Qu.:11.00  
 Median :15.00   Median :26.50   Median :22.00   Median :18.00  
 Mean   :16.49   Mean   :25.93   Mean   :22.54   Mean   :18.26  
 3rd Qu.:23.00   3rd Qu.:38.00   3rd Qu.:33.75   3rd Qu.:26.00  
 Max.   :42.00   Max.   :53.00   Max.   :49.00   Max.   :39.00  
                                                                
     X22_6      
 Min.   : 1.00  
 1st Qu.:12.25  
 Median :25.00  
 Mean   :24.85  
 3rd Qu.:36.00  
 Max.   :52.00  
                </code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(data) <span class="co"># column names</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "ID"             "kjonn"          "Alder"          "BMI"           
 [5] "time"           "tot.kol"        "LDL.kol"        "HDL.kol"       
 [9] "triglyserider"  "Hexenal"        "Nonenal"        "X8isoPGF.creat"
[13] "GSH"            "GR"             "GPx"            "CAT"           
[17] "X18_1_9"        "X18_2"          "X18_3"          "X20_4"         
[21] "X20_5"          "X22_5"          "X22_6"         </code></pre>
</div>
</div>
<p>Ddescriptive plots: boxplot and histogram (for one specific variable across groups)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)) <span class="co"># plot 2 figures side by side</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(BMI <span class="sc">~</span> kjonn, data)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(tot.kol <span class="sc">~</span> kjonn, data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day2_testing_files/figure-html/ttest-brain-plots-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(data<span class="sc">$</span>BMI)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># histogram</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>BMI_group1 <span class="ot">&lt;-</span> data[<span class="fu">which</span>(data<span class="sc">$</span>kjonn<span class="sc">==</span><span class="dv">0</span>),<span class="st">'BMI'</span>]</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>BMI_group2 <span class="ot">&lt;-</span> data[<span class="fu">which</span>(data<span class="sc">$</span>kjonn<span class="sc">==</span><span class="dv">1</span>),<span class="st">'BMI'</span>]</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(BMI_group1) <span class="co">#men</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day2_testing_files/figure-html/ttest-brain-plots-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(BMI_group2) <span class="co">#females</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day2_testing_files/figure-html/ttest-brain-plots-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<section id="t-test" class="level3">
<h3 class="anchored" data-anchor-id="t-test">t-test</h3>
<p>We carry out a t-test to find out whether there is a significant difference for BMI between the two genders.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">t.test</span>(data[<span class="fu">which</span>(data<span class="sc">$</span>kjonn<span class="sc">==</span><span class="dv">0</span>),<span class="st">'BMI'</span>],data[<span class="fu">which</span>(data<span class="sc">$</span>kjonn<span class="sc">==</span><span class="dv">1</span>),<span class="st">'BMI'</span>], <span class="at">alternative=</span><span class="st">'greater'</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co"># much more clever syntax</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">t.test</span>(BMI <span class="sc">~</span> kjonn, data, <span class="at">alternative=</span><span class="st">'greater'</span>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>out</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Welch Two Sample t-test

data:  BMI by kjonn
t = 2.5952, df = 67.68, p-value = 0.005791
alternative hypothesis: true difference in means between group 0 and group 1 is greater than 0
95 percent confidence interval:
 0.443913      Inf
sample estimates:
mean in group 0 mean in group 1 
       23.52667        22.28462 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(out)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "statistic"   "parameter"   "p.value"     "conf.int"    "estimate"   
 [6] "null.value"  "stderr"      "alternative" "method"      "data.name"  </code></pre>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>out<span class="sc">$</span>p.value</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.00579067</code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>pvalue <span class="ot">&lt;-</span> out<span class="sc">$</span>p.value</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>pvalue</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.00579067</code></pre>
</div>
</div>
<p>You can also just run the t-test and not save it to an object. But it is often useful to save it, because you have access to the p-values and do not need to copy paste it from the console.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(BMI <span class="sc">~</span> kjonn, data, <span class="at">alternative=</span><span class="st">'greater'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Welch Two Sample t-test

data:  BMI by kjonn
t = 2.5952, df = 67.68, p-value = 0.005791
alternative hypothesis: true difference in means between group 0 and group 1 is greater than 0
95 percent confidence interval:
 0.443913      Inf
sample estimates:
mean in group 0 mean in group 1 
       23.52667        22.28462 </code></pre>
</div>
</div>
<p>Check normality assumption.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># QQ-plots:</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)) <span class="co"># plot-window with two columns</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="fu">qqnorm</span>(BMI_group1, <span class="at">main =</span> <span class="st">'BMI - male'</span>)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="fu">qqline</span>(BMI_group1)</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="fu">qqnorm</span>(BMI_group2, <span class="at">main =</span> <span class="st">'BMI - female'</span>)</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="fu">qqline</span>(BMI_group2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day2_testing_files/figure-html/ttest-brain-qq-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># test data for normality</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">shapiro.test</span>(BMI_group1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Shapiro-Wilk normality test

data:  BMI_group1
W = 0.89644, p-value = 0.0007354</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">shapiro.test</span>(BMI_group2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Shapiro-Wilk normality test

data:  BMI_group2
W = 0.88549, p-value = 5.161e-08</code></pre>
</div>
</div>
</section>
</section>
<section id="exercise-2" class="level2">
<h2 class="anchored" data-anchor-id="exercise-2">Exercise 2</h2>
<p>Load the data from the file “Testfil_Rcourse.xlsx” and consider the variable <code>vitD_v1</code>.</p>
<ol type="1">
<li>plot the histogram of this variable and save it, and then also plot the boxplot of vitD_v1 stratified according to gender.</li>
</ol>
<p>Does this variable look normally distributed?</p>
<ol start="2" type="1">
<li>Perform a t-test to verify that <code>vitD_v1</code> is different across gender groups</li>
</ol>
<p>(Solution see R script <a href="https://github.com/ocbe-uio/course_med3007/blob/main/lab/code/MED3007_Lab1_exercise_solution.R">MED3007_Lab1_exercise_solution.R</a>)</p>
</section>
<section id="exercise-3-nci60" class="level2">
<h2 class="anchored" data-anchor-id="exercise-3-nci60">Exercise 3: NCI60</h2>
<p>Now we move to a typical large-scale biological data set which will be our running example throughout the course.</p>
<p>The NCI60 cancer cell line microarray data set consists of 6830 gene expression measurements on 64 cancer cell lines. It is available in the R package ISLR, which is the compendium R package to the book by James et al.&nbsp;(2013).</p>
<p>(It is not really a testing example, but we use it this way for illustration)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">'data/NCI60.RData'</span>)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>nci.labs <span class="ot">&lt;-</span> NCI60<span class="sc">$</span>labs <span class="co"># Sample labels (tissue type)</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>nci.data <span class="ot">&lt;-</span> NCI60<span class="sc">$</span>data <span class="co"># Gene expression data set</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a><span class="co"># some basic summary</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(nci.labs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 64</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(nci.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]   64 6830</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(nci.labs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>nci.labs
     BREAST         CNS       COLON K562A-repro K562B-repro    LEUKEMIA 
          7           5           7           1           1           6 
MCF7A-repro MCF7D-repro    MELANOMA       NSCLC     OVARIAN    PROSTATE 
          1           1           8           9           6           2 
      RENAL     UNKNOWN 
          9           1 </code></pre>
</div>
</div>
<p>We create a grouping in 2 macro-groups of cancers, to test for gene expression differences across these 2 groups only. Those 2 groups should be relatively far in terms of genomic characteristics.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>ind1 <span class="ot">&lt;-</span> <span class="fu">which</span>(nci.labs<span class="sc">==</span><span class="st">'BREAST'</span> <span class="sc">|</span> nci.labs<span class="sc">==</span><span class="st">'OVARIAN'</span> <span class="sc">|</span> nci.labs<span class="sc">==</span><span class="st">'PROSTATE'</span>)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>ind2 <span class="ot">&lt;-</span> <span class="fu">which</span>(nci.labs<span class="sc">==</span><span class="st">'LEUKEMIA'</span> <span class="sc">|</span> nci.labs<span class="sc">==</span><span class="st">'MELANOMA'</span> <span class="sc">|</span> nci.labs<span class="sc">==</span><span class="st">'NSCLC'</span> <span class="sc">|</span> nci.labs<span class="sc">==</span><span class="st">'RENAL'</span>) <span class="co"># blood, skin, lung, kidney</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>cancer.type <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">1</span>, <span class="fu">length</span>(ind1)), <span class="fu">rep</span>(<span class="dv">2</span>, <span class="fu">length</span>(ind2))) <span class="co"># we have two "cancer type" groups</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>mydata <span class="ot">&lt;-</span> nci.data[<span class="fu">c</span>(ind1, ind2) , ]</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(mydata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]   47 6830</code></pre>
</div>
</div>
<p>Some descriptive statistics</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># mean across patients for each gene</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>genes.means <span class="ot">&lt;-</span> <span class="fu">apply</span>(mydata, <span class="dv">2</span>, mean) <span class="co">#1=rows, 2=columns</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>genes.var <span class="ot">&lt;-</span> <span class="fu">apply</span>(mydata, <span class="dv">2</span>, var)</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(genes.means, <span class="at">xlab =</span> <span class="st">'genes'</span>, <span class="at">main =</span> <span class="st">'mean across samples'</span>, <span class="at">ylab =</span> <span class="st">'mean expr'</span>)</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h=</span><span class="dv">0</span>, <span class="at">col=</span><span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day2_testing_files/figure-html/ttest-nci60-summary-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># we can try to compute the within group means and plot in different colors..</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>genes.gr.means <span class="ot">&lt;-</span> <span class="fu">apply</span>(mydata, <span class="dv">2</span>, <span class="cf">function</span>(x){<span class="fu">tapply</span>(x, cancer.type, mean)})</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(genes.gr.means[<span class="dv">1</span>,], <span class="at">xlab =</span> <span class="st">'genes'</span>, <span class="at">main =</span> <span class="st">'within-group mean across samples'</span>, </span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylim =</span> <span class="fu">range</span>(genes.gr.means), <span class="at">ylab =</span> <span class="st">'mean expr'</span>)</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(genes.gr.means[<span class="dv">2</span>,], <span class="at">col=</span><span class="dv">2</span>)</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h=</span><span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day2_testing_files/figure-html/ttest-nci60-summary-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<section id="t-test-for-genes" class="level3">
<h3 class="anchored" data-anchor-id="t-test-for-genes">t-test for genes</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>alpha <span class="ot">&lt;-</span> .<span class="dv">05</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>pval.ttest <span class="ot">&lt;-</span> <span class="fu">apply</span>(mydata, <span class="dv">2</span>, <span class="cf">function</span>(x){<span class="fu">t.test</span>(x[<span class="fu">which</span>(cancer.type<span class="sc">==</span><span class="dv">1</span>)], x[<span class="fu">which</span>(cancer.type<span class="sc">==</span><span class="dv">2</span>)])<span class="sc">$</span>p.value})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Investigate closely on one gene</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> mydata[,<span class="dv">1</span>] <span class="co">#gene 1</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>x1 <span class="ot">&lt;-</span> x[<span class="fu">which</span>(cancer.type<span class="sc">==</span><span class="dv">1</span>)] <span class="co">#select values for group 1</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>x2 <span class="ot">&lt;-</span> x[<span class="fu">which</span>(cancer.type<span class="sc">==</span><span class="dv">2</span>)] <span class="co">#select values for group 2</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">t.test</span>(x1, x2) <span class="co">#do the t-test</span></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>res <span class="co">#show a summary of the test result</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Welch Two Sample t-test

data:  x1 and x2
t = 0.27023, df = 23.473, p-value = 0.7893
alternative hypothesis: true difference in means is not equal to 0
95 percent confidence interval:
 -0.2562172  0.3333137
sample estimates:
  mean of x   mean of y 
 0.01433073 -0.02421753 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>p.value <span class="co">#extract the p-value</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.7893453</code></pre>
</div>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(pval.ttest)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day2_testing_files/figure-html/ttest-nci60-onegene-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># bit more details for nicer plotting</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(pval.ttest, <span class="at">breaks =</span> <span class="dv">1</span><span class="sc">/</span>alpha, <span class="at">xlab =</span> <span class="st">'p-values'</span>, <span class="at">ylab =</span> <span class="st">'counts'</span>, </span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">'Histogram of t-test p-values across genes'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab_day2_testing_files/figure-html/ttest-nci60-onegene-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The histogram of the p-values shows nearly uniform values but we see a slight peak around 0.</p>
</section>
<section id="multiple-testing-correction" class="level3">
<h3 class="anchored" data-anchor-id="multiple-testing-correction">Multiple testing correction</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># How many significant p-values (without any correction)</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(pval.ttest <span class="sc">&lt;</span> alpha)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 562</code></pre>
</div>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># how many expected false positives?</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>Vexp <span class="ot">&lt;-</span> alpha<span class="sc">*</span><span class="fu">dim</span>(mydata)[<span class="dv">2</span>]  <span class="co"># this is the expected value of V from the slides</span></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>Vexp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 341.5</code></pre>
</div>
</div>
<p>We adjust the p-values</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Simple way of calculation adjusted p-values (using p.adjust()):</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>pval.fwer <span class="ot">&lt;-</span> <span class="fu">p.adjust</span>(pval.ttest, <span class="at">method =</span> <span class="st">"bonferroni"</span>)</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>pval.fdr <span class="ot">&lt;-</span> <span class="fu">p.adjust</span>(pval.ttest, <span class="at">method =</span> <span class="st">"BH"</span>)</span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of significant p-values after Bonferroni correction</span></span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(pval.fwer <span class="sc">&lt;</span> alpha) <span class="co"># conservative</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of significant p-values after BH correction</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(pval.fdr  <span class="sc">&lt;</span> alpha)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0</code></pre>
</div>
</div>
<p>Conclusion: no significant difference in gene expression across the groups!</p>
</section>
</section>
<section id="exercise-4" class="level2">
<h2 class="anchored" data-anchor-id="exercise-4">Exercise 4</h2>
<p>Consider the gene expression data set “Ch10Ex11.csv” that consists of 40 tissue samples with measurements on 1,000 genes. The first 20 samples are from healthy patients, while the second 20 are from a diseased group.</p>
<ol type="1">
<li><p>Load in the data using read.csv(). You will need to select header=F.</p></li>
<li><p>Have a look at the data and describe them with appropriate descriptive measures.</p></li>
<li><p>Your collaborator wants to know which genes differ the most across the two groups.</p></li>
</ol>
<p>Suggest a way to answer this question, and apply it here.</p>
<p>(Solution see R script <a href="https://github.com/ocbe-uio/course_med3007/blob/main/lab/code/MED3007_Lab1_exercise_solution.R">MED3007_Lab1_exercise_solution.R</a>)</p>


<!-- -->

</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb50" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "R Lab (day 2): Multiple Testing"</span></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span><span class="co"> </span></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a><span class="co">    code-fold: false</span></span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools: true</span></span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a>Download datasets <span class="co">[</span><span class="ot">here</span><span class="co">](https://github.com/ocbe-uio/course_med3007/tree/main/lab/data)</span> or from Canvas.</span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-11"><a href="#cb50-11" aria-hidden="true" tabindex="-1"></a>R scripts</span>
<span id="cb50-12"><a href="#cb50-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-13"><a href="#cb50-13" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="co">[</span><span class="ot">Code</span><span class="co">](https://github.com/ocbe-uio/course_med3007/blob/main/lab/code/MED3007_Lab1.R)</span></span>
<span id="cb50-14"><a href="#cb50-14" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="co">[</span><span class="ot">Code (solution)</span><span class="co">](https://github.com/ocbe-uio/course_med3007/blob/main/lab/code/MED3007_Lab1_exercise_solution.R)</span></span>
<span id="cb50-15"><a href="#cb50-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-16"><a href="#cb50-16" aria-hidden="true" tabindex="-1"></a><span class="fu">## Exercise 1: brain data</span></span>
<span id="cb50-17"><a href="#cb50-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-20"><a href="#cb50-20" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-21"><a href="#cb50-21" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: ttest-brain-loaddata</span></span>
<span id="cb50-22"><a href="#cb50-22" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb50-23"><a href="#cb50-23" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb50-24"><a href="#cb50-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-25"><a href="#cb50-25" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">'data/data_brainshake.RData'</span>)</span>
<span id="cb50-26"><a href="#cb50-26" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(mydatanew)</span>
<span id="cb50-27"><a href="#cb50-27" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(data)</span>
<span id="cb50-28"><a href="#cb50-28" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-29"><a href="#cb50-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-30"><a href="#cb50-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-31"><a href="#cb50-31" aria-hidden="true" tabindex="-1"></a>Some descriptive statistics of each variable</span>
<span id="cb50-32"><a href="#cb50-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-35"><a href="#cb50-35" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-36"><a href="#cb50-36" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: ttest-brain-summary</span></span>
<span id="cb50-37"><a href="#cb50-37" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb50-38"><a href="#cb50-38" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb50-39"><a href="#cb50-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-40"><a href="#cb50-40" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(data) <span class="co"># summary for each variable</span></span>
<span id="cb50-41"><a href="#cb50-41" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(data) <span class="co"># column names</span></span>
<span id="cb50-42"><a href="#cb50-42" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-43"><a href="#cb50-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-44"><a href="#cb50-44" aria-hidden="true" tabindex="-1"></a>Ddescriptive plots: boxplot and histogram (for one specific variable across groups)</span>
<span id="cb50-45"><a href="#cb50-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-48"><a href="#cb50-48" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-49"><a href="#cb50-49" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: ttest-brain-plots</span></span>
<span id="cb50-50"><a href="#cb50-50" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb50-51"><a href="#cb50-51" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb50-52"><a href="#cb50-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-53"><a href="#cb50-53" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)) <span class="co"># plot 2 figures side by side</span></span>
<span id="cb50-54"><a href="#cb50-54" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(BMI <span class="sc">~</span> kjonn, data)</span>
<span id="cb50-55"><a href="#cb50-55" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(tot.kol <span class="sc">~</span> kjonn, data)</span>
<span id="cb50-56"><a href="#cb50-56" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(data<span class="sc">$</span>BMI)</span>
<span id="cb50-57"><a href="#cb50-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-58"><a href="#cb50-58" aria-hidden="true" tabindex="-1"></a><span class="co"># histogram</span></span>
<span id="cb50-59"><a href="#cb50-59" aria-hidden="true" tabindex="-1"></a>BMI_group1 <span class="ot">&lt;-</span> data[<span class="fu">which</span>(data<span class="sc">$</span>kjonn<span class="sc">==</span><span class="dv">0</span>),<span class="st">'BMI'</span>]</span>
<span id="cb50-60"><a href="#cb50-60" aria-hidden="true" tabindex="-1"></a>BMI_group2 <span class="ot">&lt;-</span> data[<span class="fu">which</span>(data<span class="sc">$</span>kjonn<span class="sc">==</span><span class="dv">1</span>),<span class="st">'BMI'</span>]</span>
<span id="cb50-61"><a href="#cb50-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-62"><a href="#cb50-62" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(BMI_group1) <span class="co">#men</span></span>
<span id="cb50-63"><a href="#cb50-63" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(BMI_group2) <span class="co">#females</span></span>
<span id="cb50-64"><a href="#cb50-64" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-65"><a href="#cb50-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-66"><a href="#cb50-66" aria-hidden="true" tabindex="-1"></a><span class="fu">### t-test</span></span>
<span id="cb50-67"><a href="#cb50-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-68"><a href="#cb50-68" aria-hidden="true" tabindex="-1"></a>We carry out a t-test to find out whether there is a significant difference for BMI between the two genders.</span>
<span id="cb50-69"><a href="#cb50-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-72"><a href="#cb50-72" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-73"><a href="#cb50-73" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: ttest-brain-ttest1</span></span>
<span id="cb50-74"><a href="#cb50-74" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb50-75"><a href="#cb50-75" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb50-76"><a href="#cb50-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-77"><a href="#cb50-77" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">t.test</span>(data[<span class="fu">which</span>(data<span class="sc">$</span>kjonn<span class="sc">==</span><span class="dv">0</span>),<span class="st">'BMI'</span>],data[<span class="fu">which</span>(data<span class="sc">$</span>kjonn<span class="sc">==</span><span class="dv">1</span>),<span class="st">'BMI'</span>], <span class="at">alternative=</span><span class="st">'greater'</span>)</span>
<span id="cb50-78"><a href="#cb50-78" aria-hidden="true" tabindex="-1"></a><span class="co"># much more clever syntax</span></span>
<span id="cb50-79"><a href="#cb50-79" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">t.test</span>(BMI <span class="sc">~</span> kjonn, data, <span class="at">alternative=</span><span class="st">'greater'</span>)</span>
<span id="cb50-80"><a href="#cb50-80" aria-hidden="true" tabindex="-1"></a>out</span>
<span id="cb50-81"><a href="#cb50-81" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(out)</span>
<span id="cb50-82"><a href="#cb50-82" aria-hidden="true" tabindex="-1"></a>out<span class="sc">$</span>p.value</span>
<span id="cb50-83"><a href="#cb50-83" aria-hidden="true" tabindex="-1"></a>pvalue <span class="ot">&lt;-</span> out<span class="sc">$</span>p.value</span>
<span id="cb50-84"><a href="#cb50-84" aria-hidden="true" tabindex="-1"></a>pvalue</span>
<span id="cb50-85"><a href="#cb50-85" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-86"><a href="#cb50-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-87"><a href="#cb50-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-88"><a href="#cb50-88" aria-hidden="true" tabindex="-1"></a>You can also just run the t-test and not save it to an object. But it is often useful to save it, because you have access to the p-values and do not need to copy paste it from the console.</span>
<span id="cb50-89"><a href="#cb50-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-92"><a href="#cb50-92" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-93"><a href="#cb50-93" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: ttest-brain-ttest2</span></span>
<span id="cb50-94"><a href="#cb50-94" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb50-95"><a href="#cb50-95" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb50-96"><a href="#cb50-96" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(BMI <span class="sc">~</span> kjonn, data, <span class="at">alternative=</span><span class="st">'greater'</span>)</span>
<span id="cb50-97"><a href="#cb50-97" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-98"><a href="#cb50-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-99"><a href="#cb50-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-100"><a href="#cb50-100" aria-hidden="true" tabindex="-1"></a>Check normality assumption.</span>
<span id="cb50-101"><a href="#cb50-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-104"><a href="#cb50-104" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-105"><a href="#cb50-105" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: ttest-brain-qq</span></span>
<span id="cb50-106"><a href="#cb50-106" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb50-107"><a href="#cb50-107" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb50-108"><a href="#cb50-108" aria-hidden="true" tabindex="-1"></a><span class="co"># QQ-plots:</span></span>
<span id="cb50-109"><a href="#cb50-109" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)) <span class="co"># plot-window with two columns</span></span>
<span id="cb50-110"><a href="#cb50-110" aria-hidden="true" tabindex="-1"></a><span class="fu">qqnorm</span>(BMI_group1, <span class="at">main =</span> <span class="st">'BMI - male'</span>)</span>
<span id="cb50-111"><a href="#cb50-111" aria-hidden="true" tabindex="-1"></a><span class="fu">qqline</span>(BMI_group1)</span>
<span id="cb50-112"><a href="#cb50-112" aria-hidden="true" tabindex="-1"></a><span class="fu">qqnorm</span>(BMI_group2, <span class="at">main =</span> <span class="st">'BMI - female'</span>)</span>
<span id="cb50-113"><a href="#cb50-113" aria-hidden="true" tabindex="-1"></a><span class="fu">qqline</span>(BMI_group2)</span>
<span id="cb50-114"><a href="#cb50-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-115"><a href="#cb50-115" aria-hidden="true" tabindex="-1"></a><span class="co"># test data for normality</span></span>
<span id="cb50-116"><a href="#cb50-116" aria-hidden="true" tabindex="-1"></a><span class="fu">shapiro.test</span>(BMI_group1)</span>
<span id="cb50-117"><a href="#cb50-117" aria-hidden="true" tabindex="-1"></a><span class="fu">shapiro.test</span>(BMI_group2)</span>
<span id="cb50-118"><a href="#cb50-118" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-119"><a href="#cb50-119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-120"><a href="#cb50-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-121"><a href="#cb50-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-122"><a href="#cb50-122" aria-hidden="true" tabindex="-1"></a><span class="fu">## Exercise 2</span></span>
<span id="cb50-123"><a href="#cb50-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-124"><a href="#cb50-124" aria-hidden="true" tabindex="-1"></a>Load the data from the file "Testfil_Rcourse.xlsx" and consider the variable <span class="in">`vitD_v1`</span>.</span>
<span id="cb50-125"><a href="#cb50-125" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-126"><a href="#cb50-126" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>plot the histogram of this variable and save it, and then also plot the boxplot of vitD_v1 stratified according to gender.</span>
<span id="cb50-127"><a href="#cb50-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-128"><a href="#cb50-128" aria-hidden="true" tabindex="-1"></a>Does this variable look normally distributed?</span>
<span id="cb50-129"><a href="#cb50-129" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-130"><a href="#cb50-130" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Perform a t-test to verify that <span class="in">`vitD_v1`</span> is different across gender groups</span>
<span id="cb50-131"><a href="#cb50-131" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-132"><a href="#cb50-132" aria-hidden="true" tabindex="-1"></a>(Solution see R script <span class="co">[</span><span class="ot">MED3007_Lab1_exercise_solution.R</span><span class="co">](https://github.com/ocbe-uio/course_med3007/blob/main/lab/code/MED3007_Lab1_exercise_solution.R)</span>)</span>
<span id="cb50-133"><a href="#cb50-133" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-134"><a href="#cb50-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-135"><a href="#cb50-135" aria-hidden="true" tabindex="-1"></a><span class="fu">## Exercise 3: NCI60</span></span>
<span id="cb50-136"><a href="#cb50-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-137"><a href="#cb50-137" aria-hidden="true" tabindex="-1"></a>Now we move to a typical large-scale biological data set which will be our running example throughout the course.</span>
<span id="cb50-138"><a href="#cb50-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-139"><a href="#cb50-139" aria-hidden="true" tabindex="-1"></a>The NCI60 cancer cell line microarray data set consists of 6830 gene expression measurements on 64 cancer cell lines. It is available in the R package ISLR, which is the compendium R package to the book by James et al. (2013).</span>
<span id="cb50-140"><a href="#cb50-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-141"><a href="#cb50-141" aria-hidden="true" tabindex="-1"></a>(It is not really a testing example, but we use it this way for illustration)</span>
<span id="cb50-142"><a href="#cb50-142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-145"><a href="#cb50-145" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-146"><a href="#cb50-146" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: ttest-nci60-loaddata</span></span>
<span id="cb50-147"><a href="#cb50-147" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb50-148"><a href="#cb50-148" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb50-149"><a href="#cb50-149" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">'data/NCI60.RData'</span>)</span>
<span id="cb50-150"><a href="#cb50-150" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-151"><a href="#cb50-151" aria-hidden="true" tabindex="-1"></a>nci.labs <span class="ot">&lt;-</span> NCI60<span class="sc">$</span>labs <span class="co"># Sample labels (tissue type)</span></span>
<span id="cb50-152"><a href="#cb50-152" aria-hidden="true" tabindex="-1"></a>nci.data <span class="ot">&lt;-</span> NCI60<span class="sc">$</span>data <span class="co"># Gene expression data set</span></span>
<span id="cb50-153"><a href="#cb50-153" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-154"><a href="#cb50-154" aria-hidden="true" tabindex="-1"></a><span class="co"># some basic summary</span></span>
<span id="cb50-155"><a href="#cb50-155" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(nci.labs)</span>
<span id="cb50-156"><a href="#cb50-156" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(nci.data)</span>
<span id="cb50-157"><a href="#cb50-157" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(nci.labs)</span>
<span id="cb50-158"><a href="#cb50-158" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-159"><a href="#cb50-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-160"><a href="#cb50-160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-161"><a href="#cb50-161" aria-hidden="true" tabindex="-1"></a>We create a grouping in 2 macro-groups of cancers, to test for gene expression differences across these 2 groups only. Those 2 groups should be relatively far in terms of genomic characteristics.</span>
<span id="cb50-162"><a href="#cb50-162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-165"><a href="#cb50-165" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-166"><a href="#cb50-166" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: ttest-nci60-process</span></span>
<span id="cb50-167"><a href="#cb50-167" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb50-168"><a href="#cb50-168" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb50-169"><a href="#cb50-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-170"><a href="#cb50-170" aria-hidden="true" tabindex="-1"></a>ind1 <span class="ot">&lt;-</span> <span class="fu">which</span>(nci.labs<span class="sc">==</span><span class="st">'BREAST'</span> <span class="sc">|</span> nci.labs<span class="sc">==</span><span class="st">'OVARIAN'</span> <span class="sc">|</span> nci.labs<span class="sc">==</span><span class="st">'PROSTATE'</span>)</span>
<span id="cb50-171"><a href="#cb50-171" aria-hidden="true" tabindex="-1"></a>ind2 <span class="ot">&lt;-</span> <span class="fu">which</span>(nci.labs<span class="sc">==</span><span class="st">'LEUKEMIA'</span> <span class="sc">|</span> nci.labs<span class="sc">==</span><span class="st">'MELANOMA'</span> <span class="sc">|</span> nci.labs<span class="sc">==</span><span class="st">'NSCLC'</span> <span class="sc">|</span> nci.labs<span class="sc">==</span><span class="st">'RENAL'</span>) <span class="co"># blood, skin, lung, kidney</span></span>
<span id="cb50-172"><a href="#cb50-172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-173"><a href="#cb50-173" aria-hidden="true" tabindex="-1"></a>cancer.type <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">1</span>, <span class="fu">length</span>(ind1)), <span class="fu">rep</span>(<span class="dv">2</span>, <span class="fu">length</span>(ind2))) <span class="co"># we have two "cancer type" groups</span></span>
<span id="cb50-174"><a href="#cb50-174" aria-hidden="true" tabindex="-1"></a>mydata <span class="ot">&lt;-</span> nci.data[<span class="fu">c</span>(ind1, ind2) , ]</span>
<span id="cb50-175"><a href="#cb50-175" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(mydata)</span>
<span id="cb50-176"><a href="#cb50-176" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-177"><a href="#cb50-177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-178"><a href="#cb50-178" aria-hidden="true" tabindex="-1"></a>Some descriptive statistics</span>
<span id="cb50-179"><a href="#cb50-179" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-182"><a href="#cb50-182" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-183"><a href="#cb50-183" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: ttest-nci60-summary</span></span>
<span id="cb50-184"><a href="#cb50-184" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb50-185"><a href="#cb50-185" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb50-186"><a href="#cb50-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-187"><a href="#cb50-187" aria-hidden="true" tabindex="-1"></a><span class="co"># mean across patients for each gene</span></span>
<span id="cb50-188"><a href="#cb50-188" aria-hidden="true" tabindex="-1"></a>genes.means <span class="ot">&lt;-</span> <span class="fu">apply</span>(mydata, <span class="dv">2</span>, mean) <span class="co">#1=rows, 2=columns</span></span>
<span id="cb50-189"><a href="#cb50-189" aria-hidden="true" tabindex="-1"></a>genes.var <span class="ot">&lt;-</span> <span class="fu">apply</span>(mydata, <span class="dv">2</span>, var)</span>
<span id="cb50-190"><a href="#cb50-190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-191"><a href="#cb50-191" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(genes.means, <span class="at">xlab =</span> <span class="st">'genes'</span>, <span class="at">main =</span> <span class="st">'mean across samples'</span>, <span class="at">ylab =</span> <span class="st">'mean expr'</span>)</span>
<span id="cb50-192"><a href="#cb50-192" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h=</span><span class="dv">0</span>, <span class="at">col=</span><span class="dv">3</span>)</span>
<span id="cb50-193"><a href="#cb50-193" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-194"><a href="#cb50-194" aria-hidden="true" tabindex="-1"></a><span class="co"># we can try to compute the within group means and plot in different colors..</span></span>
<span id="cb50-195"><a href="#cb50-195" aria-hidden="true" tabindex="-1"></a>genes.gr.means <span class="ot">&lt;-</span> <span class="fu">apply</span>(mydata, <span class="dv">2</span>, <span class="cf">function</span>(x){<span class="fu">tapply</span>(x, cancer.type, mean)})</span>
<span id="cb50-196"><a href="#cb50-196" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(genes.gr.means[<span class="dv">1</span>,], <span class="at">xlab =</span> <span class="st">'genes'</span>, <span class="at">main =</span> <span class="st">'within-group mean across samples'</span>, </span>
<span id="cb50-197"><a href="#cb50-197" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylim =</span> <span class="fu">range</span>(genes.gr.means), <span class="at">ylab =</span> <span class="st">'mean expr'</span>)</span>
<span id="cb50-198"><a href="#cb50-198" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(genes.gr.means[<span class="dv">2</span>,], <span class="at">col=</span><span class="dv">2</span>)</span>
<span id="cb50-199"><a href="#cb50-199" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h=</span><span class="dv">0</span>)</span>
<span id="cb50-200"><a href="#cb50-200" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-201"><a href="#cb50-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-202"><a href="#cb50-202" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-203"><a href="#cb50-203" aria-hidden="true" tabindex="-1"></a><span class="fu">### t-test for genes</span></span>
<span id="cb50-204"><a href="#cb50-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-207"><a href="#cb50-207" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-208"><a href="#cb50-208" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: ttest-nci60-test</span></span>
<span id="cb50-209"><a href="#cb50-209" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb50-210"><a href="#cb50-210" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb50-211"><a href="#cb50-211" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-212"><a href="#cb50-212" aria-hidden="true" tabindex="-1"></a>alpha <span class="ot">&lt;-</span> .<span class="dv">05</span></span>
<span id="cb50-213"><a href="#cb50-213" aria-hidden="true" tabindex="-1"></a>pval.ttest <span class="ot">&lt;-</span> <span class="fu">apply</span>(mydata, <span class="dv">2</span>, <span class="cf">function</span>(x){<span class="fu">t.test</span>(x[<span class="fu">which</span>(cancer.type<span class="sc">==</span><span class="dv">1</span>)], x[<span class="fu">which</span>(cancer.type<span class="sc">==</span><span class="dv">2</span>)])<span class="sc">$</span>p.value})</span>
<span id="cb50-214"><a href="#cb50-214" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-215"><a href="#cb50-215" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-216"><a href="#cb50-216" aria-hidden="true" tabindex="-1"></a>Investigate closely on one gene</span>
<span id="cb50-217"><a href="#cb50-217" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-220"><a href="#cb50-220" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-221"><a href="#cb50-221" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: ttest-nci60-onegene</span></span>
<span id="cb50-222"><a href="#cb50-222" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb50-223"><a href="#cb50-223" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb50-224"><a href="#cb50-224" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-225"><a href="#cb50-225" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> mydata[,<span class="dv">1</span>] <span class="co">#gene 1</span></span>
<span id="cb50-226"><a href="#cb50-226" aria-hidden="true" tabindex="-1"></a>x1 <span class="ot">&lt;-</span> x[<span class="fu">which</span>(cancer.type<span class="sc">==</span><span class="dv">1</span>)] <span class="co">#select values for group 1</span></span>
<span id="cb50-227"><a href="#cb50-227" aria-hidden="true" tabindex="-1"></a>x2 <span class="ot">&lt;-</span> x[<span class="fu">which</span>(cancer.type<span class="sc">==</span><span class="dv">2</span>)] <span class="co">#select values for group 2</span></span>
<span id="cb50-228"><a href="#cb50-228" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">t.test</span>(x1, x2) <span class="co">#do the t-test</span></span>
<span id="cb50-229"><a href="#cb50-229" aria-hidden="true" tabindex="-1"></a>res <span class="co">#show a summary of the test result</span></span>
<span id="cb50-230"><a href="#cb50-230" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>p.value <span class="co">#extract the p-value</span></span>
<span id="cb50-231"><a href="#cb50-231" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-232"><a href="#cb50-232" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(pval.ttest)</span>
<span id="cb50-233"><a href="#cb50-233" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-234"><a href="#cb50-234" aria-hidden="true" tabindex="-1"></a><span class="co"># bit more details for nicer plotting</span></span>
<span id="cb50-235"><a href="#cb50-235" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(pval.ttest, <span class="at">breaks =</span> <span class="dv">1</span><span class="sc">/</span>alpha, <span class="at">xlab =</span> <span class="st">'p-values'</span>, <span class="at">ylab =</span> <span class="st">'counts'</span>, </span>
<span id="cb50-236"><a href="#cb50-236" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">'Histogram of t-test p-values across genes'</span>)</span>
<span id="cb50-237"><a href="#cb50-237" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-238"><a href="#cb50-238" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-239"><a href="#cb50-239" aria-hidden="true" tabindex="-1"></a>The histogram of the p-values shows nearly uniform values but we see a slight peak around 0.</span>
<span id="cb50-240"><a href="#cb50-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-241"><a href="#cb50-241" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-242"><a href="#cb50-242" aria-hidden="true" tabindex="-1"></a><span class="fu">### Multiple testing correction</span></span>
<span id="cb50-243"><a href="#cb50-243" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-246"><a href="#cb50-246" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-247"><a href="#cb50-247" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: ttest-nci60-pval</span></span>
<span id="cb50-248"><a href="#cb50-248" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb50-249"><a href="#cb50-249" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb50-250"><a href="#cb50-250" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-251"><a href="#cb50-251" aria-hidden="true" tabindex="-1"></a><span class="co"># How many significant p-values (without any correction)</span></span>
<span id="cb50-252"><a href="#cb50-252" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(pval.ttest <span class="sc">&lt;</span> alpha)</span>
<span id="cb50-253"><a href="#cb50-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-254"><a href="#cb50-254" aria-hidden="true" tabindex="-1"></a><span class="co"># how many expected false positives?</span></span>
<span id="cb50-255"><a href="#cb50-255" aria-hidden="true" tabindex="-1"></a>Vexp <span class="ot">&lt;-</span> alpha<span class="sc">*</span><span class="fu">dim</span>(mydata)[<span class="dv">2</span>]  <span class="co"># this is the expected value of V from the slides</span></span>
<span id="cb50-256"><a href="#cb50-256" aria-hidden="true" tabindex="-1"></a>Vexp</span>
<span id="cb50-257"><a href="#cb50-257" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-258"><a href="#cb50-258" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-259"><a href="#cb50-259" aria-hidden="true" tabindex="-1"></a>We adjust the p-values</span>
<span id="cb50-260"><a href="#cb50-260" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-263"><a href="#cb50-263" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-264"><a href="#cb50-264" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: ttest-nci60-pvaladjust</span></span>
<span id="cb50-265"><a href="#cb50-265" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb50-266"><a href="#cb50-266" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb50-267"><a href="#cb50-267" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-268"><a href="#cb50-268" aria-hidden="true" tabindex="-1"></a><span class="co"># Simple way of calculation adjusted p-values (using p.adjust()):</span></span>
<span id="cb50-269"><a href="#cb50-269" aria-hidden="true" tabindex="-1"></a>pval.fwer <span class="ot">&lt;-</span> <span class="fu">p.adjust</span>(pval.ttest, <span class="at">method =</span> <span class="st">"bonferroni"</span>)</span>
<span id="cb50-270"><a href="#cb50-270" aria-hidden="true" tabindex="-1"></a>pval.fdr <span class="ot">&lt;-</span> <span class="fu">p.adjust</span>(pval.ttest, <span class="at">method =</span> <span class="st">"BH"</span>)</span>
<span id="cb50-271"><a href="#cb50-271" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-272"><a href="#cb50-272" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of significant p-values after Bonferroni correction</span></span>
<span id="cb50-273"><a href="#cb50-273" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(pval.fwer <span class="sc">&lt;</span> alpha) <span class="co"># conservative</span></span>
<span id="cb50-274"><a href="#cb50-274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-275"><a href="#cb50-275" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of significant p-values after BH correction</span></span>
<span id="cb50-276"><a href="#cb50-276" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(pval.fdr  <span class="sc">&lt;</span> alpha)</span>
<span id="cb50-277"><a href="#cb50-277" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-278"><a href="#cb50-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-279"><a href="#cb50-279" aria-hidden="true" tabindex="-1"></a>Conclusion: no significant difference in gene expression across the groups! </span>
<span id="cb50-280"><a href="#cb50-280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-281"><a href="#cb50-281" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-282"><a href="#cb50-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-283"><a href="#cb50-283" aria-hidden="true" tabindex="-1"></a><span class="fu">## Exercise 4</span></span>
<span id="cb50-284"><a href="#cb50-284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-285"><a href="#cb50-285" aria-hidden="true" tabindex="-1"></a>Consider the gene expression data set "Ch10Ex11.csv" that consists of 40 tissue samples with measurements on 1,000 genes. The first 20 samples are from healthy patients, while the second 20 are from a diseased group.</span>
<span id="cb50-286"><a href="#cb50-286" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-287"><a href="#cb50-287" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Load in the data using read.csv(). You will need to select header=F.</span>
<span id="cb50-288"><a href="#cb50-288" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-289"><a href="#cb50-289" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Have a look at the data and describe them with appropriate descriptive measures.</span>
<span id="cb50-290"><a href="#cb50-290" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-291"><a href="#cb50-291" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>Your collaborator wants to know which genes differ the most across the two groups.</span>
<span id="cb50-292"><a href="#cb50-292" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-293"><a href="#cb50-293" aria-hidden="true" tabindex="-1"></a>Suggest a way to answer this question, and apply it here.</span>
<span id="cb50-294"><a href="#cb50-294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-295"><a href="#cb50-295" aria-hidden="true" tabindex="-1"></a>(Solution see R script <span class="co">[</span><span class="ot">MED3007_Lab1_exercise_solution.R</span><span class="co">](https://github.com/ocbe-uio/course_med3007/blob/main/lab/code/MED3007_Lab1_exercise_solution.R)</span>)</span>
<span id="cb50-296"><a href="#cb50-296" aria-hidden="true" tabindex="-1"></a></span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->




</body></html>