<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>MED3007 V24 - R Lab (day 3): Data exploration, Principal Component Analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">MED3007 V24</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../get_started/get_started.html" rel="" target="">
 <span class="menu-text">Get Started</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../lab/overview.html" rel="" target="">
 <span class="menu-text">Lab and Code</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/ocbe-uio/course_med3007" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#exercise-1-food" id="toc-exercise-1-food" class="nav-link active" data-scroll-target="#exercise-1-food">Exercise 1: Food</a>
  <ul class="collapse">
  <li><a href="#explore-the-dataset" id="toc-explore-the-dataset" class="nav-link" data-scroll-target="#explore-the-dataset">Explore the dataset</a></li>
  <li><a href="#principal-component-analysis" id="toc-principal-component-analysis" class="nav-link" data-scroll-target="#principal-component-analysis">Principal component analysis</a></li>
  <li><a href="#visualize-pca-results" id="toc-visualize-pca-results" class="nav-link" data-scroll-target="#visualize-pca-results">Visualize PCA results</a></li>
  </ul></li>
  <li><a href="#exercise-2-nci60" id="toc-exercise-2-nci60" class="nav-link" data-scroll-target="#exercise-2-nci60">Exercise 2: NCI60</a>
  <ul class="collapse">
  <li><a href="#carry-out-pca" id="toc-carry-out-pca" class="nav-link" data-scroll-target="#carry-out-pca">Carry out PCA</a></li>
  <li><a href="#how-many-pcs-to-keep" id="toc-how-many-pcs-to-keep" class="nav-link" data-scroll-target="#how-many-pcs-to-keep">How many PCs to keep?</a></li>
  </ul></li>
  <li><a href="#exercise-3-gene-expression-data" id="toc-exercise-3-gene-expression-data" class="nav-link" data-scroll-target="#exercise-3-gene-expression-data">Exercise 3: Gene expression data</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">R Lab (day 3): Data exploration, Principal Component Analysis</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p>Datasets</p>
<p>The following datasets are used in the exercises. Download datasets from GitHub repository, or from Canvas.</p>
<ul>
<li><code>Food.txt</code> (<a href="">link</a>)</li>
<li><code>nci60</code></li>
<li><code>ch12ex13.csv</code></li>
</ul>
<p>R script</p>
<p>Lab lecture</p>
<section id="exercise-1-food" class="level2">
<h2 class="anchored" data-anchor-id="exercise-1-food">Exercise 1: Food</h2>
<p>In the first exercise, we explore a low-dimensional dataset, <code>Food.txt</code>.</p>
<p>(This example is not a genomics example, however it is useful to illustrate important concepts in data exploration and dimensional reduction via PCA.)</p>
<p><code>Food.txt</code> contains data on food consumption of a variety of food categories in European counries. First load the dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>food <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">'data/Food.txt'</span>, <span class="at">header=</span>T)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co"># we change the name from pulses to a more common name, legume</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(food)[<span class="dv">7</span>] <span class="ot">&lt;-</span> <span class="st">'Legume'</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(food) <span class="co"># print first 6 lines </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               Meat Pigs Eggs Milk Fish Cereals Legume Fruit
Albania        10.1  1.4  0.5  8.9  0.2    42.3    5.5   1.7
Austria         8.9 14.0  4.3 19.9  2.1    28.0    1.3   4.3
Belg.Lux.      13.5  9.3  4.1 17.5  4.5    26.6    2.1   4.0
Bulgaria        7.8  6.0  1.6  8.3  1.2    56.7    3.7   4.2
Czechoslovakia  9.7 11.4  2.8 12.5  2.0    34.3    1.1   4.0
Denmark        10.6 10.8  3.7 25.0  9.9    21.9    0.7   2.4</code></pre>
</div>
</div>
<section id="explore-the-dataset" class="level3">
<h3 class="anchored" data-anchor-id="explore-the-dataset">Explore the dataset</h3>
<p>The first thing to do after loading the dataset is to get to know its content better.</p>
<p>Try to find out the following information about the data:</p>
<ul>
<li>the <strong>dimension</strong> (numbers of columns and rows)</li>
<li>the <strong>column names</strong>, and <strong>row names</strong></li>
</ul>
<p>We can also look at one food in particular, <code>Fish</code>: find the mean, maximum and minimum value for this variable. This is to investigate one column of the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(food)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 25  8</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(food)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Meat"    "Pigs"    "Eggs"    "Milk"    "Fish"    "Cereals" "Legume" 
[8] "Fruit"  </code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(food)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "Albania"        "Austria"        "Belg.Lux."      "Bulgaria"      
 [5] "Czechoslovakia" "Denmark"        "East.Germany"   "Finland"       
 [9] "France"         "Greece"         "Hungary"        "Ireland"       
[13] "Italy"          "Netherlands"    "Norway"         "Poland"        
[17] "Portugal"       "Romania"        "Spain"          "Sweden"        
[21] "Switzerland"    "United.Kingdom" "USSR"           "West.Germany"  
[25] "Yugoslavia"    </code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># explore one variable</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(food<span class="sc">$</span>Fish)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4.284</code></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">max</span>(food<span class="sc">$</span>Fish)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 14.2</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">min</span>(food<span class="sc">$</span>Fish)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.2</code></pre>
</div>
</div>
<p>Now we investigate the data by row. Each row is a country. We can extract the data (‘filter’) for Norway, for exmaple.</p>
<p>It is also easy to get data for more than one country, such as from Denmark and Sweden.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subsetting</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>food[<span class="fu">rownames</span>(food) <span class="sc">==</span> <span class="st">'Norway'</span>,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       Meat Pigs Eggs Milk Fish Cereals Legume Fruit
Norway  9.4  4.7  2.7 23.4  9.7      23    1.6   2.7</code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>food[<span class="fu">rownames</span>(food) <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">'Norway'</span>, <span class="st">'Denmark'</span>, <span class="st">'Sweden'</span>),]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Meat Pigs Eggs Milk Fish Cereals Legume Fruit
Denmark 10.6 10.8  3.7 25.0  9.9    21.9    0.7   2.4
Norway   9.4  4.7  2.7 23.4  9.7    23.0    1.6   2.7
Sweden   9.9  7.8  3.5 24.7  7.5    19.5    1.4   2.0</code></pre>
</div>
</div>
<p>Now we can make some simple visualizations. We can look at the distribution for each variable by making a histogram.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(food<span class="sc">$</span>Fish, <span class="at">main =</span> <span class="st">'Histogram of Fish'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab_day3_pca_files/figure-html/pca-food-histogram-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(food<span class="sc">$</span>Fish, <span class="at">breaks =</span> <span class="dv">20</span>, <span class="at">main =</span> <span class="st">'Histogram of Fish'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab_day3_pca_files/figure-html/pca-food-histogram-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(food<span class="sc">$</span>Meat, <span class="at">main =</span> <span class="st">'Histogram of Meat'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab_day3_pca_files/figure-html/pca-food-histogram-3.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>A box plot is another commonly used tool for presenting data distribution, for a few variables together. This way you can compare different foods.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># las rotates the axis text 90 degrees</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(food, <span class="at">las =</span> <span class="dv">2</span>,</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">'Food consumption across European countries'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab_day3_pca_files/figure-html/pca-food-boxplot-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We can look at pairs of food, to identify the whether there is some kind of relationship between them. For example, do countries that consume more fish also consume more meat? What about cereal and fruit?</p>
<p>A scatter plot with one food on its x-axis and another on the y-axis can be useful.</p>
<p>We can add the country names on top of the scatter plot to provide more information.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># two variables</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(food<span class="sc">$</span>Fish, food<span class="sc">$</span>Meat, <span class="at">pch=</span><span class="dv">20</span>, <span class="at">main =</span> <span class="st">'Fish vs Meat'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab_day3_pca_files/figure-html/pca-food-scatter-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># to add country: run this line as well</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="co"># text(food$Fish, food$Meat, labels = rownames(food))</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="co"># choose aother pair of food, add country label</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(food<span class="sc">$</span>Cereals, food<span class="sc">$</span>Fruit, <span class="at">main =</span> <span class="st">'Cereals vs Fruit'</span>)</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(food<span class="sc">$</span>Cereals, food<span class="sc">$</span>Fruit, <span class="at">labels =</span> <span class="fu">rownames</span>(food))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab_day3_pca_files/figure-html/pca-food-scatter-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>You can make pair-wise scatter plots for all the food pairs. However when you have many variables, these plots become less easy to read.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># pair-wise scatter</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(food, <span class="at">main =</span> <span class="st">'Pair-wise scatter plot'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab_day3_pca_files/figure-html/pca-food-scatter2-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="principal-component-analysis" class="level3">
<h3 class="anchored" data-anchor-id="principal-component-analysis">Principal component analysis</h3>
<p>We can use principal component analysis PCA to explore the dataset, and reveal more information beyond the original data points.</p>
<p>The command we use is <code>prcomp(data, scale)</code>.</p>
<ul>
<li><code>data</code> is the dataset to carry out PCA. Sometimes you need to do some processing such as centering and scaling.</li>
<li><code>scale</code> is an argument that asks the program to scale the data for us automatically. We specify it to be TRUE.</li>
</ul>
<p>After running the command, you can print out the results using <code>summary</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># need to scale the data</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>pc_food <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(food, <span class="at">scale=</span><span class="cn">TRUE</span>)</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="co"># pc_food</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(pc_food)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Importance of components:
                          PC1    PC2    PC3    PC4     PC5     PC6     PC7
Standard deviation     1.9251 1.2073 1.0595 0.9315 0.57322 0.52889 0.35617
Proportion of Variance 0.4632 0.1822 0.1403 0.1085 0.04107 0.03497 0.01586
Cumulative Proportion  0.4632 0.6454 0.7857 0.8942 0.93527 0.97024 0.98609
                           PC8
Standard deviation     0.33354
Proportion of Variance 0.01391
Cumulative Proportion  1.00000</code></pre>
</div>
</div>
<p>Examine the <strong>loading</strong> (rotation). This is a <span class="math inline">\(p \times n_{pc}\)</span> matrix where each row corresponds to one feature from the original data, and each column corresponds to one principal component PC.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>loading_food <span class="ot">&lt;-</span> pc_food<span class="sc">$</span>rotation</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="co"># print out the result, but only keep 2 digits</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(loading_food, <span class="at">digits =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          PC1   PC2   PC3   PC4   PC5   PC6   PC7   PC8
Meat    -0.33 -0.05 -0.20 -0.72 -0.48 -0.20 -0.12 -0.21
Pigs    -0.31  0.15  0.68  0.20 -0.06 -0.06 -0.37 -0.48
Eggs    -0.44 -0.07  0.23 -0.26  0.27  0.53  0.57 -0.06
Milk    -0.41  0.15 -0.36 -0.03  0.70 -0.38 -0.15 -0.13
Fish    -0.13 -0.67 -0.32  0.40 -0.13  0.02  0.11 -0.49
Cereals  0.45  0.29  0.05 -0.13  0.06 -0.32  0.52 -0.56
Legume   0.43 -0.17 -0.05 -0.36  0.34  0.46 -0.46 -0.32
Fruit    0.13 -0.62  0.45 -0.25  0.24 -0.46  0.06  0.23</code></pre>
</div>
</div>
<p><strong>Scores (projections)</strong> are produced by multiplying the original <span class="math inline">\(n \times p\)</span> data matrix and <span class="math inline">\(p \times n_{pc}\)</span> loading matrix. The result is a <span class="math inline">\(n \times n_{pc}\)</span> matrix where rows correspond to the rows in the original data (country names), and columns are the principal components.</p>
<p>You can visualize the points using a few selected PCs rather than the original data (next section).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>scores_food <span class="ot">&lt;-</span> pc_food<span class="sc">$</span>x</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(scores_food, <span class="at">digits =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 PC1   PC2   PC3   PC4   PC5   PC6   PC7   PC8
Albania         2.94  1.40 -1.60 -0.52 -1.02  0.22 -0.73  0.52
Austria        -1.61  0.65  1.60  0.30  0.41  0.11  0.19 -0.03
Belg.Lux.      -1.43 -0.13  0.18 -0.70 -0.46  0.27  0.28 -0.08
Bulgaria        2.70  1.02  0.32 -0.10 -0.50 -0.63  0.69 -0.26
Czechoslovakia -0.23  0.78  1.09  0.36 -0.80 -0.38  0.15  0.17
Denmark        -2.36 -0.34 -0.72  1.25 -0.11  0.07  0.08 -0.73
East.Germany   -1.03 -0.04  1.01  1.07 -0.82  0.55  0.38  0.20
Finland        -1.49  0.92 -2.26  0.91  0.87 -0.55  0.00  0.21
France         -1.46 -1.17  0.27 -1.73 -0.78 -1.10 -0.28 -0.40
Greece          1.97 -1.50 -0.65 -1.44  1.16  0.21 -0.10 -0.52
Hungary         1.50  0.86  1.84  0.26  0.44  0.83 -0.45 -0.36
Ireland        -2.46  0.84 -0.06 -0.92  0.28  0.30  0.20  0.03
Italy           1.22 -0.88  0.39 -0.70  0.43 -0.24  0.37  0.53
Netherlands    -1.73  0.69  0.94  0.40  0.47 -0.02 -0.62  0.03
Norway         -0.94 -0.66 -1.85  1.17 -0.02 -0.05  0.12  0.10
Poland          0.22 -0.20  1.22  0.45  0.75 -1.10  0.18  0.34
Portugal        1.83 -3.70  0.15  1.49 -0.64 -0.15 -0.35 -0.19
Romania         2.68  1.32 -0.10  0.24  0.04  0.19 -0.16 -0.30
Spain           1.71 -2.26  0.19 -0.27  0.44  0.77  0.23  0.63
Sweden         -1.81  0.04 -1.17  0.95  0.12  0.40 -0.05  0.03
Switzerland    -1.24  0.18  0.26 -0.79  0.21 -0.69 -0.61  0.22
United.Kingdom -1.78 -0.21 -0.96 -2.03 -0.38  0.78  0.31 -0.08
USSR            1.23  0.86 -0.90 -0.05 -0.12 -0.25  0.32 -0.06
West.Germany   -2.01  0.20  0.88  0.03 -0.22  0.41 -0.29  0.24
Yugoslavia      3.57  1.35 -0.05  0.37  0.26  0.07  0.13 -0.24</code></pre>
</div>
</div>
<p>Another important result from the <code>prcomp()</code> command is the explained variance (PVE, proportion of variance explained). All PCs explain 100% of the variance; while the first PC explains the largest proportion (46%), second PC explains the second largest proportion (18.2%) and so on.</p>
<p>You can find the variance (squared standard deviation, <code>sdev</code>) using the following commands. You can compare with the results from <code>summary()</code>, see if they correspond.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># variance explained by each PC</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>pc_food_var <span class="ot">&lt;-</span> pc_food<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="co"># proportion</span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>pc_food_pve <span class="ot">&lt;-</span> pc_food_var<span class="sc">/</span><span class="fu">sum</span>(pc_food_var)</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a><span class="co"># print out, keep 3 digits</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(pc_food_pve, <span class="at">digits =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.463 0.182 0.140 0.108 0.041 0.035 0.016 0.014</code></pre>
</div>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># cumulative of 1st, 2nd, ... 8th PC</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cumsum</span>(pc_food_pve)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.4632302 0.6454168 0.7857372 0.8941976 0.9352708 0.9702365 0.9860940
[8] 1.0000000</code></pre>
</div>
</div>
</section>
<section id="visualize-pca-results" class="level3">
<h3 class="anchored" data-anchor-id="visualize-pca-results">Visualize PCA results</h3>
<p>Biplot of the principal components displays two principal components with their labels. By default, it plots the first 2 PC; but you can also select PC1, PC3 with the <code>choices = c(1,3)</code> argument.</p>
<p>The country labels indicate the position where the original datapoints are projected onto PC1 and PC2. The closer points are clustered together, the more similar the points are. You can find the arrows pointing towards different directions, which tells additional information as in what way countries are similar.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># by default, pc1 pc2</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="co"># set x-axis limit</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="co"># same as biplot(pc_food, choices = c(1,2)) </span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a><span class="fu">biplot</span>(pc_food, <span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.5</span>, <span class="fl">0.5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab_day3_pca_files/figure-html/pca-food-biplot-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># choose pc1 and pc3</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="fu">biplot</span>(pc_food, <span class="at">choices =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab_day3_pca_files/figure-html/pca-food-biplot-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Now we can visualize the variance explained by the PCs, and compare with the original data features (columns). We use bar plots on the variance already computed from the previous section.</p>
<p>It can be seen that the first PC corresponds to the largest variance explained, the second PC corresponds to the second largest variance explained, etc.</p>
<p>On the other hand, the original data features have different variances, and do not seem to have obvious patterns.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="co"># variance of the PCs</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(pc_food_var, <span class="at">las=</span><span class="dv">2</span>,</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">main=</span><span class="st">'Principal components'</span>, <span class="at">ylab=</span><span class="st">'Variances'</span>,</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">names.arg =</span> <span class="fu">paste</span>(<span class="st">'PC '</span>,<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(pc_food_var),<span class="at">sep=</span><span class="st">''</span>))</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a><span class="co"># variance of the original data variables</span></span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a><span class="co"># apply command here computes sd by column</span></span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(<span class="fu">apply</span>(food, <span class="dv">2</span>, sd)<span class="sc">^</span><span class="dv">2</span>, <span class="at">las=</span><span class="dv">2</span>, <span class="at">main=</span><span class="st">'Original Variables'</span>, <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">150</span>), <span class="at">ylab=</span><span class="st">'Variances'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab_day3_pca_files/figure-html/pca-food-plot_var-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Lastly, you can visualize the <strong>cumulative</strong> variance explained well.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="co"># PVE</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">cumsum</span>(pc_food_pve), <span class="at">type=</span><span class="st">'b'</span>, <span class="at">axes=</span>F, <span class="at">xlab=</span><span class="st">'number of components'</span>,</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">'contribution to total variance'</span>, <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>))</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h=</span><span class="dv">1</span>, <span class="at">col=</span><span class="st">'blue'</span>)</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h=</span><span class="fl">0.8</span>, <span class="at">lty=</span><span class="dv">2</span>, <span class="at">col=</span><span class="st">'blue'</span>)</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a><span class="fu">box</span>()</span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">2</span>,<span class="at">at=</span><span class="dv">0</span><span class="sc">:</span><span class="dv">10</span><span class="sc">/</span><span class="dv">10</span>,<span class="at">labels=</span><span class="dv">0</span><span class="sc">:</span><span class="dv">10</span><span class="sc">/</span><span class="dv">10</span>)</span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">1</span>,<span class="at">at=</span><span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(food),<span class="at">labels=</span><span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(food),<span class="at">las=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab_day3_pca_files/figure-html/pca-food-plot_pve-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="exercise-2-nci60" class="level2">
<h2 class="anchored" data-anchor-id="exercise-2-nci60">Exercise 2: NCI60</h2>
<p>Now we move to a typical large-scale biological data set.</p>
<p>We have already seen the NCI60 cancer cell line microarray data set, consisting of 6830 gene expression measurements on 64 cancer cell lines.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ISLR)</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>nci.labs <span class="ot">&lt;-</span> NCI60<span class="sc">$</span>labs <span class="co"># Sample labels (tissue type)</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>nci.data <span class="ot">&lt;-</span> NCI60<span class="sc">$</span>data <span class="co"># Gene expression data set</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>nci.data</code> contains the gene expression level measured on 6830 genes. We can find out the dimension with <code>dim()</code> command.</p>
<p><code>nci.labs</code> contains the labels for the cancer types. Note that some cancer types have more than one cell line, we can get a quick overview using <code>table()</code>. This counts how many times each cancer type is present in the labels.</p>
<p>Since the focus of today’s topic is unsupervised learning, we do not need to pay too much attention to the labels. However we can use the labels in the end to check whether the PCA and clustering produce meaningful results.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(nci.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]   64 6830</code></pre>
</div>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(nci.labs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>nci.labs
     BREAST         CNS       COLON K562A-repro K562B-repro    LEUKEMIA 
          7           5           7           1           1           6 
MCF7A-repro MCF7D-repro    MELANOMA       NSCLC     OVARIAN    PROSTATE 
          1           1           8           9           6           2 
      RENAL     UNKNOWN 
          9           1 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># what if I would like to compute the mean of each gene within each tissue type?</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>tissue.means <span class="ot">&lt;-</span> <span class="fu">apply</span>(nci.data, <span class="dv">2</span>, <span class="cf">function</span>(x){<span class="fu">tapply</span>(x, nci.labs, mean)})</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(tissue.means)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]   14 6830</code></pre>
</div>
</div>
<p>You can and should carry out some exploratory analysis on the data, such as finding out the means and extreme values; this is left for home practice.</p>
<section id="carry-out-pca" class="level3">
<h3 class="anchored" data-anchor-id="carry-out-pca">Carry out PCA</h3>
<p>The dataset is a high-dimensional one, that is the number of columns (features) are much larger than the number of rows (measurements). It is inpractical to analyze and search for hidden structures by looking at each one of the feature, and PCA is a useful tool to move forward.</p>
<p>We use <code>prcomp()</code> to produce PCA results. Set the <code>scale = T</code> to make sure the data is standardized to have standard deviation equal to 1 among each column.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA analysis after scaling the variables to standard deviation one:</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>pr.out <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(nci.data, <span class="at">scale=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Print the summary output with <code>summary()</code>. Note that the result is quite long, so we didn’t print out everything; but you can execute the line and see what it does.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(pr.out)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we investigate the variance explained by the PCs. The y-axis in the plots below corresponds to the percentage (from 0 to 100%) explained.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>pr.var <span class="ot">&lt;-</span> pr.out<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>pve <span class="ot">&lt;-</span> pr.var<span class="sc">/</span><span class="fu">sum</span>(pr.var)</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>pve <span class="ot">&lt;-</span> <span class="dv">100</span><span class="sc">*</span>pve <span class="co"># display percentage</span></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pve,  <span class="at">type=</span><span class="st">"o"</span>, <span class="at">ylab=</span><span class="st">"PVE (%)"</span>, <span class="at">xlab=</span><span class="st">"Principal Component"</span>, <span class="at">col=</span><span class="st">"blue"</span>)</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">cumsum</span>(pve), <span class="at">type=</span><span class="st">"o"</span>, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">100</span>),<span class="at">ylab=</span><span class="st">"Cumulative PVE (%)"</span>, <span class="at">xlab=</span><span class="st">"Principal Component"</span>, <span class="at">col=</span><span class="st">"brown3"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab_day3_pca_files/figure-html/pca-nci-plotpc-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="how-many-pcs-to-keep" class="level3">
<h3 class="anchored" data-anchor-id="how-many-pcs-to-keep">How many PCs to keep?</h3>
<p>The maximum number of PC computed in principal component analysis is the <span class="math inline">\(min\{n, p\}\)</span>. This means, for a high dimensional problem like this one, we have at most 64 PCs.</p>
<p>For the purpose of dimensional reduction, we do not need to keep all the PCs. We can check how many PCs explain at least a certain amount of the variance. For example, we found out that the first 24 PC and first 32 PC explained 70% and 80% of the variance. We can mark it on the plots from before.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>mysel70 <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">cumsum</span>(pve) <span class="sc">&gt;</span> <span class="dv">70</span>)[<span class="dv">1</span>] <span class="co"># explains 70% </span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>mysel80 <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">cumsum</span>(pve) <span class="sc">&gt;</span> <span class="dv">80</span>)[<span class="dv">1</span>] <span class="co"># explains 80% </span></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(mysel70, mysel80)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 24 32</code></pre>
</div>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)) <span class="co"># plot contains two smaller plots next to each other</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pve,  <span class="at">type=</span><span class="st">"o"</span>, <span class="at">ylab=</span><span class="st">"PVE"</span>, <span class="at">xlab=</span><span class="st">"Principal Component"</span>, <span class="at">col=</span><span class="st">"blue"</span>)</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> mysel80)</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> mysel70, <span class="at">col=</span><span class="dv">3</span>)</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">cumsum</span>(pve), <span class="at">type=</span><span class="st">"o"</span>, <span class="at">ylab=</span><span class="st">"Cumulative PVE"</span>, <span class="at">xlab=</span><span class="st">"Principal Component"</span>, <span class="at">col=</span><span class="st">"brown3"</span>)</span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> mysel80)</span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> <span class="dv">80</span>)</span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> mysel70, <span class="at">col=</span><span class="dv">3</span>)</span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> <span class="dv">70</span>, <span class="at">col=</span><span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab_day3_pca_files/figure-html/pca-nci-keeppc-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>If we decide to only keep the principal components that explains 70% of the variance, we end up with 24 components, which we can further analyse to better understand the relationships between the variables. For simplicity we only look at the first few components.</p>
<p>We plot the first few principal component score vectors, to visualize the results. The observations (cell lines) corresponding to a given cancer type will be plotted in the same colour.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>Cols<span class="ot">=</span><span class="cf">function</span>(vec){</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>  cols<span class="ot">=</span><span class="fu">rainbow</span>(<span class="fu">length</span>(<span class="fu">unique</span>(vec)))</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(cols[<span class="fu">as.numeric</span>(<span class="fu">as.factor</span>(vec))])</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the first vs second and first vs third principal component score vectors,</span></span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a><span class="co"># with colors associated to labels (using the Cols() helper function)</span></span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pr.out<span class="sc">$</span>x[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="at">col=</span><span class="fu">Cols</span>(nci.labs), <span class="at">pch=</span><span class="dv">19</span>,<span class="at">xlab=</span><span class="st">"PC 1"</span>,<span class="at">ylab=</span><span class="st">" PC 2"</span>)</span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pr.out<span class="sc">$</span>x[,<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>)], <span class="at">col=</span><span class="fu">Cols</span>(nci.labs), <span class="at">pch=</span><span class="dv">19</span>,<span class="at">xlab=</span><span class="st">"PC 1"</span>,<span class="at">ylab=</span><span class="st">" PC 3"</span>)</span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">'topleft'</span>, <span class="at">col=</span><span class="fu">rainbow</span>(<span class="fu">length</span>(<span class="fu">unique</span>(nci.labs))), <span class="at">legend=</span><span class="fu">unique</span>(nci.labs), <span class="at">bty=</span><span class="st">'n'</span>, <span class="at">lwd=</span><span class="dv">2</span>, <span class="at">cex=</span>.<span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab_day3_pca_files/figure-html/pca-nci-plot2-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Why not use biplot?
</div>
</div>
<div class="callout-body-container callout-body">
<p>You might have realized that the plots above are essentially what <code>biplot</code> did in the first exercise: plotting PC1, PC2 againt each other. The reason we do not use it here is simple: by default <code>biplot</code> displays also the loading vectors (rotation). Given that we have 64 PCs, the final figure would be unreadable.</p>
<p>You can try to remove the loading vectors from biplot as an exercise.</p>
</div>
</div>
</section>
</section>
<section id="exercise-3-gene-expression-data" class="level2">
<h2 class="anchored" data-anchor-id="exercise-3-gene-expression-data">Exercise 3: Gene expression data</h2>
<p>(CH12Ex13 from statistical learning)</p>
<p>Consider again the gene expression data set “Ch12Ex13.csv” (which can be also found on the <a href="www.StatLearning.com">book website</a>) that consists of 40 tissue samples with measurements on 1,000 genes.</p>
<p>The first 20 samples are from healthy patients, while the second 20 are from a diseased group.</p>
<p>Load in the data using read.csv(). You will need to select header=F.</p>
<p>Alternatively: load in the data using “Import dataset” in the upper right window, and click “no” on the “Heading” option.</p>
<p>Perform a PCA of these data and visualize the results.</p>
<p>Note: remember to check if the variables (genes) are on the columns in the dataset before running the PCA. If they are not: use t() to transform the dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set the path to your own!</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>exp.data <span class="ot">&lt;-</span>  <span class="fu">read.csv</span>(<span class="st">"data/Ch12Ex13.csv"</span>,<span class="at">header=</span><span class="cn">FALSE</span>)</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a><span class="co"># I want each row to represent a sample, and each column a gene</span></span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>exp.data <span class="ot">&lt;-</span> <span class="fu">t</span>(exp.data)</span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(exp.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]   40 1000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># should have n=40 samples/rows, and 1000 columns --&gt; OK!</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>groups <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">1</span>,<span class="dv">20</span>), <span class="fu">rep</span>(<span class="dv">2</span>,<span class="dv">20</span>)) <span class="co"># group variable</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Carry out PCA</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>pr.exp <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(exp.data, <span class="at">scale=</span><span class="cn">TRUE</span>)</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot proportion of variance explained</span></span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>pr.var <span class="ot">&lt;-</span> pr.exp<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>pve <span class="ot">&lt;-</span> pr.var<span class="sc">/</span><span class="fu">sum</span>(pr.var)</span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a>pve <span class="ot">&lt;-</span> <span class="dv">100</span><span class="sc">*</span>pve</span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pve, <span class="at">type=</span><span class="st">"o"</span>, <span class="at">ylab=</span><span class="st">"PVE"</span>, <span class="at">xlab=</span><span class="st">"Principal Component"</span>, <span class="at">col=</span><span class="st">"blue"</span>)</span>
<span id="cb57-10"><a href="#cb57-10" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">cumsum</span>(pve), <span class="at">type=</span><span class="st">"o"</span>, <span class="at">ylab=</span><span class="st">"Cumulative PVE"</span>, <span class="at">xlab=</span><span class="st">"Principal Component"</span>, <span class="at">col=</span><span class="st">"red"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab_day3_pca_files/figure-html/pca-ge-pca-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Looks like most of the principal components are needed to explain the data well. Maybe we can decide to keep 25-30 components?</p>
<p>Can also plot some of the first principal components</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Remember the use the helper-function to get colours</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>Cols<span class="ot">=</span><span class="cf">function</span>(vec){</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>  cols<span class="ot">=</span><span class="fu">rainbow</span>(<span class="fu">length</span>(<span class="fu">unique</span>(vec)))</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(cols[<span class="fu">as.numeric</span>(<span class="fu">as.factor</span>(vec))])</span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)) <span class="co"># plot-window has two small plots next to each other</span></span>
<span id="cb58-8"><a href="#cb58-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pr.exp<span class="sc">$</span>x[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="at">col=</span><span class="fu">Cols</span>(groups), <span class="at">pch=</span><span class="dv">19</span>, <span class="at">xlab=</span><span class="st">"PC 1"</span>, <span class="at">ylab=</span><span class="st">" PC 2"</span>)</span>
<span id="cb58-9"><a href="#cb58-9" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pr.exp<span class="sc">$</span>x[,<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>)], <span class="at">col=</span><span class="fu">Cols</span>(groups), <span class="at">pch=</span><span class="dv">19</span>,<span class="at">xlab=</span><span class="st">"PC 1"</span>,<span class="at">ylab=</span><span class="st">" PC 3"</span>)</span>
<span id="cb58-10"><a href="#cb58-10" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">'topleft'</span>, <span class="at">col=</span><span class="fu">rainbow</span>(<span class="fu">length</span>(<span class="fu">unique</span>(groups))), <span class="at">legend=</span><span class="fu">paste</span>(<span class="st">'group '</span>,<span class="fu">unique</span>(groups),<span class="at">sep=</span><span class="st">''</span>), <span class="at">bty=</span><span class="st">'n'</span>, <span class="at">lwd=</span><span class="dv">2</span>, <span class="at">cex=</span>.<span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab_day3_pca_files/figure-html/pca-ge-plot-1.png" class="img-fluid" width="672"></p>
</div>
</div>


<!-- -->

</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb59" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "R Lab (day 3): Data exploration, Principal Component Analysis"</span></span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span><span class="co"> </span></span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a><span class="co">    code-fold: false</span></span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools: true</span></span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb59-8"><a href="#cb59-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-9"><a href="#cb59-9" aria-hidden="true" tabindex="-1"></a>Datasets</span>
<span id="cb59-10"><a href="#cb59-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-11"><a href="#cb59-11" aria-hidden="true" tabindex="-1"></a>The following datasets are used in the exercises. Download datasets from GitHub repository, or from Canvas.</span>
<span id="cb59-12"><a href="#cb59-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-13"><a href="#cb59-13" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`Food.txt`</span> (<span class="co">[</span><span class="ot">link</span><span class="co">]()</span>)</span>
<span id="cb59-14"><a href="#cb59-14" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`nci60`</span></span>
<span id="cb59-15"><a href="#cb59-15" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`ch12ex13.csv`</span></span>
<span id="cb59-16"><a href="#cb59-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-17"><a href="#cb59-17" aria-hidden="true" tabindex="-1"></a>R script </span>
<span id="cb59-18"><a href="#cb59-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-19"><a href="#cb59-19" aria-hidden="true" tabindex="-1"></a>Lab lecture </span>
<span id="cb59-20"><a href="#cb59-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-21"><a href="#cb59-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-22"><a href="#cb59-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-23"><a href="#cb59-23" aria-hidden="true" tabindex="-1"></a><span class="fu">## Exercise 1: Food</span></span>
<span id="cb59-24"><a href="#cb59-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-25"><a href="#cb59-25" aria-hidden="true" tabindex="-1"></a>In the first exercise, we explore a low-dimensional dataset, <span class="in">`Food.txt`</span>. </span>
<span id="cb59-26"><a href="#cb59-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-27"><a href="#cb59-27" aria-hidden="true" tabindex="-1"></a>(This example is not a genomics example, however it is useful to illustrate important concepts in data exploration and dimensional reduction via PCA.)  </span>
<span id="cb59-28"><a href="#cb59-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-29"><a href="#cb59-29" aria-hidden="true" tabindex="-1"></a><span class="in">`Food.txt`</span> contains data on food consumption of a variety of food categories in European counries. First load the dataset. </span>
<span id="cb59-30"><a href="#cb59-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-31"><a href="#cb59-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-34"><a href="#cb59-34" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-35"><a href="#cb59-35" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-food-loaddata</span></span>
<span id="cb59-36"><a href="#cb59-36" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-37"><a href="#cb59-37" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-38"><a href="#cb59-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-39"><a href="#cb59-39" aria-hidden="true" tabindex="-1"></a>food <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">'data/Food.txt'</span>, <span class="at">header=</span>T)</span>
<span id="cb59-40"><a href="#cb59-40" aria-hidden="true" tabindex="-1"></a><span class="co"># we change the name from pulses to a more common name, legume</span></span>
<span id="cb59-41"><a href="#cb59-41" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(food)[<span class="dv">7</span>] <span class="ot">&lt;-</span> <span class="st">'Legume'</span></span>
<span id="cb59-42"><a href="#cb59-42" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(food) <span class="co"># print first 6 lines </span></span>
<span id="cb59-43"><a href="#cb59-43" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-44"><a href="#cb59-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-45"><a href="#cb59-45" aria-hidden="true" tabindex="-1"></a><span class="fu">### Explore the dataset</span></span>
<span id="cb59-46"><a href="#cb59-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-47"><a href="#cb59-47" aria-hidden="true" tabindex="-1"></a>The first thing to do after loading the dataset is to get to know its content better. </span>
<span id="cb59-48"><a href="#cb59-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-49"><a href="#cb59-49" aria-hidden="true" tabindex="-1"></a>Try to find out the following information about the data:</span>
<span id="cb59-50"><a href="#cb59-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-51"><a href="#cb59-51" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>the **dimension** (numbers of columns and rows)</span>
<span id="cb59-52"><a href="#cb59-52" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>the **column names**, and **row names**</span>
<span id="cb59-53"><a href="#cb59-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-54"><a href="#cb59-54" aria-hidden="true" tabindex="-1"></a>We can also look at one food in particular, <span class="in">`Fish`</span>: find the mean, maximum and minimum value for this variable. This is to investigate one column of the data.</span>
<span id="cb59-55"><a href="#cb59-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-56"><a href="#cb59-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-59"><a href="#cb59-59" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-60"><a href="#cb59-60" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-food-summary</span></span>
<span id="cb59-61"><a href="#cb59-61" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-62"><a href="#cb59-62" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-63"><a href="#cb59-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-64"><a href="#cb59-64" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(food)</span>
<span id="cb59-65"><a href="#cb59-65" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(food)</span>
<span id="cb59-66"><a href="#cb59-66" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(food)</span>
<span id="cb59-67"><a href="#cb59-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-68"><a href="#cb59-68" aria-hidden="true" tabindex="-1"></a><span class="co"># explore one variable</span></span>
<span id="cb59-69"><a href="#cb59-69" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(food<span class="sc">$</span>Fish)</span>
<span id="cb59-70"><a href="#cb59-70" aria-hidden="true" tabindex="-1"></a><span class="fu">max</span>(food<span class="sc">$</span>Fish)</span>
<span id="cb59-71"><a href="#cb59-71" aria-hidden="true" tabindex="-1"></a><span class="fu">min</span>(food<span class="sc">$</span>Fish)</span>
<span id="cb59-72"><a href="#cb59-72" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-73"><a href="#cb59-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-74"><a href="#cb59-74" aria-hidden="true" tabindex="-1"></a>Now we investigate the data by row. Each row is a country. We can extract the data ('filter') for Norway, for exmaple.</span>
<span id="cb59-75"><a href="#cb59-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-76"><a href="#cb59-76" aria-hidden="true" tabindex="-1"></a>It is also easy to get data for more than one country, such as from Denmark and Sweden. </span>
<span id="cb59-77"><a href="#cb59-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-80"><a href="#cb59-80" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-81"><a href="#cb59-81" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-food-country</span></span>
<span id="cb59-82"><a href="#cb59-82" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-83"><a href="#cb59-83" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-84"><a href="#cb59-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-85"><a href="#cb59-85" aria-hidden="true" tabindex="-1"></a><span class="co"># subsetting</span></span>
<span id="cb59-86"><a href="#cb59-86" aria-hidden="true" tabindex="-1"></a>food[<span class="fu">rownames</span>(food) <span class="sc">==</span> <span class="st">'Norway'</span>,]</span>
<span id="cb59-87"><a href="#cb59-87" aria-hidden="true" tabindex="-1"></a>food[<span class="fu">rownames</span>(food) <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">'Norway'</span>, <span class="st">'Denmark'</span>, <span class="st">'Sweden'</span>),]</span>
<span id="cb59-88"><a href="#cb59-88" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-89"><a href="#cb59-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-90"><a href="#cb59-90" aria-hidden="true" tabindex="-1"></a>Now we can make some simple visualizations. We can look at the distribution for each variable by making a histogram.</span>
<span id="cb59-91"><a href="#cb59-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-94"><a href="#cb59-94" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-95"><a href="#cb59-95" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-food-histogram</span></span>
<span id="cb59-96"><a href="#cb59-96" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-97"><a href="#cb59-97" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-98"><a href="#cb59-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-99"><a href="#cb59-99" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(food<span class="sc">$</span>Fish, <span class="at">main =</span> <span class="st">'Histogram of Fish'</span>)</span>
<span id="cb59-100"><a href="#cb59-100" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(food<span class="sc">$</span>Fish, <span class="at">breaks =</span> <span class="dv">20</span>, <span class="at">main =</span> <span class="st">'Histogram of Fish'</span>)</span>
<span id="cb59-101"><a href="#cb59-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-102"><a href="#cb59-102" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(food<span class="sc">$</span>Meat, <span class="at">main =</span> <span class="st">'Histogram of Meat'</span>)</span>
<span id="cb59-103"><a href="#cb59-103" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-104"><a href="#cb59-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-105"><a href="#cb59-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-106"><a href="#cb59-106" aria-hidden="true" tabindex="-1"></a>A box plot is another commonly used tool for presenting data distribution, for a few variables together. This way you can compare different foods.</span>
<span id="cb59-107"><a href="#cb59-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-110"><a href="#cb59-110" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-111"><a href="#cb59-111" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-food-boxplot</span></span>
<span id="cb59-112"><a href="#cb59-112" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-113"><a href="#cb59-113" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-114"><a href="#cb59-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-115"><a href="#cb59-115" aria-hidden="true" tabindex="-1"></a><span class="co"># las rotates the axis text 90 degrees</span></span>
<span id="cb59-116"><a href="#cb59-116" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(food, <span class="at">las =</span> <span class="dv">2</span>,</span>
<span id="cb59-117"><a href="#cb59-117" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">'Food consumption across European countries'</span>)</span>
<span id="cb59-118"><a href="#cb59-118" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-119"><a href="#cb59-119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-120"><a href="#cb59-120" aria-hidden="true" tabindex="-1"></a>We can look at pairs of food, to identify the whether there is some kind of relationship between them. For example, do countries that consume more fish also consume more meat? What about cereal and fruit?</span>
<span id="cb59-121"><a href="#cb59-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-122"><a href="#cb59-122" aria-hidden="true" tabindex="-1"></a>A scatter plot with one food on its x-axis and another on the y-axis can be useful.</span>
<span id="cb59-123"><a href="#cb59-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-124"><a href="#cb59-124" aria-hidden="true" tabindex="-1"></a>We can add the country names on top of the scatter plot to provide more information.</span>
<span id="cb59-125"><a href="#cb59-125" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-128"><a href="#cb59-128" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-129"><a href="#cb59-129" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-food-scatter</span></span>
<span id="cb59-130"><a href="#cb59-130" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-131"><a href="#cb59-131" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-132"><a href="#cb59-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-133"><a href="#cb59-133" aria-hidden="true" tabindex="-1"></a><span class="co"># two variables</span></span>
<span id="cb59-134"><a href="#cb59-134" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(food<span class="sc">$</span>Fish, food<span class="sc">$</span>Meat, <span class="at">pch=</span><span class="dv">20</span>, <span class="at">main =</span> <span class="st">'Fish vs Meat'</span>)</span>
<span id="cb59-135"><a href="#cb59-135" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-136"><a href="#cb59-136" aria-hidden="true" tabindex="-1"></a><span class="co"># to add country: run this line as well</span></span>
<span id="cb59-137"><a href="#cb59-137" aria-hidden="true" tabindex="-1"></a><span class="co"># text(food$Fish, food$Meat, labels = rownames(food))</span></span>
<span id="cb59-138"><a href="#cb59-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-139"><a href="#cb59-139" aria-hidden="true" tabindex="-1"></a><span class="co"># choose aother pair of food, add country label</span></span>
<span id="cb59-140"><a href="#cb59-140" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(food<span class="sc">$</span>Cereals, food<span class="sc">$</span>Fruit, <span class="at">main =</span> <span class="st">'Cereals vs Fruit'</span>)</span>
<span id="cb59-141"><a href="#cb59-141" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(food<span class="sc">$</span>Cereals, food<span class="sc">$</span>Fruit, <span class="at">labels =</span> <span class="fu">rownames</span>(food))</span>
<span id="cb59-142"><a href="#cb59-142" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-143"><a href="#cb59-143" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-144"><a href="#cb59-144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-145"><a href="#cb59-145" aria-hidden="true" tabindex="-1"></a>You can make pair-wise scatter plots for all the food pairs. However when you have many variables, these plots become less easy to read. </span>
<span id="cb59-146"><a href="#cb59-146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-149"><a href="#cb59-149" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-150"><a href="#cb59-150" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-food-scatter2</span></span>
<span id="cb59-151"><a href="#cb59-151" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-152"><a href="#cb59-152" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-153"><a href="#cb59-153" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-154"><a href="#cb59-154" aria-hidden="true" tabindex="-1"></a><span class="co"># pair-wise scatter</span></span>
<span id="cb59-155"><a href="#cb59-155" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(food, <span class="at">main =</span> <span class="st">'Pair-wise scatter plot'</span>)</span>
<span id="cb59-156"><a href="#cb59-156" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-157"><a href="#cb59-157" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-158"><a href="#cb59-158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-159"><a href="#cb59-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-160"><a href="#cb59-160" aria-hidden="true" tabindex="-1"></a><span class="fu">### Principal component analysis </span></span>
<span id="cb59-161"><a href="#cb59-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-162"><a href="#cb59-162" aria-hidden="true" tabindex="-1"></a>We can use principal component analysis PCA to explore the dataset, and reveal more information beyond the original data points.</span>
<span id="cb59-163"><a href="#cb59-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-164"><a href="#cb59-164" aria-hidden="true" tabindex="-1"></a>The command we use is <span class="in">`prcomp(data, scale)`</span>.</span>
<span id="cb59-165"><a href="#cb59-165" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-166"><a href="#cb59-166" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`data`</span> is the dataset to carry out PCA. Sometimes you need to do some processing such as centering and scaling.</span>
<span id="cb59-167"><a href="#cb59-167" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`scale`</span> is an argument that asks the program to scale the data for us automatically. We specify it to be TRUE.</span>
<span id="cb59-168"><a href="#cb59-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-169"><a href="#cb59-169" aria-hidden="true" tabindex="-1"></a>After running the command, you can print out the results using <span class="in">`summary`</span>.</span>
<span id="cb59-170"><a href="#cb59-170" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-173"><a href="#cb59-173" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-174"><a href="#cb59-174" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-food-pca</span></span>
<span id="cb59-175"><a href="#cb59-175" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-176"><a href="#cb59-176" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-177"><a href="#cb59-177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-178"><a href="#cb59-178" aria-hidden="true" tabindex="-1"></a><span class="co"># need to scale the data</span></span>
<span id="cb59-179"><a href="#cb59-179" aria-hidden="true" tabindex="-1"></a>pc_food <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(food, <span class="at">scale=</span><span class="cn">TRUE</span>)</span>
<span id="cb59-180"><a href="#cb59-180" aria-hidden="true" tabindex="-1"></a><span class="co"># pc_food</span></span>
<span id="cb59-181"><a href="#cb59-181" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(pc_food)</span>
<span id="cb59-182"><a href="#cb59-182" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-183"><a href="#cb59-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-184"><a href="#cb59-184" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-185"><a href="#cb59-185" aria-hidden="true" tabindex="-1"></a>Examine the **loading** (rotation). This is a $p \times n_{pc}$ matrix where each row corresponds to one feature from the original data, and each column corresponds to one principal component PC. </span>
<span id="cb59-186"><a href="#cb59-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-189"><a href="#cb59-189" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-190"><a href="#cb59-190" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-food-loading</span></span>
<span id="cb59-191"><a href="#cb59-191" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-192"><a href="#cb59-192" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-193"><a href="#cb59-193" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-194"><a href="#cb59-194" aria-hidden="true" tabindex="-1"></a>loading_food <span class="ot">&lt;-</span> pc_food<span class="sc">$</span>rotation</span>
<span id="cb59-195"><a href="#cb59-195" aria-hidden="true" tabindex="-1"></a><span class="co"># print out the result, but only keep 2 digits</span></span>
<span id="cb59-196"><a href="#cb59-196" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(loading_food, <span class="at">digits =</span> <span class="dv">2</span>)</span>
<span id="cb59-197"><a href="#cb59-197" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-198"><a href="#cb59-198" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-199"><a href="#cb59-199" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-200"><a href="#cb59-200" aria-hidden="true" tabindex="-1"></a>**Scores (projections)** are produced by multiplying the original $n \times p$ data matrix and $p \times n_{pc}$ loading matrix. The result is a $n \times n_{pc}$ matrix where rows correspond to the rows in the original data (country names), and columns are the principal components. </span>
<span id="cb59-201"><a href="#cb59-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-202"><a href="#cb59-202" aria-hidden="true" tabindex="-1"></a>You can visualize the points using a few selected PCs rather than the original data (next section). </span>
<span id="cb59-203"><a href="#cb59-203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-206"><a href="#cb59-206" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-207"><a href="#cb59-207" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-food-scores</span></span>
<span id="cb59-208"><a href="#cb59-208" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-209"><a href="#cb59-209" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-210"><a href="#cb59-210" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-211"><a href="#cb59-211" aria-hidden="true" tabindex="-1"></a>scores_food <span class="ot">&lt;-</span> pc_food<span class="sc">$</span>x</span>
<span id="cb59-212"><a href="#cb59-212" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(scores_food, <span class="at">digits =</span> <span class="dv">2</span>)</span>
<span id="cb59-213"><a href="#cb59-213" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-214"><a href="#cb59-214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-215"><a href="#cb59-215" aria-hidden="true" tabindex="-1"></a>Another important result from the <span class="in">`prcomp()`</span> command is the explained variance (PVE, proportion of variance explained). All PCs explain 100% of the variance; while the first PC explains the largest proportion (46%), second PC explains the second largest proportion (18.2%) and so on. </span>
<span id="cb59-216"><a href="#cb59-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-217"><a href="#cb59-217" aria-hidden="true" tabindex="-1"></a>You can find the variance (squared standard deviation, <span class="in">`sdev`</span>) using the following commands. You can compare with the results from <span class="in">`summary()`</span>, see if they correspond.</span>
<span id="cb59-218"><a href="#cb59-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-221"><a href="#cb59-221" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-222"><a href="#cb59-222" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-food-variance</span></span>
<span id="cb59-223"><a href="#cb59-223" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-224"><a href="#cb59-224" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-225"><a href="#cb59-225" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-226"><a href="#cb59-226" aria-hidden="true" tabindex="-1"></a><span class="co"># variance explained by each PC</span></span>
<span id="cb59-227"><a href="#cb59-227" aria-hidden="true" tabindex="-1"></a>pc_food_var <span class="ot">&lt;-</span> pc_food<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb59-228"><a href="#cb59-228" aria-hidden="true" tabindex="-1"></a><span class="co"># proportion</span></span>
<span id="cb59-229"><a href="#cb59-229" aria-hidden="true" tabindex="-1"></a>pc_food_pve <span class="ot">&lt;-</span> pc_food_var<span class="sc">/</span><span class="fu">sum</span>(pc_food_var)</span>
<span id="cb59-230"><a href="#cb59-230" aria-hidden="true" tabindex="-1"></a><span class="co"># print out, keep 3 digits</span></span>
<span id="cb59-231"><a href="#cb59-231" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(pc_food_pve, <span class="at">digits =</span> <span class="dv">3</span>)</span>
<span id="cb59-232"><a href="#cb59-232" aria-hidden="true" tabindex="-1"></a><span class="co"># cumulative of 1st, 2nd, ... 8th PC</span></span>
<span id="cb59-233"><a href="#cb59-233" aria-hidden="true" tabindex="-1"></a><span class="fu">cumsum</span>(pc_food_pve)</span>
<span id="cb59-234"><a href="#cb59-234" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-235"><a href="#cb59-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-236"><a href="#cb59-236" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-237"><a href="#cb59-237" aria-hidden="true" tabindex="-1"></a><span class="fu">### Visualize PCA results</span></span>
<span id="cb59-238"><a href="#cb59-238" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-239"><a href="#cb59-239" aria-hidden="true" tabindex="-1"></a>Biplot of the principal components displays two principal components with their labels. By default, it plots the first 2 PC; but you can also select PC1, PC3 with the <span class="in">`choices = c(1,3)`</span> argument.</span>
<span id="cb59-240"><a href="#cb59-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-241"><a href="#cb59-241" aria-hidden="true" tabindex="-1"></a>The country labels indicate the position where the original datapoints are projected onto PC1 and PC2. The closer points are clustered together, the more similar the points are. You can find the arrows pointing towards different directions, which tells additional information as in what way countries are similar.</span>
<span id="cb59-242"><a href="#cb59-242" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-243"><a href="#cb59-243" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-246"><a href="#cb59-246" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-247"><a href="#cb59-247" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-food-biplot</span></span>
<span id="cb59-248"><a href="#cb59-248" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-249"><a href="#cb59-249" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-250"><a href="#cb59-250" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-251"><a href="#cb59-251" aria-hidden="true" tabindex="-1"></a><span class="co"># by default, pc1 pc2</span></span>
<span id="cb59-252"><a href="#cb59-252" aria-hidden="true" tabindex="-1"></a><span class="co"># set x-axis limit</span></span>
<span id="cb59-253"><a href="#cb59-253" aria-hidden="true" tabindex="-1"></a><span class="co"># same as biplot(pc_food, choices = c(1,2)) </span></span>
<span id="cb59-254"><a href="#cb59-254" aria-hidden="true" tabindex="-1"></a><span class="fu">biplot</span>(pc_food, <span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.5</span>, <span class="fl">0.5</span>))</span>
<span id="cb59-255"><a href="#cb59-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-256"><a href="#cb59-256" aria-hidden="true" tabindex="-1"></a><span class="co"># choose pc1 and pc3</span></span>
<span id="cb59-257"><a href="#cb59-257" aria-hidden="true" tabindex="-1"></a><span class="fu">biplot</span>(pc_food, <span class="at">choices =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>)) </span>
<span id="cb59-258"><a href="#cb59-258" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-259"><a href="#cb59-259" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-260"><a href="#cb59-260" aria-hidden="true" tabindex="-1"></a>Now we can visualize the variance explained by the PCs, and compare with the original data features (columns). We use bar plots on the variance already computed from the previous section. </span>
<span id="cb59-261"><a href="#cb59-261" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-262"><a href="#cb59-262" aria-hidden="true" tabindex="-1"></a>It can be seen that the first PC corresponds to the largest variance explained, the second PC corresponds to the second largest variance explained, etc.</span>
<span id="cb59-263"><a href="#cb59-263" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-264"><a href="#cb59-264" aria-hidden="true" tabindex="-1"></a>On the other hand, the original data features have different variances, and do not seem to have obvious patterns.</span>
<span id="cb59-265"><a href="#cb59-265" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-268"><a href="#cb59-268" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-269"><a href="#cb59-269" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-food-plot_var</span></span>
<span id="cb59-270"><a href="#cb59-270" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-271"><a href="#cb59-271" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-272"><a href="#cb59-272" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-273"><a href="#cb59-273" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb59-274"><a href="#cb59-274" aria-hidden="true" tabindex="-1"></a><span class="co"># variance of the PCs</span></span>
<span id="cb59-275"><a href="#cb59-275" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(pc_food_var, <span class="at">las=</span><span class="dv">2</span>,</span>
<span id="cb59-276"><a href="#cb59-276" aria-hidden="true" tabindex="-1"></a>        <span class="at">main=</span><span class="st">'Principal components'</span>, <span class="at">ylab=</span><span class="st">'Variances'</span>,</span>
<span id="cb59-277"><a href="#cb59-277" aria-hidden="true" tabindex="-1"></a>        <span class="at">names.arg =</span> <span class="fu">paste</span>(<span class="st">'PC '</span>,<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(pc_food_var),<span class="at">sep=</span><span class="st">''</span>))</span>
<span id="cb59-278"><a href="#cb59-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-279"><a href="#cb59-279" aria-hidden="true" tabindex="-1"></a><span class="co"># variance of the original data variables</span></span>
<span id="cb59-280"><a href="#cb59-280" aria-hidden="true" tabindex="-1"></a><span class="co"># apply command here computes sd by column</span></span>
<span id="cb59-281"><a href="#cb59-281" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(<span class="fu">apply</span>(food, <span class="dv">2</span>, sd)<span class="sc">^</span><span class="dv">2</span>, <span class="at">las=</span><span class="dv">2</span>, <span class="at">main=</span><span class="st">'Original Variables'</span>, <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">150</span>), <span class="at">ylab=</span><span class="st">'Variances'</span>)</span>
<span id="cb59-282"><a href="#cb59-282" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-283"><a href="#cb59-283" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-284"><a href="#cb59-284" aria-hidden="true" tabindex="-1"></a>Lastly, you can visualize the **cumulative** variance explained well.</span>
<span id="cb59-285"><a href="#cb59-285" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-288"><a href="#cb59-288" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-289"><a href="#cb59-289" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-food-plot_pve</span></span>
<span id="cb59-290"><a href="#cb59-290" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-291"><a href="#cb59-291" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-292"><a href="#cb59-292" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-293"><a href="#cb59-293" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="cb59-294"><a href="#cb59-294" aria-hidden="true" tabindex="-1"></a><span class="co"># PVE</span></span>
<span id="cb59-295"><a href="#cb59-295" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">cumsum</span>(pc_food_pve), <span class="at">type=</span><span class="st">'b'</span>, <span class="at">axes=</span>F, <span class="at">xlab=</span><span class="st">'number of components'</span>,</span>
<span id="cb59-296"><a href="#cb59-296" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">'contribution to total variance'</span>, <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>))</span>
<span id="cb59-297"><a href="#cb59-297" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h=</span><span class="dv">1</span>, <span class="at">col=</span><span class="st">'blue'</span>)</span>
<span id="cb59-298"><a href="#cb59-298" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h=</span><span class="fl">0.8</span>, <span class="at">lty=</span><span class="dv">2</span>, <span class="at">col=</span><span class="st">'blue'</span>)</span>
<span id="cb59-299"><a href="#cb59-299" aria-hidden="true" tabindex="-1"></a><span class="fu">box</span>()</span>
<span id="cb59-300"><a href="#cb59-300" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">2</span>,<span class="at">at=</span><span class="dv">0</span><span class="sc">:</span><span class="dv">10</span><span class="sc">/</span><span class="dv">10</span>,<span class="at">labels=</span><span class="dv">0</span><span class="sc">:</span><span class="dv">10</span><span class="sc">/</span><span class="dv">10</span>)</span>
<span id="cb59-301"><a href="#cb59-301" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">1</span>,<span class="at">at=</span><span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(food),<span class="at">labels=</span><span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(food),<span class="at">las=</span><span class="dv">2</span>)</span>
<span id="cb59-302"><a href="#cb59-302" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-303"><a href="#cb59-303" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-304"><a href="#cb59-304" aria-hidden="true" tabindex="-1"></a><span class="fu">## Exercise 2: NCI60</span></span>
<span id="cb59-305"><a href="#cb59-305" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-306"><a href="#cb59-306" aria-hidden="true" tabindex="-1"></a>Now we move to a typical large-scale biological data set.</span>
<span id="cb59-307"><a href="#cb59-307" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-308"><a href="#cb59-308" aria-hidden="true" tabindex="-1"></a>We have already seen the NCI60 cancer cell line microarray data set, consisting of 6830 gene expression measurements on 64 cancer cell lines.</span>
<span id="cb59-309"><a href="#cb59-309" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-310"><a href="#cb59-310" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-313"><a href="#cb59-313" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-314"><a href="#cb59-314" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-nci-loaddata</span></span>
<span id="cb59-315"><a href="#cb59-315" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-316"><a href="#cb59-316" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-317"><a href="#cb59-317" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-318"><a href="#cb59-318" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ISLR)</span>
<span id="cb59-319"><a href="#cb59-319" aria-hidden="true" tabindex="-1"></a>nci.labs <span class="ot">&lt;-</span> NCI60<span class="sc">$</span>labs <span class="co"># Sample labels (tissue type)</span></span>
<span id="cb59-320"><a href="#cb59-320" aria-hidden="true" tabindex="-1"></a>nci.data <span class="ot">&lt;-</span> NCI60<span class="sc">$</span>data <span class="co"># Gene expression data set</span></span>
<span id="cb59-321"><a href="#cb59-321" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-322"><a href="#cb59-322" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-323"><a href="#cb59-323" aria-hidden="true" tabindex="-1"></a><span class="in">`nci.data`</span> contains the gene expression level measured on 6830 genes. We can find out the dimension with <span class="in">`dim()`</span> command.</span>
<span id="cb59-324"><a href="#cb59-324" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-325"><a href="#cb59-325" aria-hidden="true" tabindex="-1"></a><span class="in">`nci.labs`</span> contains the labels for the cancer types. Note that some cancer types have more than one cell line, we can get a quick overview using <span class="in">`table()`</span>. This counts how many times each cancer type is present in the labels.</span>
<span id="cb59-326"><a href="#cb59-326" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-327"><a href="#cb59-327" aria-hidden="true" tabindex="-1"></a>Since the focus of today's topic is unsupervised learning, we do not need to pay too much attention to the labels. However we can use the labels in the end to check whether the PCA and clustering produce meaningful results. </span>
<span id="cb59-328"><a href="#cb59-328" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-331"><a href="#cb59-331" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-332"><a href="#cb59-332" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-nci-summary</span></span>
<span id="cb59-333"><a href="#cb59-333" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-334"><a href="#cb59-334" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-335"><a href="#cb59-335" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-336"><a href="#cb59-336" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(nci.data)</span>
<span id="cb59-337"><a href="#cb59-337" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(nci.labs)</span>
<span id="cb59-338"><a href="#cb59-338" aria-hidden="true" tabindex="-1"></a><span class="co"># what if I would like to compute the mean of each gene within each tissue type?</span></span>
<span id="cb59-339"><a href="#cb59-339" aria-hidden="true" tabindex="-1"></a>tissue.means <span class="ot">&lt;-</span> <span class="fu">apply</span>(nci.data, <span class="dv">2</span>, <span class="cf">function</span>(x){<span class="fu">tapply</span>(x, nci.labs, mean)})</span>
<span id="cb59-340"><a href="#cb59-340" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(tissue.means)</span>
<span id="cb59-341"><a href="#cb59-341" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-342"><a href="#cb59-342" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-343"><a href="#cb59-343" aria-hidden="true" tabindex="-1"></a>You can and should carry out some exploratory analysis on the data, such as finding out the means and extreme values; this is left for home practice.</span>
<span id="cb59-344"><a href="#cb59-344" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-345"><a href="#cb59-345" aria-hidden="true" tabindex="-1"></a><span class="fu">### Carry out PCA</span></span>
<span id="cb59-346"><a href="#cb59-346" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-347"><a href="#cb59-347" aria-hidden="true" tabindex="-1"></a>The dataset is a high-dimensional one, that is the number of columns (features) are much larger than the number of rows (measurements). It is inpractical to analyze and search for hidden structures by looking at each one of the feature, and PCA is a useful tool to move forward.</span>
<span id="cb59-348"><a href="#cb59-348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-349"><a href="#cb59-349" aria-hidden="true" tabindex="-1"></a>We use <span class="in">`prcomp()`</span> to produce PCA results. Set the <span class="in">`scale = T`</span> to make sure the data is standardized to have standard deviation equal to 1 among each column.</span>
<span id="cb59-350"><a href="#cb59-350" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-353"><a href="#cb59-353" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-354"><a href="#cb59-354" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-nci-pc</span></span>
<span id="cb59-355"><a href="#cb59-355" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-356"><a href="#cb59-356" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-357"><a href="#cb59-357" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA analysis after scaling the variables to standard deviation one:</span></span>
<span id="cb59-358"><a href="#cb59-358" aria-hidden="true" tabindex="-1"></a>pr.out <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(nci.data, <span class="at">scale=</span><span class="cn">TRUE</span>)</span>
<span id="cb59-359"><a href="#cb59-359" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-360"><a href="#cb59-360" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-361"><a href="#cb59-361" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-362"><a href="#cb59-362" aria-hidden="true" tabindex="-1"></a>Print the summary output with <span class="in">`summary()`</span>. Note that the result is quite long, so we didn't print out everything; but you can execute the line and see what it does.</span>
<span id="cb59-363"><a href="#cb59-363" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb59-366"><a href="#cb59-366" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-367"><a href="#cb59-367" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-nci-pcsummary</span></span>
<span id="cb59-368"><a href="#cb59-368" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-369"><a href="#cb59-369" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-370"><a href="#cb59-370" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb59-371"><a href="#cb59-371" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(pr.out)</span>
<span id="cb59-372"><a href="#cb59-372" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-373"><a href="#cb59-373" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-374"><a href="#cb59-374" aria-hidden="true" tabindex="-1"></a>Now we investigate the variance explained by the PCs. The y-axis in the plots below corresponds to the percentage (from 0 to 100%) explained. </span>
<span id="cb59-375"><a href="#cb59-375" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-378"><a href="#cb59-378" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-379"><a href="#cb59-379" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-nci-plotpc</span></span>
<span id="cb59-380"><a href="#cb59-380" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-381"><a href="#cb59-381" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-382"><a href="#cb59-382" aria-hidden="true" tabindex="-1"></a>pr.var <span class="ot">&lt;-</span> pr.out<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb59-383"><a href="#cb59-383" aria-hidden="true" tabindex="-1"></a>pve <span class="ot">&lt;-</span> pr.var<span class="sc">/</span><span class="fu">sum</span>(pr.var)</span>
<span id="cb59-384"><a href="#cb59-384" aria-hidden="true" tabindex="-1"></a>pve <span class="ot">&lt;-</span> <span class="dv">100</span><span class="sc">*</span>pve <span class="co"># display percentage</span></span>
<span id="cb59-385"><a href="#cb59-385" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-386"><a href="#cb59-386" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb59-387"><a href="#cb59-387" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pve,  <span class="at">type=</span><span class="st">"o"</span>, <span class="at">ylab=</span><span class="st">"PVE (%)"</span>, <span class="at">xlab=</span><span class="st">"Principal Component"</span>, <span class="at">col=</span><span class="st">"blue"</span>)</span>
<span id="cb59-388"><a href="#cb59-388" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">cumsum</span>(pve), <span class="at">type=</span><span class="st">"o"</span>, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">100</span>),<span class="at">ylab=</span><span class="st">"Cumulative PVE (%)"</span>, <span class="at">xlab=</span><span class="st">"Principal Component"</span>, <span class="at">col=</span><span class="st">"brown3"</span>)</span>
<span id="cb59-389"><a href="#cb59-389" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-390"><a href="#cb59-390" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-391"><a href="#cb59-391" aria-hidden="true" tabindex="-1"></a><span class="fu">### How many PCs to keep?</span></span>
<span id="cb59-392"><a href="#cb59-392" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-393"><a href="#cb59-393" aria-hidden="true" tabindex="-1"></a>The maximum number of PC computed in principal component analysis is the $min<span class="sc">\{</span>n, p<span class="sc">\}</span>$. This means, for a high dimensional problem like this one, we have at most 64 PCs. </span>
<span id="cb59-394"><a href="#cb59-394" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-395"><a href="#cb59-395" aria-hidden="true" tabindex="-1"></a>For the purpose of dimensional reduction, we do not need to keep all the PCs. We can check how many PCs explain at least a certain amount of the variance. For example, we found out that the first 24 PC and first 32 PC explained 70% and 80% of the variance. We can mark it on the plots from before.</span>
<span id="cb59-396"><a href="#cb59-396" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-399"><a href="#cb59-399" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-400"><a href="#cb59-400" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-nci-keeppc</span></span>
<span id="cb59-401"><a href="#cb59-401" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-402"><a href="#cb59-402" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-403"><a href="#cb59-403" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-404"><a href="#cb59-404" aria-hidden="true" tabindex="-1"></a>mysel70 <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">cumsum</span>(pve) <span class="sc">&gt;</span> <span class="dv">70</span>)[<span class="dv">1</span>] <span class="co"># explains 70% </span></span>
<span id="cb59-405"><a href="#cb59-405" aria-hidden="true" tabindex="-1"></a>mysel80 <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">cumsum</span>(pve) <span class="sc">&gt;</span> <span class="dv">80</span>)[<span class="dv">1</span>] <span class="co"># explains 80% </span></span>
<span id="cb59-406"><a href="#cb59-406" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(mysel70, mysel80)</span>
<span id="cb59-407"><a href="#cb59-407" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-408"><a href="#cb59-408" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)) <span class="co"># plot contains two smaller plots next to each other</span></span>
<span id="cb59-409"><a href="#cb59-409" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pve,  <span class="at">type=</span><span class="st">"o"</span>, <span class="at">ylab=</span><span class="st">"PVE"</span>, <span class="at">xlab=</span><span class="st">"Principal Component"</span>, <span class="at">col=</span><span class="st">"blue"</span>)</span>
<span id="cb59-410"><a href="#cb59-410" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> mysel80)</span>
<span id="cb59-411"><a href="#cb59-411" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> mysel70, <span class="at">col=</span><span class="dv">3</span>)</span>
<span id="cb59-412"><a href="#cb59-412" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">cumsum</span>(pve), <span class="at">type=</span><span class="st">"o"</span>, <span class="at">ylab=</span><span class="st">"Cumulative PVE"</span>, <span class="at">xlab=</span><span class="st">"Principal Component"</span>, <span class="at">col=</span><span class="st">"brown3"</span>)</span>
<span id="cb59-413"><a href="#cb59-413" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> mysel80)</span>
<span id="cb59-414"><a href="#cb59-414" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> <span class="dv">80</span>)</span>
<span id="cb59-415"><a href="#cb59-415" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> mysel70, <span class="at">col=</span><span class="dv">3</span>)</span>
<span id="cb59-416"><a href="#cb59-416" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> <span class="dv">70</span>, <span class="at">col=</span><span class="dv">3</span>)</span>
<span id="cb59-417"><a href="#cb59-417" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-418"><a href="#cb59-418" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-419"><a href="#cb59-419" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-420"><a href="#cb59-420" aria-hidden="true" tabindex="-1"></a>If we decide to only keep the principal components that explains 70% of the variance, we end up with 24 components, which we can further analyse to better understand the relationships between the variables. For simplicity we only look at the first few components.</span>
<span id="cb59-421"><a href="#cb59-421" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-422"><a href="#cb59-422" aria-hidden="true" tabindex="-1"></a>We plot the first few principal component score vectors, to visualize the results. The observations (cell lines) corresponding to a given cancer type will be plotted in the same colour.</span>
<span id="cb59-423"><a href="#cb59-423" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-426"><a href="#cb59-426" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-427"><a href="#cb59-427" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-nci-plot2</span></span>
<span id="cb59-428"><a href="#cb59-428" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-429"><a href="#cb59-429" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-430"><a href="#cb59-430" aria-hidden="true" tabindex="-1"></a>Cols<span class="ot">=</span><span class="cf">function</span>(vec){</span>
<span id="cb59-431"><a href="#cb59-431" aria-hidden="true" tabindex="-1"></a>  cols<span class="ot">=</span><span class="fu">rainbow</span>(<span class="fu">length</span>(<span class="fu">unique</span>(vec)))</span>
<span id="cb59-432"><a href="#cb59-432" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(cols[<span class="fu">as.numeric</span>(<span class="fu">as.factor</span>(vec))])</span>
<span id="cb59-433"><a href="#cb59-433" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb59-434"><a href="#cb59-434" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-435"><a href="#cb59-435" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the first vs second and first vs third principal component score vectors,</span></span>
<span id="cb59-436"><a href="#cb59-436" aria-hidden="true" tabindex="-1"></a><span class="co"># with colors associated to labels (using the Cols() helper function)</span></span>
<span id="cb59-437"><a href="#cb59-437" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb59-438"><a href="#cb59-438" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pr.out<span class="sc">$</span>x[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="at">col=</span><span class="fu">Cols</span>(nci.labs), <span class="at">pch=</span><span class="dv">19</span>,<span class="at">xlab=</span><span class="st">"PC 1"</span>,<span class="at">ylab=</span><span class="st">" PC 2"</span>)</span>
<span id="cb59-439"><a href="#cb59-439" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pr.out<span class="sc">$</span>x[,<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>)], <span class="at">col=</span><span class="fu">Cols</span>(nci.labs), <span class="at">pch=</span><span class="dv">19</span>,<span class="at">xlab=</span><span class="st">"PC 1"</span>,<span class="at">ylab=</span><span class="st">" PC 3"</span>)</span>
<span id="cb59-440"><a href="#cb59-440" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">'topleft'</span>, <span class="at">col=</span><span class="fu">rainbow</span>(<span class="fu">length</span>(<span class="fu">unique</span>(nci.labs))), <span class="at">legend=</span><span class="fu">unique</span>(nci.labs), <span class="at">bty=</span><span class="st">'n'</span>, <span class="at">lwd=</span><span class="dv">2</span>, <span class="at">cex=</span>.<span class="dv">6</span>)</span>
<span id="cb59-441"><a href="#cb59-441" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-442"><a href="#cb59-442" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-443"><a href="#cb59-443" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-444"><a href="#cb59-444" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb59-445"><a href="#cb59-445" aria-hidden="true" tabindex="-1"></a><span class="fu">## Why not use biplot?</span></span>
<span id="cb59-446"><a href="#cb59-446" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-447"><a href="#cb59-447" aria-hidden="true" tabindex="-1"></a>You might have realized that the plots above are essentially what <span class="in">`biplot`</span> did in the first exercise: plotting PC1, PC2 againt each other. The reason we do not use it here is simple: by default <span class="in">`biplot`</span> displays also the loading vectors (rotation). Given that we have 64 PCs, the final figure would be unreadable. </span>
<span id="cb59-448"><a href="#cb59-448" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-449"><a href="#cb59-449" aria-hidden="true" tabindex="-1"></a>You can try to remove the loading vectors from biplot as an exercise. </span>
<span id="cb59-450"><a href="#cb59-450" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb59-451"><a href="#cb59-451" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-452"><a href="#cb59-452" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-453"><a href="#cb59-453" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-454"><a href="#cb59-454" aria-hidden="true" tabindex="-1"></a><span class="fu">## Exercise 3: Gene expression data</span></span>
<span id="cb59-455"><a href="#cb59-455" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-456"><a href="#cb59-456" aria-hidden="true" tabindex="-1"></a>(CH12Ex13 from statistical learning)</span>
<span id="cb59-457"><a href="#cb59-457" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-458"><a href="#cb59-458" aria-hidden="true" tabindex="-1"></a>Consider again the gene expression data set "Ch12Ex13.csv" (which can be also found on the <span class="co">[</span><span class="ot">book website</span><span class="co">](www.StatLearning.com)</span>) that consists of 40 tissue samples with measurements on 1,000 genes.</span>
<span id="cb59-459"><a href="#cb59-459" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-460"><a href="#cb59-460" aria-hidden="true" tabindex="-1"></a>The first 20 samples are from healthy patients, while the second 20 are from a diseased group.</span>
<span id="cb59-461"><a href="#cb59-461" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-462"><a href="#cb59-462" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-463"><a href="#cb59-463" aria-hidden="true" tabindex="-1"></a>Load in the data using read.csv(). You will need to select header=F.</span>
<span id="cb59-464"><a href="#cb59-464" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-465"><a href="#cb59-465" aria-hidden="true" tabindex="-1"></a>Alternatively: load in the data using "Import dataset" in the upper right window, and click "no" on the "Heading" option.</span>
<span id="cb59-466"><a href="#cb59-466" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-467"><a href="#cb59-467" aria-hidden="true" tabindex="-1"></a>Perform a PCA of these data and visualize the results. </span>
<span id="cb59-468"><a href="#cb59-468" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-469"><a href="#cb59-469" aria-hidden="true" tabindex="-1"></a>Note: remember to check if the variables (genes) are on the columns in the dataset before running the PCA. If they are not: use t() to transform the dataset.</span>
<span id="cb59-470"><a href="#cb59-470" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-471"><a href="#cb59-471" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-472"><a href="#cb59-472" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-475"><a href="#cb59-475" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-476"><a href="#cb59-476" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-ge-loaddata</span></span>
<span id="cb59-477"><a href="#cb59-477" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-478"><a href="#cb59-478" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-479"><a href="#cb59-479" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-480"><a href="#cb59-480" aria-hidden="true" tabindex="-1"></a><span class="co"># set the path to your own!</span></span>
<span id="cb59-481"><a href="#cb59-481" aria-hidden="true" tabindex="-1"></a>exp.data <span class="ot">&lt;-</span>  <span class="fu">read.csv</span>(<span class="st">"data/Ch12Ex13.csv"</span>,<span class="at">header=</span><span class="cn">FALSE</span>)</span>
<span id="cb59-482"><a href="#cb59-482" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-483"><a href="#cb59-483" aria-hidden="true" tabindex="-1"></a><span class="co"># I want each row to represent a sample, and each column a gene</span></span>
<span id="cb59-484"><a href="#cb59-484" aria-hidden="true" tabindex="-1"></a>exp.data <span class="ot">&lt;-</span> <span class="fu">t</span>(exp.data)</span>
<span id="cb59-485"><a href="#cb59-485" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(exp.data)</span>
<span id="cb59-486"><a href="#cb59-486" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-487"><a href="#cb59-487" aria-hidden="true" tabindex="-1"></a><span class="co"># should have n=40 samples/rows, and 1000 columns --&gt; OK!</span></span>
<span id="cb59-488"><a href="#cb59-488" aria-hidden="true" tabindex="-1"></a>groups <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">1</span>,<span class="dv">20</span>), <span class="fu">rep</span>(<span class="dv">2</span>,<span class="dv">20</span>)) <span class="co"># group variable</span></span>
<span id="cb59-489"><a href="#cb59-489" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-490"><a href="#cb59-490" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-491"><a href="#cb59-491" aria-hidden="true" tabindex="-1"></a>Carry out PCA</span>
<span id="cb59-492"><a href="#cb59-492" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-495"><a href="#cb59-495" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-496"><a href="#cb59-496" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-ge-pca</span></span>
<span id="cb59-497"><a href="#cb59-497" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-498"><a href="#cb59-498" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-499"><a href="#cb59-499" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-500"><a href="#cb59-500" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA</span></span>
<span id="cb59-501"><a href="#cb59-501" aria-hidden="true" tabindex="-1"></a>pr.exp <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(exp.data, <span class="at">scale=</span><span class="cn">TRUE</span>)</span>
<span id="cb59-502"><a href="#cb59-502" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-503"><a href="#cb59-503" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot proportion of variance explained</span></span>
<span id="cb59-504"><a href="#cb59-504" aria-hidden="true" tabindex="-1"></a>pr.var <span class="ot">&lt;-</span> pr.exp<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb59-505"><a href="#cb59-505" aria-hidden="true" tabindex="-1"></a>pve <span class="ot">&lt;-</span> pr.var<span class="sc">/</span><span class="fu">sum</span>(pr.var)</span>
<span id="cb59-506"><a href="#cb59-506" aria-hidden="true" tabindex="-1"></a>pve <span class="ot">&lt;-</span> <span class="dv">100</span><span class="sc">*</span>pve</span>
<span id="cb59-507"><a href="#cb59-507" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb59-508"><a href="#cb59-508" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pve, <span class="at">type=</span><span class="st">"o"</span>, <span class="at">ylab=</span><span class="st">"PVE"</span>, <span class="at">xlab=</span><span class="st">"Principal Component"</span>, <span class="at">col=</span><span class="st">"blue"</span>)</span>
<span id="cb59-509"><a href="#cb59-509" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">cumsum</span>(pve), <span class="at">type=</span><span class="st">"o"</span>, <span class="at">ylab=</span><span class="st">"Cumulative PVE"</span>, <span class="at">xlab=</span><span class="st">"Principal Component"</span>, <span class="at">col=</span><span class="st">"red"</span>)</span>
<span id="cb59-510"><a href="#cb59-510" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-511"><a href="#cb59-511" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-512"><a href="#cb59-512" aria-hidden="true" tabindex="-1"></a>Looks like most of the principal components are needed to explain the data well. Maybe we can decide to keep 25-30 components?</span>
<span id="cb59-513"><a href="#cb59-513" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-514"><a href="#cb59-514" aria-hidden="true" tabindex="-1"></a>Can also plot some of the first principal components</span>
<span id="cb59-515"><a href="#cb59-515" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-518"><a href="#cb59-518" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb59-519"><a href="#cb59-519" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pca-ge-plot</span></span>
<span id="cb59-520"><a href="#cb59-520" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb59-521"><a href="#cb59-521" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb59-522"><a href="#cb59-522" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-523"><a href="#cb59-523" aria-hidden="true" tabindex="-1"></a><span class="co"># Remember the use the helper-function to get colours</span></span>
<span id="cb59-524"><a href="#cb59-524" aria-hidden="true" tabindex="-1"></a>Cols<span class="ot">=</span><span class="cf">function</span>(vec){</span>
<span id="cb59-525"><a href="#cb59-525" aria-hidden="true" tabindex="-1"></a>  cols<span class="ot">=</span><span class="fu">rainbow</span>(<span class="fu">length</span>(<span class="fu">unique</span>(vec)))</span>
<span id="cb59-526"><a href="#cb59-526" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(cols[<span class="fu">as.numeric</span>(<span class="fu">as.factor</span>(vec))])</span>
<span id="cb59-527"><a href="#cb59-527" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb59-528"><a href="#cb59-528" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-529"><a href="#cb59-529" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)) <span class="co"># plot-window has two small plots next to each other</span></span>
<span id="cb59-530"><a href="#cb59-530" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pr.exp<span class="sc">$</span>x[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="at">col=</span><span class="fu">Cols</span>(groups), <span class="at">pch=</span><span class="dv">19</span>, <span class="at">xlab=</span><span class="st">"PC 1"</span>, <span class="at">ylab=</span><span class="st">" PC 2"</span>)</span>
<span id="cb59-531"><a href="#cb59-531" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pr.exp<span class="sc">$</span>x[,<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>)], <span class="at">col=</span><span class="fu">Cols</span>(groups), <span class="at">pch=</span><span class="dv">19</span>,<span class="at">xlab=</span><span class="st">"PC 1"</span>,<span class="at">ylab=</span><span class="st">" PC 3"</span>)</span>
<span id="cb59-532"><a href="#cb59-532" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">'topleft'</span>, <span class="at">col=</span><span class="fu">rainbow</span>(<span class="fu">length</span>(<span class="fu">unique</span>(groups))), <span class="at">legend=</span><span class="fu">paste</span>(<span class="st">'group '</span>,<span class="fu">unique</span>(groups),<span class="at">sep=</span><span class="st">''</span>), <span class="at">bty=</span><span class="st">'n'</span>, <span class="at">lwd=</span><span class="dv">2</span>, <span class="at">cex=</span>.<span class="dv">6</span>)</span>
<span id="cb59-533"><a href="#cb59-533" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb59-534"><a href="#cb59-534" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-535"><a href="#cb59-535" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-536"><a href="#cb59-536" aria-hidden="true" tabindex="-1"></a></span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>